---
title: "Count_Matrix_Stats.Filter"
author: "Samuel Gurr"
date: "February 1, 2021"
output: html_document
---

# Setup: 

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
```

### Load libraries
```{r  Load libs, include=FALSE}

library(dplyr)
library(edgeR)
library(cowplot)
library(ggplot2)
library(reshape2)
library(stringr)
library(tidyr)
library(edgeR)
```

### Set working directory
```{r  setup, include=TRUE}
### Set working directory
print(getwd())  # working directory is the scipts folder 
# path for output ting all .csv filtered count files

path = 'C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts/' # personnal computer
# path = 'C:/Users/samuel.gurr/Documents/Github_repositories/Cvriginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts' # work computer



```

# LOAD DATA:
```{r  load_data}

raw_counts_mtx <- read.csv("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/HPC_Analysis/Output/Cvirginica_transcript_count_matrix.csv", sep=',', header=TRUE) # read the output count matrix from sedna
head(raw_counts_mtx) # notice that there are some instances of NAs - prepDE.py3 appeared to skips instances when 
ncol(raw_counts_mtx) -1 # 70 total columns without the transcript ID; 35 total samples (duplicated for sequencing lanes 1 and 2)
# View(raw_counts_mtx)

```

# FORMAT AND EXPORT READ COUNT MATRIX (sum lanes of common sample ID):
```{r  format_export_readcount}
# Edit raw count matrix ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
# two major alterations before starting read filtering...
# (1) convert NAs to 0 
# (2) read counts are parsed between sequencing lanes (L001 and L002) - sum within sample ID

# (1) 
raw_counts_mtx[is.na(raw_counts_mtx)] <- 0 # all instances of NA are now 0 

# (2) 
colnames(raw_counts_mtx) # sequencing lanes can be ommitted from the header so samples can be merged by common ID (sum of gene counts) 
raw_counts_melt <- raw_counts_mtx %>%  # convert the matrix from wide to long format 
                          reshape2::melt(id.vars = "transcript_id",
                                        variable.name = "sample_id", 
                                        value.name = "transcript_count")
raw_counts_melt$sample_id <- gsub("_L00.*","",raw_counts_melt$sample_id) # omit the lane identifier from sample_id


# 'merged' seq lanes - meaning the sum of reads counts within gene and sample ID (no longer L001 and L002 for each sample ID)
merged_raw_counts <- raw_counts_melt %>% # call the long form data
                          mutate(sample_id = as.factor(sub('.[^.]*$', '',raw_counts_melt$sample_id)) ) %>% # grab all of the sample name before the last occurance of a period (i.e. now A1.spat, a2.spat, etc...)
                          mutate(transcript_sample_IDs = paste(transcript_id, sample_id, sep='~')) %>%  # merge the transcript ID and sample Id with unique deliminator '~' 
                          dplyr::group_by(transcript_sample_IDs) %>%  # group by the merged IDs 
                          dplyr::summarise(sum_count = sum(transcript_count)) %>% # sum together the transcript counts by this grouped variable 
                          dplyr::mutate(transcript_id = as.factor(gsub("~.*","",transcript_sample_IDs)) ) %>%  # mutate transcript ID (back from the pasted merge character to group_by)
                          dplyr::mutate(sample_id = as.factor(gsub(".*~","",transcript_sample_IDs)) ) # mutate sample ID (back from the pasted merge character to group_by)
merged_raw_counts <- subset(merged_raw_counts, select = -c(transcript_sample_IDs) ) # omit the merged column 
# dcast back to the wide format
merged_raw_counts_matrix <- merged_raw_counts %>%
                          dcast(transcript_id  ~ sample_id, value.var = "sum_count")

# Write new raw read file
write.csv(merged_raw_counts_matrix,paste(path,"merged_raw_count_matrix.csv", sep =''))


```


# LOAD THE merged_raw_count_matrix (from previous cluster) - edit for mRNA transcripts +  SEQ REFERENCE KEY -FORMAT IT 
# OUTPUT FILES FROM THIS CLUSTER: 'merged_raw_counts_matrix.mRNA' and 'seqIDs_master'


```{r load annotation and experiment metadata}
# merged_raw_counts_matrix.mRNA :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
merged_raw_counts_matrix        <- read.csv("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_counts/merged_raw_count_matrix.csv", sep=',', header=TRUE) # read the output count matrix formatted in the previous cluster..
nrow(merged_raw_counts_matrix) # 67334 total unique transcript IDs
# View(merged_raw_counts_matrix[2]) # tRNA == 'rna-NC_....'  'rna-Trna..'    pseudogenes == 'gene-LOC111.....'  mitochondrion  == 'gene-'      tRNA == 'rna-Trna'


# Note omission of these may be questionable... 
# pseuodogenes share common evolutionary history to the to functional genes with a high similarity - may regulate genes through generation of lncRNAs (long non coding RNAs)
tRNA_pseudogenes               <- c("gene-", "rna-NC_", "rna-Trna")
merged_raw_counts_matrix.mRNA  <- merged_raw_counts_matrix %>% 
                                dplyr::filter( !grepl(paste(tRNA_pseudogenes, collapse="|"), transcript_id)) %>%  # omit all rows with IDs pertaining to tRNA, pseudogenes, and mitchondrion
                                dplyr::select(!X) %>%  # omit the X row number column
                                dplyr::mutate(transcript_id = sub(".*-", "", transcript_id)) # omit 'rna-" from the start of each transcript ID 
nrow(merged_raw_counts_matrix.mRNA) #  66071 total transcripts - without pseuodogenes, tRNAs and mitchondrion
# View(merged_raw_counts_matrix.mRNA)
# Write new raw read file
write.csv(merged_raw_counts_matrix.mRNA,paste(path,"merged_raw_count_matrix_mRNAs.csv", sep =''))





# seqIDs_master :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

# load seq_id.txt - contains all transcript IDs and putative functoinal annotation 
SeqIDs <- as.data.frame(read.delim2(file = "C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Seq_details/seq_id.txt",header=F)) %>% 
                  dplyr::mutate(fullID = sub(".*>", "", V1)) %>% 
                  dplyr::mutate(TranscriptID = sub("PREDICTED.*","",fullID)) %>% 
                  dplyr::mutate(Function = sub(".*PREDICTED: Crassostrea virginica", "", fullID)) %>% 
                  dplyr::select(!V1)
# column 'GeneIDs' as the LOC... string between parenthesis () in each line 'fullID'
for (i in 1:nrow(SeqIDs)) {  # this takes an obnoxious few minutes to run (not efficient code?) or the objective of just calling another unique identifier for downstream gene calls
parenthesis.strings           <- as.data.frame(list(regmatches(SeqIDs$fullID[i], gregexpr("\\(.*?\\)", SeqIDs$fullID[i])))) %>% # for each row.. create 1 column data frame parsing strings between parenthesis
                                    dplyr::rename("string"=1) %>% # rename the column for clarity
                                    dplyr::filter(grepl('LOC1', string)) %>%  # call only the row containing LOC1 (LOC1... is the gene name) - several occurances of () with other functional annotation
                                    dplyr::mutate(string =  gsub("[\\(\\)]", "",string)) # remove the parenthesis from the now isolated (LOC1...) string
                                    
  SeqIDs$GeneID[i]              <- parenthesis.strings[[1]] # print to row i 
}
# View(SeqIDs) # view the final master data
length(unique(SeqIDs$GeneID)) # 38264 unique gene IDs
# Write new raw read file
write.csv(SeqIDs,paste(path,"../Seq_details/seq_id_master.csv", sep = ''))

```


# Checkpoint to load all necessary data
  * skip to this cluster if the above has already been run 
  
```{r load all data}

SeqIDs             <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Seq_details/seq_id_master.csv", sep=',', header=TRUE) # 

raw.countDataframe <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_counts/merged_raw_count_matrix_mRNAs.csv", sep=',', header=TRUE) %>% dplyr::select(-X) # 

# Reference; sample IDs and corresponding treatments 
Metadata           <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Seq_details/sample_Key.csv", sep=',', header=TRUE) # reference 

```


```{r  experiment data}
#================= #
# ALL TIMEPOINTS:
# call all experiment design treatments as 'exp.data'
exp.data <- Metadata %>% dplyr::select(!c("SampleName","Treatment", "ID")) # we already have treatments as separate columns
colnames(exp.data) # SapleName_readmatrix" "Temperature"  "OA" "Salinity" "Replicate" "Age.days"
write.csv(exp.data,paste(path,"../all.exp.metadata.csv", sep = ''))
#================= #
# DAY 2 
exp.data.d2 <- Metadata %>% dplyr::select(!c("SampleName","Treatment", "ID")) %>% dplyr::filter(Age.days %in% 2) # all data on day 2
nrow(exp.data.d2) # 24 total samples on Day 2
write.csv(exp.data.d2,paste(path,"../day2.exp.data.csv", sep =''))
#================= #
# DAY 18 
exp.data.d18 <- Metadata %>% dplyr::select(!c("SampleName","Treatment", "ID")) %>% dplyr::filter(Age.days %in% 18) # all data on day 2
nrow(exp.data.d18) # 11 total samples on Day 18
write.csv(exp.data.d18,paste(path,"../day18.exp.data.csv", sep =''))

```


# INITALL ASSESSMENT OF THE RAW COUNT MATRIX 
  
  * colsums
  * 

```{r assess teh raw count matrix}
nrow(raw.countDataframe) # 66071 total transcripts
ncol(raw.countDataframe)-1 # 35 samples (not counting gene ID column) 
ncol(raw.countDataframe)-1 == nrow(exp.data) # MUST be true

raw.countMatrix           <- as.matrix(raw.countDataframe[-1])
rownames(raw.countMatrix) <- raw.countDataframe$transcript_id

# READ COUNTS 
dim(raw.countMatrix) # 66071 total genes 35 samples
sum(raw.countMatrix) # 52002750 total read counts 

gene_sums <- data.frame(rowSums(raw.countMatrix))  # all gene.IDs and the sum of unique reads

transcript_sums <- colSums(raw.countMatrix) # sum of reads for each sample
mean(transcript_sums) # 1485793 == average raw read counts for each sample
sd(transcript_sums) # 377305.7 == standard deviation of raw read counts per sample

gene_sums_gtr0 <- rowSums(raw.countMatrix) > 0 # all gene.IDs with at least one unique read
sum(gene_sums_gtr0 == TRUE) # 10862 total genes with unique transcript reads 
( sum(gene_sums_gtr0 == TRUE) / dim(raw.countMatrix)[1] ) *100 # 16.43989 % of genes have a unique mapped read
```


```{r  raw_reads_by_Day}
raw.cts.all <- raw.countMatrix

# ========================================================== 
# DAY 2  raw counts
# ========================================================== 
# About: run dyplr 'antijoin' to call cts columns that match 'Sample.Name' in the data frame 'exp.data.d0'
raw.cts.d2 <- raw.countDataframe[,c(1,na.omit(match(exp.data.d2$SapleName_readmatrix, colnames(raw.countDataframe))))] # call the read count matrix JUST for day2 (omit col mathces to IDs in exp data Day 18)
raw.cts.d2 <- data.frame(raw.cts.d2[,-1], row.names=raw.cts.d2[,1])
raw.cts.d2 <- as.matrix(raw.cts.d2, row.names="transcript_id")
dim(raw.cts.d2) # 66071 transcripts and  24 total samples
paste( (mean(colSums(raw.cts.d2))) , (sd(colSums(raw.cts.d2))), sep ='+/-')  # 1,470,603 +/- 376,821 mean SD read count per sample
sort(colnames(raw.cts.d2)) == sort(exp.data.d2$SapleName_readmatrix) # check if TRUE, means the same as the exp/design dataframe exp.data.d2

# ========================================================== 
# DAY 18  raw counts
# ========================================================== 
raw.cts.d18 <- raw.countDataframe[,c(1,na.omit(match(exp.data.d18$SapleName_readmatrix, colnames(raw.countDataframe))))] # call the read count matrix JUST for day18 (omit col mathces to IDs in exp data Day 18)
raw.cts.d18 <- data.frame(raw.cts.d18[,-1], row.names=raw.cts.d18[,1])
raw.cts.d18  <-as.matrix(raw.cts.d18, row.names="transcript_id")
dim(raw.cts.d18) # 66071 transcripts and  11 total samples
paste( (mean(colSums(raw.cts.d18))) , (sd(colSums(raw.cts.d18))), sep ='+/-')  # 1,540,836 +/- 401,511 mean SD read count per sample
sort(colnames(raw.cts.d18)) == sort(exp.data.d18$SapleName_readmatrix) # check if TRUE, means the same as the exp/design dataframe exp.data.d18

# write csv
write.csv(raw.cts.d2,"C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts/raw.day2.counts.csv") # 'path' called in previous # write .csv section
write.csv(raw.cts.d18,"C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts/raw.day18.counts.csv") # 'path' called in previous # write .csv section

```




# Filter read counts 
About: usiing edgeR function 'cpm' here to filter by counts bper million 

Below I am making two cut-offs (thresholds) as (1) lenient low-end (2) strict/more conservative 
(1) < 3 counts per million in at least 50% of samples 
(2) < 10 counts per million in at least 50% of samples 

Resulting umber of genes summarized in knitr table and output to TagSeq\Analysis\Data\filtered_counts'

### For loop in 1 CPM increments to optimize the filtering threshold 
Output is a grid filture for  days 0, 7, 14, 21 
```{r CPM LOOP - Rarefaction-like appraoch to read filtering threshold}

# Load data from previous ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::;;; #

# Reference; sample IDs and corresponding treatments 
raw.cts.d2  <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts/raw.day2.counts.csv", sep=',', header=TRUE) # day 2

raw.cts.d18  <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts/raw.day18.counts.csv", sep=',', header=TRUE) # day 18

# For loop ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::;;; #
df_total <- data.frame() # start dataframe 
loop.table <- data.frame(matrix(nrow = 2, ncol = 4)) # create dataframe, columns for the number of sampling days
colnames(loop.table)<-c('CPM_threshold', 'Day', 'total_genes', 'mean_reads_per_gene') # names for comuns in the for loop
CPM_loop <- c(0:50) # run the loop for CPM from 1 - 10 (counts per million) 
CPM_loop <- data.frame(CPM_loop)

for (i in 1:nrow(CPM_loop)) {
  
  var <- CPM_loop[i,1]
  
  
  # ========================================================== 
  #
  # DAY 2 
  # ========================================================== 
  cts.d2        <- data.frame(raw.cts.d2[,-1], row.names=raw.cts.d2[,1])
  cts.d2.matrix <- as.matrix(cts.d2, row.names="transcript_id")
  CPM.d2        <- cpm(cts.d2.matrix) # Obtain CPMs (counts oer million) using egdeR
  thresh.d2     <- CPM.d2 > var # Which values in myCPM are greater than 3?
  keep.d2       <- rowSums(thresh.d2) >= (ncol(thresh.d2)/2) # we would like to keep genes that have at least 50% TRUES in each row of thresh
  rowSums(head(thresh.d2)) # Summary of how many TRUEs there are in each row
  table(rowSums(thresh.d2)) # 6880 genes with TRUE in all 36 samples 
  cts.matrix.d2.LOOPED     <- cts.d2.matrix[keep.d2,]
  
  d2_total_genes           <- nrow(cts.matrix.d2.LOOPED)
  d2_mean_reads_per_gene   <- mean( (rowSums(cts.matrix.d2.LOOPED) / ncol(cts.matrix.d2.LOOPED) ) ) 
  
  # ========================================================== 
  #
  # DAY 18 
  # ========================================================== 
  cts.d18        <- data.frame(raw.cts.d18[,-1], row.names=raw.cts.d18[,1])
  cts.d18.matrix <- as.matrix(cts.d18, row.names="transcript_id")
  CPM.d18        <- cpm(cts.d18.matrix) # Obtain CPMs (counts oer million) using egdeR
  thresh.d18     <- CPM.d18 > var # Which values in myCPM are greater than 3?
  keep.d18       <- rowSums(thresh.d18) >= (ncol(thresh.d18)/2) # we would like to keep genes that have at least 50% TRUES in each row of thresh
  rowSums(head(thresh.d18)) # Summary of how many TRUEs there are in each row
  table(rowSums(thresh.d18)) # 6880 genes with TRUE in all 36 samples 
  cts.matrix.d18.LOOPED     <- cts.d18.matrix[keep.d18,]
  
  d18_total_genes           <- nrow(cts.matrix.d18.LOOPED)
  d18_mean_reads_per_gene   <- mean( (rowSums(cts.matrix.d18.LOOPED) / ncol(cts.matrix.d18.LOOPED) ) ) 

  # table  
  loop.table$CPM_threshold           <- c(var, var)
  loop.table$Day                     <- c(2, 18)
  loop.table$total_genes             <- c(d2_total_genes, d18_total_genes)
  loop.table$mean_reads_per_gene   <- c(d2_mean_reads_per_gene, d18_mean_reads_per_gene)
    
  df <- data.frame(loop.table) # name dataframe for this singl e row
  df_total <- rbind(df_total,df) #bind to a cumulative list dataframe
  #print(df_total) # print to monitor progress
}

```

```{r}
DAY2_filterfig <- df_total %>% 
  dplyr::filter(Day %in% 2) %>% 
  tidyr::gather(key = "variable", value = "measurement",
       total_genes, mean_reads_per_gene, -Day) %>% 
  # Start with a usual ggplot2 call:
  ggplot(aes(x = CPM_threshold)) +
  geom_line(aes(y = measurement)) +
  ggtitle("Day 2") +
  geom_vline(xintercept = c(3,5,10), color = 'red') +
  facet_wrap(. ~ variable, scales="free")
# DAY2_filterfig

DAY18_filterfig <- df_total %>% 
  dplyr::filter(Day %in% 18) %>% 
  tidyr::gather(key = "variable", value = "measurement",
       total_genes, mean_reads_per_gene, -Day) %>% 
  # Start with a usual ggplot18 call:
  ggplot(aes(x = CPM_threshold)) +
  geom_line(aes(y = measurement)) +
  ggtitle("Day 18") +
  geom_vline(xintercept = c(3,5,10), color = 'red') +
  facet_wrap(. ~ variable, scales="free")
# DAY18_filterfig



png("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts/CPM_Read_Filter_Curve.png", 2000, 1000, pointsize=20)
plot_grid(DAY2_filterfig, DAY18_filterfig)
dev.off()

```


### Results of Rarefaction-like nonbias approach to read filtering
Rationale: in the previous cluster, CPMs 1 - 50 were run (including 'in 50% of samples' due to the binary initial accliamtion present in half of ALL matrices)
in a for loop to produce a rare-faction-like curve for read filtering. Observe the output "../Data/Filtered_Counts/CPM_Read_Filter_Curve.png"
Figure shows that the 3CPM threshold looks like the best filter for consistant criteria in the folloing:
a) close to the asymmptote when low reads were filtered out
b) all genes have >50 mean read count sample-1
c) 13k-15k genes selected across matrices - ~40% decrease in the total reference genome
### LENIENT CUT-OFF [ 3 CPM IN 50% SAMPLES ]
```{r  5 CPM 50% SAMPLES}

# ========================================================== 
#
# DAY 2  (5 CPM in 50% samples using edgeR)
# ========================================================== 
raw.cts.d2          <- data.frame(raw.cts.d2[,-1], row.names=raw.cts.d2[,1])
raw.countMatrix.d2  <- as.matrix(raw.cts.d2, row.names="transcript_id")
ncol(raw.countMatrix.d2) # 24  samples from just Day 0
nrow(raw.countMatrix.d2) # 66071 total genes
# pre-filtering; genes ommitted if < 5 counts per million reads in 50% of samples
 # view the colSums of our Day0 samples  - notice the read sums are around 1 million
CPM.d2 <- cpm(raw.countMatrix.d2) # Obtain CPMs (counts oer million) using egdeR
head(CPM.d2) # Have a look at the output
thresh.d2 <- CPM.d2 > 5 # Which values in myCPM are greater than 5?
head(thresh.d2) # This produces a logical matrix with TRUEs and FALSES
rowSums(head(thresh.d2)) # Summary of how many TRUEs there are in each row
table(rowSums(thresh.d2)) # 2452 genes with TRUE in all 24 samples; 57498 genes with false in all 24 samples
#     0     1     2     3     4     5     6     7     8     9    10    11    12    13    14    15    16    17    18    19    20    21    22    23    24 
# 57485  1076   571   396   318   274   238   205   179   179   167   163   159   158   153   146   155   156   142   184   191   212   279   445  2440
keep.d2 <- rowSums(thresh.d2) >= floor(ncol(thresh.d2)/2) # we would like to keep genes that have at least 50% TRUES in each row of thresh
summary(keep.d2) # only 8% of genes passed
#    Mode   FALSE    TRUE 
# logical   61251    4820 
countMatrix.d2.filtered <- raw.countMatrix.d2[keep.d2,] # Subset the rows of countdata to keep the more highly expressed genes
dim(countMatrix.d2.filtered) # 4820 seq  24 samples

hist_1_d2     <- qplot((colSums(countMatrix.d2.filtered)) , geom="histogram")
d2_1_numgenes <- dim(countMatrix.d2.filtered)[1]
d2_1_samp     <- dim(countMatrix.d2.filtered)[2]

# % Genes with annotation == (filtered genes with annotation / filtered genes) * 100 
SeqIDs$TranscriptID <- sub(" ,*","",SeqIDs$TranscriptID) #omits te space so we can match the rownames in countMatrix.d2.filtered
Day2_PercAnnot_5cpm <- paste( ( (nrow(SeqIDs %>% dplyr::filter(TranscriptID  %in% (rownames(countMatrix.d2.filtered))))) / nrow(SeqIDs) * 100), " %", sep = '') # "7.39362101313321 %"

# ========================================================== 
#
# DAY 18 (5 CPM in 50% samples using edgeR)
# ========================================================== 
raw.cts.d18          <- data.frame(raw.cts.d18[,-1], row.names=raw.cts.d18[,1])
raw.countMatrix.d18  <- as.matrix(raw.cts.d18, row.names="transcript_id")
ncol(raw.countMatrix.d18) # 11  samples from just Day 18
nrow(raw.countMatrix.d18) # 66071 total genes
# pre-filtering; genes ommitted if < 5 counts per million reads in 50% of samples
 # view the colSums of our Day0 samples  - notice the read sums are around 1 million
CPM.d18 <- cpm(raw.countMatrix.d18) # Obtain CPMs (counts oer million) using egdeR
head(CPM.d18) # Have a look at the output
thresh.d18 <- CPM.d18 > 5 # Which values in myCPM are greater than 5?
head(thresh.d18) # This produces a logical matrix with TRUEs and FALSES
rowSums(head(thresh.d18)) # Summary of how many TRUEs there are in each row
table(rowSums(thresh.d18)) # 2300  genes with TRUE in all 11 samples; 57789  genes with false in all 11 samples
#     0     1     2     3     4     5     6     7     8     9    10    11 
# 57789  1711   704   525   406   353   348   316   368   470   781  2300
keep.d18 <- rowSums(thresh.d18) >= floor(ncol(thresh.d18)/2) # we would like to keep genes that have at least 50% TRUES in each row of thresh
summary(keep.d18) # only 8% of genes passed
#    Mode   FALSE    TRUE 
# logical   61135    4936 
countMatrix.d18.filtered <- raw.countMatrix.d18[keep.d18,] # Subset the rows of countdata to keep the more highly expressed genes
dim(countMatrix.d18.filtered) # 4936 fenes  23 samples

hist_1_d18     <- qplot((colSums(countMatrix.d18.filtered)) , geom="histogram")
d18_1_numgenes <- dim(countMatrix.d18.filtered)[1]
d18_1_samp     <- dim(countMatrix.d18.filtered)[2]

# % Genes with annotation == (filtered genes with annotation / filtered genes) * 100 
SeqIDs$TranscriptID <- sub(" ,*","",SeqIDs$TranscriptID) #omits te space so we can match the rownames in countMatrix.d18.filtered
Day18_PercAnnot_5cpm <- paste( ( (nrow(SeqIDs %>% dplyr::filter(TranscriptID  %in% (rownames(countMatrix.d18.filtered))))) / nrow(SeqIDs) * 100), " %", sep = '') # "7.40863039399625 %"




png("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts/Histograms_5cpm_50perc.png", 1000, 1000, pointsize=20)
plot_grid(hist_1_d2, hist_1_d18)
dev.off()


# write csv
write.csv(countMatrix.d2.filtered,"C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts/filtered_counts_5cpm_50perc/day2.filtered_5cpm50perc.csv") # 'path' called in previous # write .csv section

write.csv(countMatrix.d18.filtered,"C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/Filtered_Counts/filtered_counts_5cpm_50perc/day18.counts.filtered_5cpm50perc.csv") # 'path' called in previous # write .csv section

```
