---
title: "GO_Analysis_DESeq2_10cpm"
author: "Samuel Gurr"
date: "February 1, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


NOTE: after runinng this once... you can load the liibraries (below) and skip to 'the 'Find GOslim terms' to load the .csv files for plotting
Load libraries
```{r load_libraries, include = TRUE}
# load libraries - notes show the install command needed to install (pre installed)
library(goseq)
library(dplyr)
library(forcats)
library(ggplot2)
library(gridExtra)
library(tidyr)
library(grDevices)
library(reshape2)
library(Rmisc)
library(ggpubr)
library(tibble)
library(hrbrthemes)
library(gridExtra)
library(tidyr)
library(zoo)
library(ComplexHeatmap)
library(circlize)
library(GSEABase)
library(data.table)
library(stringr)
library(tidyverse)
library(readxl)
```

SETWD & LOAD DATA:
Annotation and DE data (from DESeq2 Rmd script)
```{r Load_annot_and_DE_files, include = TRUE}



# SET WORKING DIRECTORY AND LOAD DATA
setwd("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis")
path_out = 'C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Output/WGCNA/' # personnal computer




# master reference view te master ref R script for details
Master_ref  <- read.csv(file= "Data/TagSeq/Seq_details/Seq_Reference_Master.csv", sep=',', header=TRUE) %>% 
                     dplyr::select(c('Cvirginica_TranscriptID','Annotation_GO_ID', 'Cvirginica_length'))


Cvirginica_annot_reference  <- read.csv(file="Data/TagSeq/Seq_details/seq_id_master.csv", sep=',', header=TRUE) %>% 
                                  dplyr::select(c('TranscriptID','Function','GeneID')) %>% 
                                  dplyr::mutate(TranscriptID = gsub(" ", "", TranscriptID)) %>% # remove the space at the end of each transcript ID
                                  dplyr::mutate(Protein_name = gsub("\\s\\(LOC.*|\\sLOC111.*", "", perl=TRUE, Function)) %>% 
                                  dplyr::select(!Function)


# WGCNA results -fromat and merge with the GO terms
d2_Annot_ModuleMembership      <- read.csv("Output/WGCNA/day2_larvae/d2.WGCNA_ModulMembership.csv") %>% 
                                       dplyr::select(c('TranscriptID','geneSymbol','Protein_name','moduleColor')) %>%  
                                       na.omit() %>% 
                                       dplyr::mutate(Day = "Day2")

d2ModCols                      <- data.frame(moduleColor = unique(d2_Annot_ModuleMembership$moduleColor)) %>% # call all unique module colors 
                                      dplyr:: filter(moduleColor %in% c('pink', 'blue', 'turquoise', 'brown', 'black', 'red')) %>% # sig modules
                                      dplyr::mutate(Day = "Day2")

d18_Annot_ModuleMembership      <- read.csv("Output/WGCNA/day18_spat/d18.WGCNA_ModulMembership.csv") %>% 
                                       dplyr::select(c('TranscriptID','geneSymbol','Protein_name','moduleColor')) %>%  
                                       na.omit() %>% 
                                       dplyr::mutate(Day = "Day18")

d18ModCols                      <- data.frame(moduleColor = unique(d18_Annot_ModuleMembership$moduleColor)) %>% # call all unique module colors 
                                      dplyr:: filter(moduleColor %in% c('green', 'purple', 'tan', 'red', 'blue')) %>% # sig modules
                                      dplyr::mutate(Day = "Day18")




# master WGCNA module data for for loops! 
WGCNA_MasterModData   <-  merge( (as.data.frame(rbind(d2_Annot_ModuleMembership, d18_Annot_ModuleMembership)) %>% 
                                      dplyr::rename(Cvirginica_TranscriptID=TranscriptID)), 
                                   Master_ref, by="Cvirginica_TranscriptID")

WGCNA_ColorList       <-  rbind(d2ModCols, d18ModCols) # master WGCNA color list - use this to loop all the analysis 



#  mapped read counts for each Day (addressed separately in DESeq2)
d2.counts   <- read.csv(file="Data/TagSeq/Filtered_counts/filtered_counts_5cpm_50perc/day2.filtered_5cpm50perc.csv", sep=',', header=TRUE) %>%   
                  dplyr::rename("TranscriptID" = "X")

d18.counts  <- read.csv(file="Data/TagSeq/Filtered_counts/filtered_counts_5cpm_50perc/day18.filtered_5cpm50perc.csv", sep=',', header=TRUE) %>% 
                  dplyr::rename("TranscriptID" = "X")



# GOslim
slim        <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database - # call goslim_generic.obo terms as 'slim'

```


goseq
Prepare dataframe(s) and vectors for goseq 
(1) Format 'GO.term' for goseq from the P.generosa annotation .fna file 'Geoduck_annotation'
```{r GO.terms_data}
# call the GO terms
Cvirginica_GOterms                <- as.data.frame(Master_ref) %>% dplyr::select(c('Cvirginica_TranscriptID','Annotation_GO_ID'))
colnames(Cvirginica_GOterms)[1:2] <- c('transcript.ID', 'GO.terms') # call gene name and the GO terms - (Uniprot ID 'V5')
splitted                          <- strsplit(as.character(Cvirginica_GOterms$GO.terms), ";") #slit into multiple GO ids by delimiter'; ' remember the space after ; is needed here! without this you will only call the first listed GO term for each gene!
GO.terms                          <- data.frame(v1 = rep.int(Cvirginica_GOterms$transcript.ID, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
```


(2) Unique Genes - vector based on all unique mapped reads  - Remember this is based on the genes for EACH sampling day becasue DEeq2 was run separately (i.e. days 0, 7, 14, and 21)
(3) Gene length 
Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
```{r LENGTH VECTORS}
# Prepare dataframe(s) and vectors for goseq 
# Format 'GO.term' for goseq from the P.generosa annotation .fna file 'Geoduck_annotation'
IDvector.d2         <- as.vector(unique(d2.counts$TranscriptID))   # call unique genes (those filtered and used in DESEq2) on day0 - 'IDvector'
IDvector.d18        <- as.vector(unique(d18.counts$TranscriptID))  # call unique genes (those filtered and used in DESEq2) on day7 - 'IDvector'
GO_unique.genes.all <- as.vector(unique(Master_ref$Cvirginica_TranscriptID)) # call all unique genes for GO analysis (goseq)

# Gene length 
GO_gene.length  <- unique(Master_ref %>% dplyr::select(c("Cvirginica_TranscriptID","Cvirginica_length")))

# merge length with counts data
length_vector   <- GO_gene.length$Cvirginica_length

GeneLength.d2   <-  dplyr::inner_join(d2.counts, unique((GO_gene.length %>% dplyr::rename(TranscriptID = Cvirginica_TranscriptID)), by='TranscriptID'))

GeneLength.d18  <- dplyr::inner_join(d18.counts, unique((GO_gene.length %>% dplyr::rename(TranscriptID = Cvirginica_TranscriptID)), by='TranscriptID'))

# call length values for goseq - confirms that the IDvector and length_vector are the same!!!
length_vector.d2 <- GeneLength.d2$Cvirginica_length    # length vector for all unique reads address in DESeq2 on day 0 
sum(sapply(length_vector.d2,length)) == dim(d2.counts)[1] #should be TRUE

length_vector.d18 <- GeneLength.d18$Cvirginica_length    # length vector for all unique reads address in DESeq2 on day 7
sum(sapply(length_vector.d18,length)) == dim(d18.counts)[1] #should be TRUE

```


It's goseq time!!!
Calculate Probability Weighting Function
```{r RUN GOSEQ}

# =================================================================================================
# FOR LOOP goseq
# for loop for all goseq analysis - if/else to loop through the day 7, 14, and 21 WGCNA modules separately 
# Objectives: 
# - run goseq - go enrichment analysis calling just those with adj P value < 0.05
# - run GOslim
# - plot all modules (within each sampling day) with heatmaos fshowing the number of genes within each GOslim
#
# =================================================================================================


for (i in 1:nrow(WGCNA_ColorList)) {
        if (WGCNA_ColorList[i,2] == "Day2") {
        Mod <- d2_Annot_ModuleMembership %>% dplyr::filter(moduleColor %in% WGCNA_ColorList[i,1]) # call the WGCNA Day - essential here!
        Modgenes    <- Mod[1] # TranscriptID
        Mod_integer <- as.integer(GO_unique.genes.all %in% (Modgenes$TranscriptID)) # w/o day-specific ID vector
        names(Mod_integer)=GO_unique.genes.all # rename
        
        pwf       <- nullp(Mod_integer,    id=GO_unique.genes.all, bias.data=length_vector) # make figure margins large enough for this to run...

        goseq     <- goseq(pwf, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
        
        GO.05           <- data.frame(goseq$category[goseq$over_represented_pvalue<.05]) # change twice here
        colnames(GO.05) <- c("category")
        GO.05           <- merge(GO.05, goseq, by="category") # change here
        GO.05           <- GO.05[order(GO.05$ontology, GO.05$over_represented_pvalue,-GO.05$numDEInCat),] %>% 
                              dplyr::mutate(term = as.factor(term)) %>% 
                              dplyr::mutate(moduleColor = WGCNA_ColorList[i,1]) %>% 
                              dplyr::mutate(Day = 'Day2')

        # remove Biological Process GO terms with < 10 genesand  Molecular Function terms with < 2 genes in the module (with that term)
        GO.05_filtered <- GO.05 %>% filter(!(numDEInCat<10 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
        
        # write csv
        write.csv(GO.05, file = paste("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Output/WGCNA/day2_larvae/GO_analysis/GO.05",WGCNA_ColorList[i,1], "Module_unfiltered.csv", sep ='')) # save csv
        
        write.csv(GO.05_filtered, file = paste("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Output/WGCNA/day2_larvae/GO_analysis/GO.05",WGCNA_ColorList[i,1], "Module.csv", sep ='')) # save csv 
        
        #print as we go
        print(paste(WGCNA_ColorList[i,2], WGCNA_ColorList[i,1], "done", sep = ' '))
        
        
        } else if (WGCNA_ColorList[i,2] == "Day18") {
        Mod      <- d18_Annot_ModuleMembership %>% dplyr::filter(moduleColor %in% WGCNA_ColorList[i,1]) # call the WGCNA Day - essential here!
        Modgenes    <- Mod[1] # TranscriptID
        Mod_integer <- as.integer(GO_unique.genes.all %in% (Modgenes$TranscriptID)) # w/o day-specific ID vector
        names(Mod_integer)=GO_unique.genes.all # rename
        
        pwf       <- nullp(Mod_integer,    id=GO_unique.genes.all, bias.data=length_vector) # make figure margins large enough for this to run...

        goseq     <- goseq(pwf, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
        
        GO.05           <- data.frame(goseq$category[goseq$over_represented_pvalue<.05]) # change twice here
        colnames(GO.05) <- c("category")
        GO.05           <- merge(GO.05, goseq, by="category") # change here
        GO.05           <- GO.05[order(GO.05$ontology, GO.05$over_represented_pvalue,-GO.05$numDEInCat),] %>% 
                              dplyr::mutate(term = as.factor(term)) %>% 
                              dplyr::mutate(moduleColor = WGCNA_ColorList[i,1]) %>% 
                              dplyr::mutate(Day = 'Day18')

        # remove Biological Process GO terms with < 10 genesand  Molecular Function terms with < 2 genes in the module (with that term)
        GO.05_filtered <- GO.05 %>% filter(!(numDEInCat<10 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
       
         # write csv
        write.csv(GO.05, file = paste("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Output/WGCNA/day18_spat/GO_analysis/GO.05",WGCNA_ColorList[i,1], "Module_unfiltered.csv", sep ='')) # save csv 
       
         write.csv(GO.05_filtered, file = paste("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Output/WGCNA/day18_spat/GO_analysis/GO.05",WGCNA_ColorList[i,1], "Module.csv", sep ='')) # save csv 
       
        #print as we go
        print(paste(WGCNA_ColorList[i,2], WGCNA_ColorList[i,1], "done", sep = ' ')) }
             
  
   else {}
  
}
          

```

Prep for goSlim
```{r RUN GOSLIM}
#===================================================================================================
#
#
#  GOslim analysis 
# - data reduction from the many GO terms to more broad functions/processes
#
#
#===================================================================================================
# load the output from the previous for loop
setwd("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis")

d2_GO.05blackModule        <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05blackModule.csv")
d2_GO.05blueModule         <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05blueModule.csv")
d2_GO.05brownModule        <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05brownModule.csv")
d2_GO.05pinkModule         <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05pinkModule.csv")
d2_GO.05redModule          <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05redModule.csv")
d2_GO.05turquoiseModule    <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05turquoiseModule.csv")

d18_GO.05blueModule        <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05blueModule.csv")
d18_GO.05greenModule       <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05greenModule.csv")
d18_GO.05purpleModule      <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05purpleModule.csv")
d18_GO.05redModule         <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05redModule.csv")
d18_GO.05tanModule         <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05tanModule.csv")


# Master ALL WGCNA significant modules with treatment
Master_goseq_results      <- rbind(d2_GO.05blackModule, d2_GO.05blueModule, d2_GO.05brownModule,
                                   d2_GO.05pinkModule,d2_GO.05redModule,d2_GO.05turquoiseModule,
                                   
                                   d18_GO.05blueModule, d18_GO.05greenModule, d18_GO.05purpleModule,
                                   d18_GO.05redModule, d18_GO.05tanModule)
# View(Master_goseq_results)


# call all the module colors and days to loop GOslim analysis 
GOslimLoop_vars <- unique(Master_goseq_results[c(9,10)]) 







#  RUN GO SLIM FOR LOOP! 

for (i in 1:nrow(GOslimLoop_vars)) {
  # call the target dataset
  goseq_res       <- Master_goseq_results %>%  dplyr::filter(Day %in% GOslimLoop_vars[i,2], moduleColor %in% GOslimLoop_vars[i,1])
  WGCNA_res       <- WGCNA_MasterModData  %>%  dplyr::filter(Day %in% GOslimLoop_vars[i,2], moduleColor %in% GOslimLoop_vars[i,1])
  
  transcript_ids  <- WGCNA_res$Cvirginica_TranscriptID

  
  # Biological Function - run GOslim
  goseq_res_BP        <- goseq_res %>% dplyr::mutate(category = as.character(category)) %>%  filter(ontology=="BP") # BP - all GO terms upregulated
  BP_GOcollection     <- GOCollection(goseq_res_BP$category)
  GOslims_BP          <- data.frame(goSlim(BP_GOcollection, slim, "BP")) #Find common parent terms to slim down our list
  GOslims_BP$category <- row.names(GOslims_BP) #save rownames as category
  
  # Molecular Function - run GOslim
  goseq_res_MF        <- goseq_res %>%  filter(ontology=="MF") # BP - all GO terms upregulated
  MF_GOcollection     <- GOCollection(goseq_res_MF$category)
  GOslims_MF          <- data.frame(goSlim(MF_GOcollection, slim, "MF")) #Find common parent terms to slim down our list
  GOslims_MF$category <- row.names(GOslims_MF) #save rownames as category
  
  # Get mapped terms - add to the GOslims datatable 
  # from Sam White's Biostars [post](https://support.bioconductor.org/p/128407/#128409).
  # Write function mappedIds to get the query terms that mapped to the slim categories 
  # ...in other words, add a column to your slim dataframe with all the GO terms from goseq
  mappedIds <-  function(df, collection, OFFSPRING) {  #the command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
    map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
    mapped <- lapply(map, intersect, ids(collection)) #Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) #Add column "go_terms" with matching terms 
    df #show resulting dataframe
  }
   
  BPslim_Mapped <- mappedIds(GOslims_BP, BP_GOcollection, GOBPOFFSPRING)
  MFslim_Mapped <- mappedIds(GOslims_MF, MF_GOcollection, GOMFOFFSPRING)
  

  
            # BIOLOGICAL PROCESS ================================================================================================================================= #
  
  
            BPslim             <- filter(BPslim_Mapped, Term!="biological_process") #filter out empty slims and term "biological process" 
            BPsplitted         <- strsplit(as.character(BPslim$go_terms), ";") #split into multiple GO ids
            BPslim$BPsplitted  <- BPsplitted
            
            for (n in 1:nrow(BPslim)) {
              table       <- data.frame(GOlist = unlist(BPslim[n,6])) # call the BPsplitted column of characters and create a small table to filter
              table       <- unique(table)
                            
              Cvirg_module         <- GO.terms %>% dplyr::filter(v1 %in% transcript_ids) # 'GO.terms' has v1 = transcript ids and v2 = GO terms
              Cvirg_loop           <- Cvirg_module %>% dplyr::filter(v2 %in% table$GOlist) # filter Gene IDs with the GO term
              Cvirg_transcriptIDs  <- Cvirg_loop[-2] # ommit the GO.terms to call unique gene calls 
              Cvirg_transcriptIDs  <- unique(Cvirg_transcriptIDs) # call unique Gene calls (unique genes that had the GO term within each of the GOslim bins)
                            
              BPslim$Gene.Count[[n]] <- nrow(Cvirg_transcriptIDs)  # count of unique GeneIDs in each GOslim bin
              BPslim$Gene.IDs[[n]]   <- vapply((Cvirg_transcriptIDs$v1), paste, collapse = ";", character(1L))
              } # end m in 1:nrow
                          
            BPslim_A <- data.frame(Term = rep.int(BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
            BPslim_B <- merge(BPslim_A, BPslim, by="Term") #Add back counts, term, and category info
            BPslim_C <- unique(setDT(BPslim_B)[order(go_term)], by = "category") #remove duplicate offspring terms, keeping only those in the larger umbrella term (Count number)
            BPslim_final <- BPslim_C[,c(1,2,6,9)]
            colnames(BPslim_final) <- c("slim_term", "slim_cat",  "GO_list",  "Gene_IDs")
            BPslim_final[["Gene_IDs"]] <- vapply(unname(BPslim_final$Gene_IDs), paste, collapse = ";", character(1L)) # convert from a list to simply a charaacter string with ; delimiter
            BPslim_final$module_day <- paste(GOslimLoop_vars[i,2], GOslimLoop_vars[i,1], sep = '_')
                          
                 if (nrow(BPslim_final) >0) {
                 BPslim_GOterm_summary       <- BPslim_final
                 s <- strsplit(BPslim_GOterm_summary$GO_list, split = ";")
                 BPslim_GOterm_summary_2     <- data.frame(slim_term = rep(BPslim_GOterm_summary$slim_term, sapply(s, length)),
                                                                              Gene_IDs  = rep(BPslim_GOterm_summary$Gene_IDs, sapply(s, length)),
                                                                              GO_terms = unlist(s))
                 colnames(goseq_res_BP)[2]   <- 'GO_terms'
                 BPslim_GOterm_summary_final <- merge(goseq_res_BP, BPslim_GOterm_summary_2, by = 'GO_terms')
                                    
                 s_2 <- strsplit(BPslim_GOterm_summary_final$Gene_IDs, split = ";")
                 BPslim_GOterm_gene_annotation <- data.frame(slim_term      = rep(BPslim_GOterm_summary_final$slim_term, sapply(s_2, length)),
                                                             over_represented_pvalue = rep(BPslim_GOterm_summary_final$over_represented_pvalue, sapply(s_2, length)),
                                                             GO_terms       = rep(BPslim_GOterm_summary_final$GO_terms, sapply(s_2, length)),
                                                             term           = rep(BPslim_GOterm_summary_final$term, sapply(s_2, length)),
                                                             ontology       = rep(BPslim_GOterm_summary_final$ontology, sapply(s_2, length)),
                                                             moduleColor    = rep(BPslim_GOterm_summary_final$moduleColor, sapply(s_2, length)),
                                                             TranscriptID   = unlist(s_2))
                 BP_master_gene_reference <- merge(BPslim_GOterm_gene_annotation, Cvirginica_annot_reference, by = "TranscriptID")
                            
                 } else (c(BPslim_GOterm_summary_final = NULL, BP_master_gene_reference = NULL))
  
            
            
                          # MOLECULAR FUNCTION ========================================================================================================= #
                          
            
                          MFslim             <- filter(MFslim_Mapped, Term!="molecular_function") #filter out empty slims and term "biological process" (omitted the Count>=2)
                          MFsplitted         <- strsplit(as.character(MFslim$go_terms), ";") #split into multiple GO ids
                          MFslim$MFsplitted  <- MFsplitted
                          for (m in 1:nrow(MFslim)) {
                            table       <- data.frame(GOlist = unlist(MFslim[m,6])) # call the MFsplitted column of characters and create a small table to filter
                            table       <- unique(table)
                            
                            Cvirg_module         <- GO.terms %>% dplyr::filter(v1 %in% transcript_ids) # 'GO.terms' has v1 = transcript ids and v2 = GO terms
                            Cvirg_loop           <- Cvirg_module %>% dplyr::filter(v2 %in% table$GOlist) # filter Gene IDs with the GO term
                            Cvirg_transcriptIDs  <- Cvirg_loop[-2] # ommit the GO.terms to call unique gene calls 
                            Cvirg_transcriptIDs  <- unique(Cvirg_transcriptIDs) # call unique Gene calls (unique genes that had the GO term within each of the GOslim bins)
                            
                            MFslim$Gene.Count[[m]] <- nrow(Cvirg_transcriptIDs)  # count of unique GeneIDs in each GOslim bin
                            MFslim$Gene.IDs[[m]]   <- vapply((Cvirg_transcriptIDs$v1), paste, collapse = ";", character(1L))
                            } # end m in 1:nrow
                          
                            MFslim_A <- data.frame(Term = rep.int(MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all
                            MFslim_B <- merge(MFslim_A, MFslim, by="Term") #Add back counts, term, and category info
                            MFslim_C <- unique(setDT(MFslim_B)[order(go_term)], by = "category") #remove duplicate offspring terms, keeping only those in the larger umbrella term (Count number)
                            MFslim_final <- MFslim_C[,c(1,2,6,9)]
                            colnames(MFslim_final) <- c("slim_term", "slim_cat",  "GO_list",  "Gene_IDs")
                            MFslim_final[["Gene_IDs"]] <- vapply(unname(MFslim_final$Gene_IDs), paste, collapse = ";", character(1L)) # convert from a list to simply a charaacter string with ; delimiter
                            MFslim_final$module_day <- paste(GOslimLoop_vars[i,2], GOslimLoop_vars[i,1], sep = '_')
                          
                                    if (nrow(MFslim_final) >0) {
                                    MFslim_GOterm_summary       <- MFslim_final
                                    s <- strsplit(MFslim_GOterm_summary$GO_list, split = ";")
                                    MFslim_GOterm_summary_2     <- data.frame(slim_term = rep(MFslim_GOterm_summary$slim_term, sapply(s, length)),
                                                                              Gene_IDs  = rep(MFslim_GOterm_summary$Gene_IDs, sapply(s, length)),
                                                                              GO_terms = unlist(s))
                                    colnames(goseq_res_MF)[2]   <- 'GO_terms'
                                    MFslim_GOterm_summary_final <- merge(goseq_res_MF, MFslim_GOterm_summary_2, by = 'GO_terms')
                                    
                                    s_2 <- strsplit(MFslim_GOterm_summary_final$Gene_IDs, split = ";")
                                    MFslim_GOterm_gene_annotation <- data.frame(slim_term      = rep(MFslim_GOterm_summary_final$slim_term, sapply(s_2, length)),
                                                                                over_represented_pvalue = rep(MFslim_GOterm_summary_final$over_represented_pvalue, sapply(s_2, length)),
                                                                                GO_terms       = rep(MFslim_GOterm_summary_final$GO_terms, sapply(s_2, length)),
                                                                                term           = rep(MFslim_GOterm_summary_final$term, sapply(s_2, length)),
                                                                                ontology       = rep(MFslim_GOterm_summary_final$ontology, sapply(s_2, length)),
                                                                                moduleColor    = rep(MFslim_GOterm_summary_final$moduleColor, sapply(s_2, length)),
                                                                                TranscriptID   = unlist(s_2))
                                    MF_master_gene_reference <- merge(MFslim_GOterm_gene_annotation, Cvirginica_annot_reference, by = "TranscriptID")
                            
                                    } else (c(MFslim_GOterm_summary_final = NULL, MF_master_gene_reference = NULL))
    
    
                            
                            # save GOslim final datasets for BP and MF of each  module in their respective folder(s) by Day
    if (GOslimLoop_vars[i,2] == 'Day2') {
          if (nrow(MFslim_final) > 0) {
                  write.csv(MFslim_final, file = paste("Output/WGCNA/day2_larvae/GO_analysis/GOslim/GOslim_MolFunction_",GOslimLoop_vars[i,1], "Module.csv", sep ='')) # save csv file
                  write.csv(MFslim_GOterm_summary_final, file = paste("Output/WGCNA/day2_larvae/GO_analysis/GOslim/GOterms_and_GOslim_MolFunction_",GOslimLoop_vars[i,1], "Module.csv", sep =''))
                  write.csv(MF_master_gene_reference, file = paste("Output/WGCNA/day2_larvae/GO_analysis/GOslim/GOterms_and_GOslim_MolFunction_",GOslimLoop_vars[i,1], "Module_GENE_REFERENCE.csv", sep ='')) 
            } else { }
          
          
          if (nrow(BPslim_final) > 0) {
                  write.csv(BPslim_final, file = paste("Output/WGCNA/day2_larvae/GO_analysis/GOslim/GOslim_BiolProc_",GOslimLoop_vars[i,1], "Module.csv", sep ='')) # save csv file
                  write.csv(BPslim_GOterm_summary_final, file = paste("Output/WGCNA/day2_larvae/GO_analysis/GOslim/GOterms_and_GOslim_BiolProc_",GOslimLoop_vars[i,1], "Module.csv", sep =''))
                  write.csv(BP_master_gene_reference, file = paste("Output/WGCNA/day2_larvae/GO_analysis/GOslim/GOterms_and_GOslim_BiolProc_",GOslimLoop_vars[i,1], "Module_GENE_REFERENCE.csv", sep =''))         
          } else { }
            
    
      
          
      
      
    } else if (GOslimLoop_vars[i,2] == 'Day18') {
          if (nrow(MFslim_final) > 0) {
                  write.csv(MFslim_final, file = paste("Output/WGCNA/day18_spat/GO_analysis/GOslim/GOslim_MolFunction_",GOslimLoop_vars[i,1], "Module.csv", sep ='')) # save csv file
                  write.csv(MFslim_GOterm_summary_final, file = paste("Output/WGCNA/day18_spat/GO_analysis/GOslim/GOterms_and_GOslim_MolFunction_",GOslimLoop_vars[i,1], "Module.csv", sep =''))
                  write.csv(MF_master_gene_reference, file = paste("Output/WGCNA/day18_spat/GO_analysis/GOslim/GOterms_and_GOslim_MolFunction_",GOslimLoop_vars[i,1], "Module_GENE_REFERENCE.csv", sep =''))
            } else { }
            
      
          if (nrow(BPslim_final) > 0) {
                  write.csv(BPslim_final, file = paste("Output/WGCNA/day18_spat/GO_analysis/GOslim/GOslim_BiolProc_",GOslimLoop_vars[i,1], "Module.csv", sep ='')) # save csv file
                  write.csv(BPslim_GOterm_summary_final, file = paste("Output/WGCNA/day18_spat/GO_analysis/GOslim/GOterms_and_GOslim_BiolProc_",GOslimLoop_vars[i,1], "Module.csv", sep =''))
                  write.csv(BP_master_gene_reference, file = paste("Output/WGCNA/day18_spat/GO_analysis/GOslim/GOterms_and_GOslim_BiolProc_",GOslimLoop_vars[i,1], "Module_GENE_REFERENCE.csv", sep =''))
            } else { }
          
    }  else (NULL)

    print(paste(GOslimLoop_vars[i,2], GOslimLoop_vars[i,1], "done", sep = ' '))
} # end of GOslim for loop! (i in 1:nrow)
 
```

Find only enriched GO terms that are statistically significant at cutoff
```{r - Sig GO terms Day 2 larvae}

# load the output from the previous for loop
setwd("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis")

d2_GO.05blackModule        <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05blackModule.csv")
d2_GO.05blueModule         <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05blueModule.csv")
d2_GO.05brownModule        <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05brownModule.csv")
d2_GO.05pinkModule         <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05pinkModule.csv")
d2_GO.05redModule          <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05redModule.csv")
d2_GO.05turquoiseModule    <- read.csv("Output/WGCNA/day2_larvae/GO_analysis/GO.05turquoiseModule.csv")

d18_GO.05blueModule        <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05blueModule.csv")
d18_GO.05greenModule       <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05greenModule.csv")
d18_GO.05purpleModule      <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05purpleModule.csv")
d18_GO.05redModule         <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05redModule.csv")
d18_GO.05tanModule         <- read.csv("Output/WGCNA/day18_spat/GO_analysis/GO.05tanModule.csv")


# Master ALL WGCNA significant modules with treatment
Master_goseq_results      <- rbind(d2_GO.05blackModule, d2_GO.05blueModule, d2_GO.05brownModule,
                                   d2_GO.05pinkModule,d2_GO.05redModule,d2_GO.05turquoiseModule,
                                   
                                   d18_GO.05blueModule, d18_GO.05greenModule, d18_GO.05purpleModule,
                                   d18_GO.05redModule, d18_GO.05tanModule)

#  for the for loop...
df_total             <- data.frame() # start dataframe 
GO.table             <- data.frame(matrix(nrow = 0, ncol = 11)) # create dataframe to save cumunalitively during for loop
colnames(GO.table)   <- c('Day', 'moduleColor',  'GO_term' , 'GO_ID', 'ontology', 'pvalue',  'Gene_count', 'Gene_ratio', 'Transcript_ID', 'geneSymbol', 'Protein_name') # names for comuns in the for loop

# un the loop
for (i in 1:nrow(WGCNA_ColorList)) {
  Ref.GO.terms       <- WGCNA_MasterModData %>%  
                          dplyr::filter(moduleColor %in% WGCNA_ColorList$moduleColor[i] & Day %in% WGCNA_ColorList$Day[i]) %>% 
                          dplyr::select(-Cvirginica_length) %>% 
                          dplyr::mutate(Annotation_GO_ID = strsplit(Annotation_GO_ID, ";")) %>%  
                          unnest(Annotation_GO_ID) %>% 
                          dplyr::rename(GO_ID = Annotation_GO_ID, Transcript_ID = Cvirginica_TranscriptID)

  Moduel.GO.results  <- Master_goseq_results[-1] %>%  
                          dplyr::filter(!ontology %in% 'CC') %>% 
                          dplyr::select(!'under_represented_pvalue') %>% 
                          dplyr::rename(Gene_count = numDEInCat, GO_term = term) %>% 
                          dplyr::mutate(Gene_ratio = Gene_count/numInCat) %>% 
                          dplyr::select(-'numInCat') %>% 
                          dplyr::filter(moduleColor %in% WGCNA_ColorList$moduleColor[i] & Day %in% WGCNA_ColorList$Day[i]) %>% 
                          dplyr::rename(GO_ID = category, pvalue = over_represented_pvalue)

  Master <- rbind(GO.table, (as.data.frame(merge(Moduel.GO.results, Ref.GO.terms, by = c("GO_ID", 'moduleColor', 'Day')))[,c(3,2,6,1,7,4,5,8,9,10,11)]) )
  
  df       <- data.frame(Master) # name dataframe for this single row
  df_total <- rbind(df_total,df) #bind to a cumulative list dataframe
  print(nrow(df_total)) # print to monitor progress      
  
} 

# write csv
Day2_larvae_master <- df_total %>%  dplyr::filter(Day %in% 'Day2')  %>% dplyr::arrange(moduleColor, ontology, pvalue, GO_term, Protein_name) 
Day18_spat_master  <- df_total %>%  dplyr::filter(Day %in% 'Day18') %>% dplyr::arrange(moduleColor, ontology, pvalue, GO_term, Protein_name) 

write.csv(Day2_larvae_master, file = "Output/WGCNA/day2_larvae/GO_analysis/Day2_GO_Results_Master.csv", row.names = FALSE)
write.csv(Day18_spat_master, file = "Output/WGCNA/day18_spat/GO_analysis/Day18_GO_Results_Master.csv", row.names = FALSE)
```


# plot the data! 
```{r PLOTs} 
setwd("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis")


reorder_within <- function(x, by, within, fun = mean) {
  stats::reorder(x, by, FUN = fun) }


# Day 18 - modules 'blue', 'turquoise', 'brown'  - these have sig main effects of salinity and temperature! 
for (i in 1:nrow(WGCNA_ColorList)) {
DotPlot <-  ggplot(data =unique(df_total %>%
                              dplyr::filter(moduleColor %in% WGCNA_ColorList$moduleColor[i]  & Day %in% WGCNA_ColorList$Day[i]) %>% 
                              dplyr::filter(!ontology %in% NA) %>% 
                              dplyr::mutate(moduleColor_Ontology = paste(moduleColor, ontology, sep = '_')) %>%
                              dplyr::mutate(log_pvalue = -log(pvalue)) %>% 
                              dplyr::arrange(log_pvalue) %>% 
                              dplyr::select(-c('Protein_name', 'Transcript_ID', 'geneSymbol'))), 
                     aes(x = Gene_ratio, y = reorder_within(GO_term, Gene_ratio,moduleColor_Ontology),
                     color = `log_pvalue`, size = Gene_count)) + 
                geom_point() +
                scale_color_gradient(low = "orange", high = "blue") +
                theme_bw() + 
                ylab("") + 
                xlab("") + 
                ggtitle("GO enrichment analysis") + 
                scale_x_continuous(limits=c(0,0.5)) +
                facet_wrap(~moduleColor_Ontology, scales = "free",switch = "y") 

        if (WGCNA_ColorList$Day[i] == 'Day2') {
          pdf(paste0("Output/WGCNA/day2_larvae/GO_analysis/Day2_GO_DotPlot", WGCNA_ColorList$moduleColor[i], ".pdf", sep = ''), width=10, height=6)
          plot(DotPlot)
          dev.off()
            } else {  
              pdf(paste0("Output/WGCNA/day18_spat/GO_analysis/Day18_GO_DotPlot", WGCNA_ColorList$moduleColor[i], ".pdf", sep = ''), width=10, height=6)
              plot(DotPlot)
              dev.off() }
    }



# Day 18 - modules 

Day18_DotPlot <-  ggplot(data =unique(Day18_spat_master %>%
                              dplyr::filter(moduleColor %in% c('blue', 'purple', 'red')) %>% 
                              dplyr::filter(!ontology %in% NA) %>% 
                              dplyr::mutate(moduleColor_Ontology = paste(moduleColor, ontology, sep = '_')) %>%
                              dplyr::mutate(log_pvalue = -log(pvalue)) %>% 
                              dplyr::arrange(log_pvalue) %>% 
                              dplyr::select(-c('Protein_name', 'Transcript_ID', 'geneSymbol'))), 
                     aes(x = Gene_ratio, y = reorder_within(GO_term, Gene_ratio,moduleColor_Ontology),
                     color = `log_pvalue`, size = Gene_count)) + 
                geom_point() +
                scale_color_gradient(low = "orange", high = "blue") +
                theme_bw() + 
                ylab("") + 
                xlab("") + 
                ggtitle("GO enrichment analysis") + 
                scale_x_continuous(limits=c(0,0.5)) +
                facet_wrap(~moduleColor_Ontology, scales = "free",switch = "y") 
pdf(paste0("Output/WGCNA/day18_spat/GO_analysis/Day18_GO_DotPlot.pdf"), width=22, height=10)
plot(Day18_DotPlot)
dev.off()


```



PLOTTING  GOslim --log1- pvalue plots 
PRIMARY  EFFECT
```{r}
# read in the outputs from the previous... (note- we did not run goslim fror Day0 DEGs because there were very very few)

# D7_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day7/primary_effect/GOterms_and_GOslim_BiolProc_Upregulated.csv")
# D7_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day7/primary_effect/GOterms_and_GOslim_BiolProc_Downregulated.csv")
# D7_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day7/primary_effect/GOterms_and_GOslim_MolFunction_Upregulated.csv")
# D7_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day7/primary_effect/GOterms_and_GOslim_MolFunction_Downregulated.csv")
# 
# D14_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOterms_and_GOslim_BiolProc_Upregulated.csv")
# D14_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOterms_and_GOslim_BiolProc_Downregulated.csv")
# D14_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOterms_and_GOslim_MolFunction_Upregulated.csv")
# D14_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOterms_and_GOslim_MolFunction_Downregulated.csv")
# 
# D21_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOterms_and_GOslim_BiolProc_Upregulated.csv")
# D21_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOterms_and_GOslim_BiolProc_Downregulated.csv")
# D21_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOterms_and_GOslim_MolFunction_Upregulated.csv")
# D21_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOterms_and_GOslim_MolFunction_Downregulated.csv")



############################################################################################################## 3
#  Molecular Function  

Master_goseq_results
GOres_Day2_MolFunction <- Master_goseq_results %>% dplyr::filter(ontology =='MF' & Day == '2')

MF_Day2_Plot <- ggplot(data = GOres_Day2_MolFunction,aes(x = dir, y = forcats::fct_rev(term))) + 
                      geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
                      scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(GOres_Day2_MolFunction$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(GOres_Day2_MolFunction$over_represented_pvalue)))),by=2))  +
                      facet_grid(~challenge, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                      theme_bw() + 
                      theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                         panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                         strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                         strip.text.x = element_text(size = 8, face = "bold"),
                                         axis.title.x = element_blank(),
                                         axis.title.y = element_text(size=8),
                                         axis.text = element_text(size = 8), legend.position = "right",
                                         plot.margin = unit(c(0,1,0,0.25), "cm"))+
                      ggtitle('GO analysis Day 2 larvae - Molecular Function') 
MF_Day2_Plot

pdf(paste("../Output/DESeq2/Day2_larva/GO_analysis/Day2_larvae_DEGs_MolFunction.pdf", sep =''), width=8,height=7)
print(MF_Day2_Plot)
dev.off()


############################################################################################################## 3
#  Biological Process    

GOres_Day2_BiolProcess <- Master_goseq_results %>% dplyr::filter(ontology =='BP' & Day == '2') %>%  dplyr::filter(!term %in% 'biological_process') 

BP_Day2_Plot <- ggplot(data = GOres_Day2_BiolProcess,aes(x = dir, y = forcats::fct_rev(term))) + 
                      geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
                      scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(GOres_Day2_MolFunction$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(GOres_Day2_MolFunction$over_represented_pvalue)))),by=2))  +
                      facet_grid(~challenge, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                      theme_bw() + 
                      theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                         panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                         strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                         strip.text.x = element_text(size = 8, face = "bold"),
                                         axis.title.x = element_blank(),
                                         axis.title.y = element_text(size=8),
                                         axis.text = element_text(size = 8), legend.position = "right",
                                         plot.margin = unit(c(0,1,0,0.25), "cm"))+
                      ggtitle('GO analysis Day 2 larvae - Biological Process') 
BP_Day2_Plot

pdf(paste("../Output/DESeq2/Day2_larva/GO_analysis/Day2_larvae_DEGs_BiolProcess.pdf", sep =''), width=8,height=7)
print(BP_Day2_Plot)
dev.off()

############################################################################################################## 3
#  Molecular Function  


GOres_Day18_MolFunction <- Master_goseq_results %>% dplyr::filter(ontology =='MF' & Day == '18')

MF_Day18_Plot <- ggplot(data = GOres_Day18_MolFunction,aes(x = dir, y = forcats::fct_rev(term))) + 
                      geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
                      scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(GOres_Day18_MolFunction$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(GOres_Day18_MolFunction$over_represented_pvalue)))),by=18))  +
                      facet_grid(~challenge, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                      theme_bw() + 
                      theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                         panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                         strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                         strip.text.x = element_text(size = 8, face = "bold"),
                                         axis.title.x = element_blank(),
                                         axis.title.y = element_text(size=8),
                                         axis.text = element_text(size = 8), legend.position = "right",
                                         plot.margin = unit(c(0,1,0,0.185), "cm"))+
                      ggtitle('GO analysis Day 18 larvae - Molecular Function') 
MF_Day18_Plot

pdf(paste("../Output/DESeq2/Day18_spat/GO_analysis/Day18_spat_DEGs_MolFunction.pdf", sep =''), width=8,height=7)
print(MF_Day18_Plot)
dev.off()

############################################################################################################## 3
#  Molecular Function  

GOres_Day18_BiolProcess <- Master_goseq_results %>% dplyr::filter(ontology =='BP' & Day == '18') %>%  dplyr::filter(!term %in% 'biological_process') 

BP_Day18_Plot <- ggplot(data = GOres_Day18_BiolProcess,aes(x = dir, y = forcats::fct_rev(term))) + 
                      geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
                      scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(GOres_Day18_MolFunction$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(GOres_Day18_MolFunction$over_represented_pvalue)))),by=18))  +
                      facet_grid(~challenge, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                      theme_bw() + 
                      theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                         panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                         strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                         strip.text.x = element_text(size = 8, face = "bold"),
                                         axis.title.x = element_blank(),
                                         axis.title.y = element_text(size=8),
                                         axis.text = element_text(size = 8), legend.position = "right",
                                         plot.margin = unit(c(0,1,0,0.185), "cm"))+
                      ggtitle('GO analysis Day 18 spat - Biological Process') 
BP_Day18_Plot


pdf(paste("../Output/DESeq2/Day18_spat/GO_analysis/Day18_spat_DEGs_BiolProcess.pdf", sep =''), width=8,height=7)
print(BP_Day18_Plot)
dev.off()

```


PLOTTING  GOslim --log1- pvalue plots 
SECOND EFFECT AND GROUP EFFECT PLOTS (DAY 7)
```{r}
# read in the outputs from the previous... (note- we did not run goslim fror Day0 DEGs because there were very very few)
# DAY 7 - SECOND TREATMENT EFFECT AMBIENT > MODERATE
# D7_Goslim_secondAvM_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day7/secondAvM/GOterms_and_GOslim_BiolProc_secondAvM_Upregulated.csv") #no upregualted terms! 
D7_Goslim_secondAvM_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day7/second_effect_AvM/GOterms_and_GOslim_BiolProc_secondAvM_Downregulated.csv")
# D7_Goslim_secondAvM_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day7/secondAvM/GOterms_and_GOslim_MolFunction_secondAvM_Upregulated.csv") #no upregualted terms! 
D7_Goslim_secondAvM_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day7/second_effect_AvM/GOterms_and_GOslim_MolFunction_secondAvM_Downregulated.csv")


# DAY 7 - GROUP TREATMENT EFFECT MA > AM
D7_Goslim_MAvAM_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GOterms_and_GOslim_BiolProc_MAvAM_Upregulated.csv") 
D7_Goslim_MAvAM_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GOterms_and_GOslim_BiolProc_MAvAM_Downregulated.csv")
# D7_Goslim_MAvAM_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GOterms_and_GOslim_MolFunction_MAvAM_Upregulated.csv") # no upregualted molecular function terms!
D7_Goslim_MAvAM_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GOterms_and_GOslim_MolFunction_MAvAM_Downregulated.csv")


############################################################################################################## 
# DAY 7 - MA  versus AM - Full model effect 
#  Biological Process Upregulated 
BP_UP_Plot_MAvAM <-ggplot(data = D7_Goslim_MAvAM_UP_BP, aes(x = effect, y = forcats::fct_rev(term))) + 
  geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
  scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(BP_UP_Master$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(BP_UP_Master$over_represented_pvalue)))),by=2))  +
  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                     strip.text.x = element_text(size = 8, face = "bold"),
                     axis.title.x = element_blank(),
                     axis.title.y = element_text(size=8),
                     axis.text = element_text(size = 8), legend.position = "right",
                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
  ggtitle('Biological Process: Day 7 Full model effect MA versus AM') +
  # facet_wrap(~ (substr(slim_term, 1,30)))
  facet_wrap((substr(slim_term, 1,30)) ~., scales="free_y", ncol= 1, strip.position="right")

pdf(paste("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/Biol_Process_Upregulated_DEGs_MAvAM.pdf", sep =''), width=15, height=15)
print(BP_UP_Plot_MAvAM)
dev.off()

#  Biological Process Upregulated 
BP_DOWN_Plot_MAvAM <-ggplot(data = D7_Goslim_MAvAM_DOWN_BP, aes(x = effect, y = forcats::fct_rev(term))) + 
  geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
  scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(BP_UP_Master$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(BP_UP_Master$over_represented_pvalue)))),by=2))  +
  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                     strip.text.x = element_text(size = 8, face = "bold"),
                     axis.title.x = element_blank(),
                     axis.title.y = element_text(size=8),
                     axis.text = element_text(size = 8), legend.position = "right",
                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
  ggtitle('Biological Process: Day 7 Full model effect MA versus AM') +
  # facet_wrap(~ (substr(slim_term, 1,30)))
  facet_wrap((substr(slim_term, 1,30)) ~., scales="free_y", ncol= 1, strip.position="right")

pdf(paste("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/Biol_Process_Downregulated_DEGs_MAvAM.pdf", sep =''), width=15, height=10)
print(BP_DOWN_Plot_MAvAM)
dev.off()


############################################################################################################## 3
#  Molecular Function Upregulated 
MF_DOWN_Plot_MAvAM <-ggplot(data = D7_Goslim_MAvAM_DOWN_MF, aes(x = effect, y = forcats::fct_rev(term))) + 
  geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
  scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(MF_UP_Master$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(MF_UP_Master$over_represented_pvalue)))),by=2))  +
  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                     strip.text.x = element_text(size = 8, face = "bold"),
                     axis.title.x = element_blank(),
                     axis.title.y = element_text(size=8),
                     axis.text = element_text(size = 8), legend.position = "right",
                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
  ggtitle('Molecular Function: Day 7 Full model effect MA versus AM') +
  # facet_wrap(~ (substr(slim_term, 1,30)))
  facet_wrap((substr(slim_term, 1,30)) ~., scales="free_y", ncol= 1, strip.position="right")

pdf(paste("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/Mol_Function_Downregulated_DEGs_MAvAM.pdf", sep =''), width=15, height=10)
print(MF_DOWN_Plot_MAvAM)
dev.off()




```




PLOTTING  GOslim - gene count plots 
```{r}
# read in the outputs from the previous... (note- we did not run goslim fror Day0 DEGs because there were very very few)

D7_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day7/GOslim_BiolProc_Upregulated.csv")
D7_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day7/GOslim_BiolProc_Downregulated.csv")

D7_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day7/GOslim_MolFunction_Upregulated.csv")
D7_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day7/GOslim_MolFunction_Downregulated.csv")


D14_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOslim_BiolProc_Upregulated.csv")
D14_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOslim_BiolProc_Downregulated.csv")

D14_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOslim_MolFunction_Upregulated.csv")
D14_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOslim_MolFunction_Downregulated.csv")



D21_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOslim_BiolProc_Upregulated.csv")
D21_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOslim_BiolProc_Downregulated.csv")

D21_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOslim_MolFunction_Upregulated.csv")
D21_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOslim_MolFunction_Downregulated.csv")



#MASTER FOR PLOTTING 
# biological process upregulated 
BP_UP_Master <- rbind(D7_Goslim_UP_BP, D14_Goslim_UP_BP, D21_Goslim_UP_BP)
BP_UP_Master$module_day <- factor(BP_UP_Master$module_day , levels = c("Day7_Up", "Day14_Up", "Day21_Up"))
# biological process donwregualted 
BP_DOWN_Master <- rbind(D7_Goslim_DOWN_BP, D14_Goslim_DOWN_BP, D21_Goslim_DOWN_BP)
BP_DOWN_Master$module_day <- factor(BP_DOWN_Master$module_day , levels = c("Day7_Down", "Day14_Down", "Day21_Down"))


MF_UP_Master <- rbind(D7_Goslim_UP_MF, D14_Goslim_UP_MF, D21_Goslim_UP_MF)
MF_UP_Master$module_day <- factor(MF_UP_Master$module_day , levels = c("Day7_Up", "Day14_Up", "Day21_Up"))
# biological process donwregualted 
MF_DOWN_Master <- rbind(D7_Goslim_DOWN_MF, D14_Goslim_DOWN_MF, D21_Goslim_DOWN_MF)
MF_DOWN_Master$module_day <- factor(MF_DOWN_Master$module_day , levels = c("Day7_Down", "Day14_Down", "Day21_Down"))


# plots
BP_UP_Plot <- BP_UP_Master %>% dplyr::filter(Gene_count >= 5) %>%  
              ggplot(aes(x = "BP_Upreg", y = forcats::fct_rev(slim_term))) + 
                geom_tile(aes(fill=Gene_count, width = 1)) + 
                scale_fill_gradient(low = "grey95", high = "grey10",limits=c(0, 20), breaks=seq(0,20,by=5)) +
                facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                   panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                   strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                   strip.text.x = element_text(size = 8, face = "bold"),
                                   axis.title.x = element_blank(),
                                   axis.title.y = element_text(size=8),
                                   axis.text = element_text(size = 8), legend.position = "right",
                                   plot.margin = unit(c(0,1,0,0.25), "cm"))+
                ggtitle('GOslim Biological Process: Upregulated DEGs Primary Effect All')

BP_DOWN_Plot <- BP_DOWN_Master %>% dplyr::filter(Gene_count >= 5) %>%  
                ggplot(aes(x = "BP_Downreg", y = forcats::fct_rev(slim_term))) + 
                  geom_tile(aes(fill=Gene_count, width = 1)) + 
                  scale_fill_gradient(low = "grey95", high = "grey10",limits=c(0, 20), breaks=seq(0,20,by=5)) +
                  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                     strip.text.x = element_text(size = 8, face = "bold"),
                                     axis.title.x = element_blank(),
                                     axis.title.y = element_text(size=8),
                                     axis.text = element_text(size = 8), legend.position = "right",
                                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
                  ggtitle('GOslim Biological Process: Downregulated DEGs Primary Effect All')





MF_UP_Plot <-MF_UP_Master %>% dplyr::filter(Gene_count >= 2) %>% 
             ggplot(aes(x = "MF_Upreg", y = forcats::fct_rev(slim_term))) + 
                geom_tile(aes(fill=Gene_count, width = 1)) + 
                scale_fill_gradient(low = "grey95", high = "grey10",limits=c(0, 20), breaks=seq(0,20,by=5)) +
                facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                   panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                   strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                   strip.text.x = element_text(size = 8, face = "bold"),
                                   axis.title.x = element_blank(),
                                   axis.title.y = element_text(size=8),
                                   axis.text = element_text(size = 8), legend.position = "right",
                                   plot.margin = unit(c(0,1,0,0.25), "cm"))+
                ggtitle('GOslim Mol Function: Upregulated DEGs Primary Effect All')

MF_DOWN_Plot <-MF_DOWN_Master %>% dplyr::filter(Gene_count >= 2) %>% 
                 ggplot(aes(x = "MF_Downreg", y = forcats::fct_rev(slim_term))) + 
                  geom_tile(aes(fill=Gene_count, width = 1)) + 
                  scale_fill_gradient(low = "grey95", high = "grey10",limits=c(0, 20), breaks=seq(0,20,by=5)) +
                  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                     strip.text.x = element_text(size = 8, face = "bold"),
                                     axis.title.x = element_blank(),
                                     axis.title.y = element_text(size=8),
                                     axis.text = element_text(size = 8), legend.position = "right",
                                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
                  ggtitle('GOslim Mol Function: Downregulated DEGs Primary Effect All')




# SAVE PLOTS  ------------------------------------------------------------------------------------------------- #
pdf(paste("../Output/GO/DESeq2_goseq/GOslim_PrimEff_BiolProcess.pdf", sep =''), width=20, height=8)
print(ggarrange(BP_UP_Plot, BP_DOWN_Plot,         
                plotlist = NULL,
                ncol = 2,
                nrow = 1,
                labels = NULL))
dev.off() 


pdf(paste("../Output/GO/DESeq2_goseq/GOslim_PrimEff_MolFunction.pdf", sep =''), width=20, height=8)
print(ggarrange(MF_UP_Plot, MF_DOWN_Plot,         
                plotlist = NULL,
                ncol = 2,
                nrow = 1,
                labels = NULL))
dev.off() 

```







prep
```{r prep data}
# add ontology column  ---------------------------------------------------------------------------------------- #
All.UP_BPslim_final$Ontolgy <- "BP_Upregulated_all"
All.UP_MFslim_final$Ontolgy <- "MF_Upregulated_all"

All.DOWN_BPslim_final$Ontolgy <- "BP_Downregulated_all"
All.DOWN_MFslim_final$Ontolgy <- "MF_Downregulated_all"

# BIND DATA TO FACET THE PLOTS BY UP AND DOWN REG   ---------------------------------------------------------- #
# bp
BP_All <- rbind(All.UP_BPslim_final, All.DOWN_BPslim_final) # merge the data by binding rows 
BP_All$Ont <- "BP" # create a new common clumn to call in the plot 
BP_All$Ontolgy <- factor(BP_All$Ontolgy, levels = c("BP_Upregulated_all", "BP_Downregulated_all"))# reorder the facotr level for the facet wrap plot 
BP_All_filtered <- BP_All %>%  dplyr::filter(Gene.Count > 1) # ommit all with gene counts >1
# mf
MF_All <- rbind(All.UP_MFslim_final, All.DOWN_MFslim_final) # merge the data by binding rows 
MF_All$Ont <- "MF" # create a new common clumn to call in the plot 
MF_All$Ontolgy <- factor(MF_All$Ontolgy, levels = c("MF_Upregulated_all", "MF_Downregulated_all")) # reorder the facotr level for the facet wrap plot 
MF_All_filtered <- MF_All %>%  dplyr::filter(Gene.Count > 1) # ommit all with gene counts >1
```

Stacked bar plots -  Gene Counts within GOslim term (NOTE: there are likely duplicates (or more) BETWEEN the GOslim bins - however EACH bin contains UNIQUE genes)
```{r barplots - ALL DEGs PrimEffect} 
# create color palette  --------------------------------------------------------------------------------------- #
library(RColorBrewer)
cbp1 <- c("#999999", "#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7") # color blind pallette
OrangeRed <- brewer.pal(2, "OrRd") 
GreenBlue <- brewer.pal(2, "GnBu") 


# PLOTTING --------------------------------------------------------------------------------------------------- #
BP_Plot_PrimEffect_All <- ggplot(data = BP_All_filtered, aes(x = Ont, y = Gene.Count, fill=forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE))) + # BP plot
                               geom_bar(color = "white",size=1.5,stat="identity") + # call bar and create lines between each slim term
                               scale_fill_manual(values =  rev( colorRampPalette(OrangeRed)( (nrow(BP_All)) ) ) )+ # fill with pallette accomodating the max nmber of slim terms
                               geom_text(aes(label = forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE)), colour = "black", position="stack", size = 7, vjust = 1.25) + # add labels
                               theme_classic() + # classic theme
                               ylim(0, 300) + 
                               theme(legend.position = "none",text = element_text(size=25)) +
                               facet_wrap(~Ontolgy) # facet by the upregulated and downregulated datasets

MF_Plot_PrimEffect_All <- ggplot(data = MF_All_filtered, aes(x = Ont, y = Gene.Count, fill=forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE))) + # MF plot
                               geom_bar(color = "white",size=1.5,stat="identity") + # call bar and create lines between each slim term
                               scale_fill_manual(values =  rev( colorRampPalette(GreenBlue)( (nrow(MF_All)) ) ) )+ # fill with pallette accomodating the max nmber of slim terms
                               geom_text(aes(label = forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE)), colour = "black", position="stack", size = 7, vjust = 1.25) + # add labels
                               theme_classic() + # classic theme
                               ylim(0, 100) +
                               theme(legend.position = "none",text = element_text(size=25)) +
                               facet_wrap(~Ontolgy) # facet by the upregulated and downregulated datasets

# SAVE PLOTS  ------------------------------------------------------------------------------------------------- #
pdf(paste("../Output/GO/DESeq2_goseq/GOslim_All_PrimEff_BP.pdf", sep =''), width=15, height=20)
print(ggarrange(BP_Plot_PrimEffect_All,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()     

pdf(paste("../Output/GO/DESeq2_goseq/GOslim_All_PrimEff_MF.pdf", sep =''), width=15, height=20)
print(ggarrange(MF_Plot_PrimEffect_All,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()     

```

Geomtile Heatmaps
```{r geom_tile heatmaps- ALL DEGs PrimEffect}

# plots
BP_Plot_PrimEffect_All <-ggplot(data = BP_All_filtered, aes(x = Ont, y = slim_term)) + 
                          geom_tile(aes(fill=Gene.Count, width = 1)) + 
                          scale_fill_gradient(low = "thistle1", high = "steelblue4") +
                          facet_grid(~Ontolgy, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                          theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                             panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                             strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                             strip.text.x = element_text(size = 8, face = "bold"),
                                             axis.title.x = element_blank(),
                                             axis.title.y = element_text(size=8),
                                             axis.text = element_text(size = 8), legend.position = "right",
                                             plot.margin = unit(c(0,1,0,0.25), "cm"))+
                          ggtitle('GOslim Biological Process: DESeq2 DEGs PrimEffect')

# plot MF
MF_Plot_PrimEffect_All <-ggplot(data = MF_All_filtered, aes(x = Ont, y = slim_term)) + 
                          geom_tile(aes(fill=Gene.Count, width = 1)) + 
                          scale_fill_gradient(low = "azure2", high = "springgreen4") +
                          facet_grid(~Ontolgy, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                          theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                             panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                             strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                             strip.text.x = element_text(size = 8, face = "bold"),
                                             axis.title.x = element_blank(),
                                             axis.title.y = element_text(size=8),
                                             axis.text = element_text(size = 8), legend.position = "right",
                                             plot.margin = unit(c(0,1,0,0.25), "cm"))+
                          ggtitle('GOslim Molecular Function: DESeq2 DEGs PrimEffect')

pdf(paste("../Output/GO/DESeq2_goseq/GOslim_All_PrimEff_Heatmap.pdf", sep =''), width=20, height=8)
print(ggarrange(BP_Plot_PrimEffect_All, MF_Plot_PrimEffect_All,         
                plotlist = NULL,
                ncol = 2,
                nrow = 1,
                labels = NULL))
dev.off() 
```


Follow-up from the stacked blots above...
What are the genesin the GOslimss of interest
Extract genes in GOslim bins of interest and plots their LFC results (DESeq2)
```{r} 
# prep data for the for loop plotting...
# add column 'DE.direction'
All.DOWN_MFslim$DE.direction  <- "downregulated"
All.DOWN_BPslim$DE.direction  <- "downregulated"
All.UP_MFslim$DE.direction    <- "upregulated"
All.UP_BPslim$DE.direction    <- "upregulated"

# add column 'Experiment.note'
All.DOWN_MFslim$Experiment.note  <- "All_Days_PrimaryEfffect"
All.DOWN_BPslim$Experiment.note  <- "All_Days_PrimaryEfffect"
All.UP_MFslim$Experiment.note    <- "All_Days_PrimaryEfffect"
All.UP_BPslim$Experiment.note    <- "All_Days_PrimaryEfffect"

# ommit rownames
rownames(All.DOWN_MFslim)<-NULL
rownames(All.DOWN_BPslim)<-NULL
rownames(All.UP_MFslim)  <-NULL
rownames(All.UP_BPslim)  <-NULL

# merge data frames to build a master GOslim
Master_GOslim <- rbind(All.DOWN_MFslim[,c(10,7,9,8,3,4,5)], All.DOWN_BPslim[,c(10,7,9,8,3,4,5)], All.UP_MFslim[,c(10,7,9,8,3,4,5)], All.UP_BPslim[,c(10,7,9,8,3,4,5)]) # merge for a master table with rbind

# load DESeq2 results 
All_DEGs_Primary.Ambieant_v_Moderate<-read.csv("C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/10cpm/DE_PrimaryTreatment.All.csv")
All_DEGs_Primary.Ambieant_v_Moderate <- All_DEGs_Primary.Ambieant_v_Moderate[2:(ncol(All_DEGs_Primary.Ambieant_v_Moderate))] # ommit the first column - DESEq2 results data 
names(All_DEGs_Primary.Ambieant_v_Moderate)[1] <- 'Gene.ID' # rename the column 

Pgen_annot <- Geoduck_annotation[,c(1,7)]# call only the gene ID and gene names in the annotation csv
names(Pgen_annot) <- c('Gene.ID', 'Gene.name') # rename the columns 
Pgen_annot$Gene.name <- str_extract(Pgen_annot$Gene.name, "[^(]+") # call all gene names before the delimiter ( when the EC code is included)

# loop through the gene IDsin each GOslim bin, DESeq2 results heatmap facetted by sampling day in for loop
for (i in 1:nrow(Master_GOslim)) {
  
  genes <- data.frame(geneids = unlist(Master_GOslim[i,4]))
  
  if (Master_GOslim[i,3] == "downregulated") {
  
    DEG_data <- All_DEGs_Primary.Ambieant_v_Moderate %>% dplyr::filter(down %in% TRUE)  # filter only downregualted genes
    DEG_data_loop <- DEG_data %>% dplyr::filter(Gene.ID %in% genes$geneids)
    DEG_data_merge <- merge(Pgen_annot, DEG_data_loop, by ="Gene.ID")
    
    DEG_data_merge$Day_Trmt <- str_extract(DEG_data_merge$Day_Trmt, "[^_]+")
    
    Plot <- DEG_data_merge %>% #mutate(Day_Trmt = fct_relevel(Day_Trmt,  "Day7", "Day14", "Day21")) %>% 
              ggplot(aes(x = "NA", y = Gene.name)) + 
                geom_tile(aes(fill=log2FoldChange, width = 1)) + 
                scale_fill_gradient(low = "orangered3", high = "orange1") +
                facet_grid(~Day_Trmt, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                strip.text.y = element_text(angle=0, size = 11, face = "bold"),
                strip.text.x = element_text(size = 12, face = "bold"),
                axis.title.x = element_blank(),
                axis.title.y = element_text(size=15),
                axis.text = element_text(size = 12), legend.position = "right",
                plot.margin = unit(c(0,1,0,0.25), "cm"))+
                ggtitle(Master_GOslim[i,5])
    pdf(paste("../Output/GO/DESeq2_goseq/All_PrimEff_GOslim/Downregulated/",Master_GOslim[i,5],".pdf", sep =''), width=18, height=10)
    print(Plot)
    dev.off()
  
    } else {
        DEG_data <- All_DEGs_Primary.Ambieant_v_Moderate %>% dplyr::filter(up %in% TRUE)  # filter only downregualted genes
        DEG_data_loop <- DEG_data %>% dplyr::filter(Gene.ID %in% genes$geneids)
        DEG_data_merge <- merge(Pgen_annot, DEG_data_loop, by ="Gene.ID")
        
        DEG_data_merge$Day_Trmt <- str_extract(DEG_data_merge$Day_Trmt, "[^_]+")
        
        Plot <- DEG_data_merge %>% #mutate(Day_Trmt = fct_relevel(Day_Trmt,  "Day7", "Day14", "Day21")) %>% 
                  ggplot(aes(x = "NA", y = Gene.name)) + 
                    geom_tile(aes(fill=log2FoldChange, width = 1)) + 
                    scale_fill_gradient(low = "palegreen1", high = "green4") +
                    facet_grid( ~Day_Trmt, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                    theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                    panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                    strip.text.y = element_text(angle=0, size = 11, face = "bold"),
                    strip.text.x = element_text(size = 12, face = "bold"),
                    axis.title.x = element_blank(),
                    axis.title.y = element_text(size=15),
                    axis.text = element_text(size = 12), legend.position = "right",
                    plot.margin = unit(c(0,1,0,0.25), "cm"))+
                    ggtitle(Master_GOslim[i,5])
        pdf(paste("../Output/GO/DESeq2_goseq/All_PrimEff_GOslim/Upregulated/",Master_GOslim[i,5],".pdf", sep =''), width=18, height=10)
        print(Plot)
        dev.off()
  } # end of if else statement

} # end of for loop


```


```{r Constant.primary.GO table}
GO_Constant.UP_summary <- GO_Constant.UP %>% dplyr::select(c('category','numDEInCat','numInCat', 'term', 'ontology')) %>% dplyr::filter(!is.na(term)) %>% dplyr::filter(numDEInCat > 0)
GO_Constant.UP_summary$DE_dir <- "upregulated"
GO_Constant.DOWN_summary <- GO_Constant.DOWN %>% dplyr::select(c('category','numDEInCat','numInCat', 'term', 'ontology')) %>% dplyr::filter(!is.na(term)) %>% dplyr::filter(numDEInCat > 0)
GO_Constant.DOWN_summary$DE_dir <- "downregulated"

GO_constant.Master <- rbind(GO_Constant.UP_summary,GO_Constant.DOWN_summary)
GO_constant.Master$perc.occurance <- GO_constant.Master$numDEInCat / GO_constant.Master$numInCat *100

#knitr
knitr::kable(GO_constant.Master, caption = "Summary table of constant GO terms Days 7, 14, 21")
# Write csv
write.csv(GO_constant.Master,paste(path_out,"Constant.Primary_GO.terms.csv"))
```



Geom_Segment Plotting: -log1- p alue all primary efffect down regualted and upregulated genes (figure 2B)
# SECOND EFFECT (A v M main effect AND full group effect MA versus AM)
```{r All Upreg and Downreg GO terms in response to priming}

# Build a master list of all genes and GO terms 'Pgen_GOterms2'
Geoduck_annotation <- read.delim2(file="C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Seq_details/Panopea-generosa-genes-annotations.txt", header=F) # Load themaster Pgenerosa gene list 
Pgen_GOterms <- Geoduck_annotation %>% dplyr::select(c('V1','V8')) # select only two columns - those with the gene IDs and those with the GO terms
Pgen_GOterms2 <- strsplit(Pgen_GOterms$V8, split = "; ") # create a string splitting by delimiter '; ' - view the data to see that this separates each GO term entry in the string
Pgen_GOterms2 <- data.frame(gene.ID = rep(Pgen_GOterms$V1, sapply(Pgen_GOterms2, length)), Go.terms = unlist(Pgen_GOterms2)) # create new dataframe 'Pgen_GOterms2' listing genes for each GO term (MUCH longer!)
Pgen_GOterms2 <- na.omit(Pgen_GOterms2) # ommit the NAs  - genes without GO annotation

d7.UP.second_AvM.05    <- read.csv("../Output/GO/DESeq2_goseq/Day7/second_effect_AvM/GO.05.Day7_SecondEffect_AvM_Upregulated.csv")
d7.UP.second_AvM.05$Dir = "Upregulated"
d7.DOWN.second_AvM.05  <- read.csv("../Output/GO/DESeq2_goseq/Day7/second_effect_AvM/GO.05.Day7_SecondEffect_AvM_Downregulated.csv")
d7.DOWN.second_AvM.05$Dir = "Downregulated"

d7.UP.MAvAM.05    <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GO.05.Day7_GroupEffect_MAvAM_Upregulated.csv")
d7.UP.MAvAM.05$Dir = "Upregulated"
d7.DOWN.MAvAM.05  <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GO.05.Day7_GroupEffect_MAvAM_Downregulated.csv")
d7.DOWN.MAvAM.05$Dir = "Downregulated"

# Master ALL DESeq2 significant modules with treatment
Second_goseq_results      <- rbind(d7.UP.second_AvM.05, d7.DOWN.second_AvM.05)
MAvAM_goseq_results      <- rbind(d7.UP.MAvAM.05, d7.DOWN.MAvAM.05)



# PLOTS 
# Second Treatment effect Ambient > Moderate 
AvM_BiolProcess <- Second_goseq_results %>%  dplyr::filter(ontology %in% ('BP')) %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Biological Process: Day 7 Second Treatment Ambient v Moderate") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Dir ~., scales="free_y", ncol= 1, strip.position="right")


AvM_MolFunction <- Second_goseq_results %>%  dplyr::filter(ontology %in% ('MF')) %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Molecular Function: Day 7 Second Treatment Ambient v Moderate") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Dir ~., scales="free_y", ncol= 1, strip.position="right")

pdf(paste("../Output/GO/DESeq2_goseq/Day7_SecondEffect_AvM_plots/GOplot_DESeq2_Day7_AvM_BP.pdf", sep =''), width=10, height=12)
print(ggarrange(AvM_BiolProcess,         
                plotlist = NULL,
                ncol = 1,
                nrow = 2,
            labels = NULL))
dev.off()   

pdf(paste("../Output/GO/DESeq2_goseq/Day7_SecondEffect_AvM_plots/GOplot_DESeq2_Day7_AvM_MF.pdf", sep =''), width=25, height=8)
print(ggarrange(AvM_MolFunction,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()   


# Second Treatment effect Ambient > Moderate 
MVvAM_BiolProcess <- MAvAM_goseq_results %>%  dplyr::filter(ontology %in% ('BP')) %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Biological Process: Day 7 Full Model MA v AM") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Dir ~., scales="free_y", ncol= 1, strip.position="right")


MVvAM_MolFunction <- MAvAM_goseq_results %>%  dplyr::filter(ontology %in% ('MF')) %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Molecular Function: Day 7 Full Model MA v AM") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Dir ~., scales="free_y", ncol= 1, strip.position="right")


pdf(paste("../Output/GO/DESeq2_goseq/Day7_FullModEffect_MAvAM_plots/GOplot_DESeq2_Day7_MAvAM_BP.pdf", sep =''), width=10, height=12)
print(ggarrange(MVvAM_BiolProcess,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()   

pdf(paste("../Output/GO/DESeq2_goseq/Day7_FullModEffect_MAvAM_plots/GOplot_DESeq2_Day7_MAvAM_MF.pdf", sep =''), width=10, height=8)
print(ggarrange(MVvAM_MolFunction,         
                plotlist = NULL,
                ncol = 1,
                nrow = 2,
            labels = NULL))
dev.off()      


```

Geom_Segment Plotting: -log1- p alue all primary efffect down regualted and upregulated genes (figure 2B)
# PRIMARY Effect
```{r All Upreg and Downreg GO terms in response to priming}

# Build a master list of all genes and GO terms 'Pgen_GOterms2'
Geoduck_annotation <- read.delim2(file="C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Seq_details/Panopea-generosa-genes-annotations.txt", header=F) # Load themaster Pgenerosa gene list 
Pgen_GOterms <- Geoduck_annotation %>% dplyr::select(c('V1','V8')) # select only two columns - those with the gene IDs and those with the GO terms
Pgen_GOterms2 <- strsplit(Pgen_GOterms$V8, split = "; ") # create a string splitting by delimiter '; ' - view the data to see that this separates each GO term entry in the string
Pgen_GOterms2 <- data.frame(gene.ID = rep(Pgen_GOterms$V1, sapply(Pgen_GOterms2, length)), Go.terms = unlist(Pgen_GOterms2)) # create new dataframe 'Pgen_GOterms2' listing genes for each GO term (MUCH longer!)
Pgen_GOterms2 <- na.omit(Pgen_GOterms2) # ommit the NAs  - genes without GO annotation



d0.UP.05    <- read.csv("../Output/GO/DESeq2_goseq/Day0/GO.05.Day0_PrimEffect_Upregulated.csv")
d0.UP.05$Day = "Day0"
d0.UP.05$Dir = "Upregulated"
d0.DOWN.05  <- read.csv("../Output/GO/DESeq2_goseq/Day0/GO.05.Day0_PrimEffect_Downregulated.csv")
d0.DOWN.05$Day = "Day0"
d0.DOWN.05$Dir = "Downregulated"

d7.UP.05    <- read.csv("../Output/GO/DESeq2_goseq/Day7/GO.05.Day7_PrimEffect_Upregulated.csv")
d7.UP.05$Day = "Day7"
d7.UP.05$Dir = "Upregulated"
d7.DOWN.05  <- read.csv("../Output/GO/DESeq2_goseq/Day7/GO.05.Day7_PrimEffect_Downregulated.csv")
d7.DOWN.05$Day = "Day7"
d7.DOWN.05$Dir = "Downregulated"

d14.UP.05   <- read.csv("../Output/GO/DESeq2_goseq/Day14/GO.05.Day14_PrimEffect_Upregulated.csv")
d14.UP.05$Day = "Day14"
d14.UP.05$Dir = "Upregulated"
d14.DOWN.05 <- read.csv("../Output/GO/DESeq2_goseq/Day14/GO.05.Day14_PrimEffect_Downregulated.csv")
d14.DOWN.05$Day = "Day14"
d14.DOWN.05$Dir = "Downregulated"

d21.UP.05   <- read.csv("../Output/GO/DESeq2_goseq/Day21/GO.05.Day21_PrimEffect_Upregulated.csv")
d21.UP.05$Day = "Day21"
d21.UP.05$Dir = "Upregulated"
d21.DOWN.05 <- read.csv("../Output/GO/DESeq2_goseq/Day21/GO.05.Day21_PrimEffect_Downregulated.csv")
d21.DOWN.05$Day = "Day21"
d21.DOWN.05$Dir = "Downregulated"

# Master ALL DESeq2 significant modules with treatment
Master_goseq_results      <- rbind(d0.UP.05, d0.DOWN.05, 
                              d7.UP.05, d7.DOWN.05,
                              d14.UP.05, d14.DOWN.05, 
                              d21.UP.05, d21.DOWN.05)
# View(Master_goseq_results)


# call enriched GO terms and plot 
#  UPREGULATED GENES! 
# day 0 UPREGULATED
GO_d0.UP.05.a<-d0.UP.05$category[d0.UP.05$over_represented_pvalue<.05] # change twice here
GO_d0.UP.05<-data.frame(GO_d0.UP.05.a)
colnames(GO_d0.UP.05) <- c("category")
GO_d0.UP.05 <- merge(GO_d0.UP.05, d0.UP.05, by="category") # change here
GO_d0.UP.05 <- GO_d0.UP.05[order(-GO_d0.UP.05$numDEInCat),]
GO_d0.UP.05$term <- as.factor(GO_d0.UP.05$term)

d0_UP_MF <- GO_d0.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd0_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d0_UP'))
d0_UP_BP <- GO_d0.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd0_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d0_UP'))

# day 7 UPREGULATED
GO_d7.UP.05.a<-d7.UP.05$category[d7.UP.05$over_represented_pvalue<.05] # change twice here
GO_d7.UP.05<-data.frame(GO_d7.UP.05.a)
colnames(GO_d7.UP.05) <- c("category")
GO_d7.UP.05 <- merge(GO_d7.UP.05, d7.UP.05, by="category") # change here
GO_d7.UP.05 <- GO_d7.UP.05[order(-GO_d7.UP.05$numDEInCat),]
GO_d7.UP.05$term <- as.factor(GO_d7.UP.05$term)

d7_UP_MF <- GO_d7.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d7_UP'))
d7_UP_BP <- GO_d7.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d7_UP'))

# day 14 UPREGULATED
GO_d14.UP.05.a<-d14.UP.05$category[d14.UP.05$over_represented_pvalue<.05] # change twice here
GO_d14.UP.05<-data.frame(GO_d14.UP.05.a)
colnames(GO_d14.UP.05) <- c("category")
GO_d14.UP.05 <- merge(GO_d14.UP.05, d14.UP.05, by="category") # change here
GO_d14.UP.05 <- GO_d14.UP.05[order(-GO_d14.UP.05$numDEInCat),]
GO_d14.UP.05$term <- as.factor(GO_d14.UP.05$term)

d14_UP_MF <- GO_d14.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d14_UP'))
d14_UP_BP <- GO_d14.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d14_UP'))

# day 21 UPREGULATED
GO_d21.UP.05.a<-d21.UP.05$category[d21.UP.05$over_represented_pvalue<.05] # change twice here
GO_d21.UP.05<-data.frame(GO_d21.UP.05.a)
colnames(GO_d21.UP.05) <- c("category")
GO_d21.UP.05 <- merge(GO_d21.UP.05, d21.UP.05, by="category") # change here
GO_d21.UP.05 <- GO_d21.UP.05[order(-GO_d21.UP.05$numDEInCat),]
GO_d21.UP.05$term <- as.factor(GO_d21.UP.05$term)

d21_UP_MF <- GO_d21.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d21_UP'))
d21_UP_BP <- GO_d21.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d21_UP'))


#======================================================================= #
# bind the rows of MF and BP in the clustered modules (those twith primary treatment effect in same pattern/directionality)
MF_UP <- rbind(d0_UP_MF, d7_UP_MF, d14_UP_MF, d21_UP_MF) 
BP_UP <- rbind(d0_UP_BP, d7_UP_BP, d14_UP_BP, d21_UP_BP)


MF_UP$term <- gsub(",.*$", "", MF_UP$term)
MF_UP$Day = factor(MF_UP$Day, levels = c("Day0", "Day7", "Day14", "Day21"))
MF_UP_plot <- MF_UP  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Molecular Function: Upregulated DEGs Primary Effect (higher by ambient history)") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Day ~., scales="free_y", ncol= 1, strip.position="right")
                  #facet_wrap(~ Day,nrow = 1)

BP_UP$term <- gsub(",.*$", "", BP_UP$term)
BP_UP$Day = factor(BP_UP$Day, levels = c("Day0", "Day7", "Day14", "Day21"))
BP_UP_plot <- BP_UP  %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Biological Process: Upregulated DEGs Primary Effect (higher by ambient history)") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Day ~., scales="free_y", ncol= 1, strip.position="right")



# Downregulated plots 
# day 0 DOWNREGULATED
GO_d0.DOWN.05.a<-d0.DOWN.05$category[d0.DOWN.05$over_represented_pvalue<.05] # change twice here
GO_d0.DOWN.05<-data.frame(GO_d0.DOWN.05.a)
colnames(GO_d0.DOWN.05) <- c("category")
GO_d0.DOWN.05 <- merge(GO_d0.DOWN.05, d0.DOWN.05, by="category") # change here
GO_d0.DOWN.05 <- GO_d0.DOWN.05[order(-GO_d0.DOWN.05$numDEInCat),]
GO_d0.DOWN.05$term <- as.factor(GO_d0.DOWN.05$term)

d0_DOWN_MF <- GO_d0.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd0_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d0_DOWN'))
d0_DOWN_BP <- GO_d0.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd0_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d0_DOWN'))

# day 7 DOWNREGULATED
GO_d7.DOWN.05.a<-d7.DOWN.05$category[d7.DOWN.05$over_represented_pvalue<.05] # change twice here
GO_d7.DOWN.05<-data.frame(GO_d7.DOWN.05.a)
colnames(GO_d7.DOWN.05) <- c("category")
GO_d7.DOWN.05 <- merge(GO_d7.DOWN.05, d7.DOWN.05, by="category") # change here
GO_d7.DOWN.05 <- GO_d7.DOWN.05[order(-GO_d7.DOWN.05$numDEInCat),]
GO_d7.DOWN.05$term <- as.factor(GO_d7.DOWN.05$term)

d7_DOWN_MF <- GO_d7.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d7_DOWN'))
d7_DOWN_BP <- GO_d7.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d7_DOWN'))

# day 14 DOWNREGULATED
GO_d14.DOWN.05.a<-d14.DOWN.05$category[d14.DOWN.05$over_represented_pvalue<.05] # change twice here
GO_d14.DOWN.05<-data.frame(GO_d14.DOWN.05.a)
colnames(GO_d14.DOWN.05) <- c("category")
GO_d14.DOWN.05 <- merge(GO_d14.DOWN.05, d14.DOWN.05, by="category") # change here
GO_d14.DOWN.05 <- GO_d14.DOWN.05[order(-GO_d14.DOWN.05$numDEInCat),]
GO_d14.DOWN.05$term <- as.factor(GO_d14.DOWN.05$term)

d14_DOWN_MF <- GO_d14.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d14_DOWN'))
d14_DOWN_BP <- GO_d14.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d14_DOWN'))

# day 21 DOWNREGULATED
GO_d21.DOWN.05.a<-d21.DOWN.05$category[d21.DOWN.05$over_represented_pvalue<.05] # change twice here
GO_d21.DOWN.05<-data.frame(GO_d21.DOWN.05.a)
colnames(GO_d21.DOWN.05) <- c("category")
GO_d21.DOWN.05 <- merge(GO_d21.DOWN.05, d21.DOWN.05, by="category") # change here
GO_d21.DOWN.05 <- GO_d21.DOWN.05[order(-GO_d21.DOWN.05$numDEInCat),]
GO_d21.DOWN.05$term <- as.factor(GO_d21.DOWN.05$term)

d21_DOWN_MF <- GO_d21.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d21_DOWN'))
d21_DOWN_BP <- GO_d21.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d21_DOWN'))


#======================================================================= #
# bind the rows of MF and BP in the clustered modules (those twith primary treatment effect in same pattern/directionality)
MF_DOWN <- rbind(d0_DOWN_MF, d7_DOWN_MF, d14_DOWN_MF, d21_DOWN_MF) 
BP_DOWN <- rbind(d0_DOWN_BP, d7_DOWN_BP, d14_DOWN_BP, d21_DOWN_BP)


MF_DOWN$term <- gsub(",.*$", "", MF_DOWN$term)
MF_DOWN$Day = factor(MF_DOWN$Day, levels = c("Day0", "Day7", "Day14", "Day21"))
MF_DOWN_plot <- MF_DOWN  %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Molecular Function: Downregulated DEGs Primary Effect (higher by Moderate history)") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Day ~., scales="free_y", ncol= 1, strip.position="right")

BP_DOWN$term <- gsub(",.*$", "", BP_DOWN$term)
BP_DOWN$Day = factor(BP_DOWN$Day, levels = c("Day0", "Day7", "Day14", "Day21"))
BP_DOWN_plot <- BP_DOWN    %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Biological Process: Downregulated DEGs Primary Effect (higher by Moderate history)") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Day ~., scales="free_y", ncol= 1, strip.position="right")

# MOLECULAR FUNCTION GRAPHS OUTPUT 

pdf(paste("../Output/GO/DESeq2_goseq/GOplot_DESeq2_PrimEffect_Upregulated_MF.pdf", sep =''), width=10, height=12)
print(ggarrange(MF_UP_plot,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()      


pdf(paste("../Output/GO/DESeq2_goseq/GOplot_DESeq2_PrimEffect_Downregulated_MF.pdf", sep =''), width=10, height=12)
print(ggarrange(MF_DOWN_plot,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()      


# BIOLOFICAL PROCESS GRAPHS OUTPUT

pdf(paste("../Output/GO/DESeq2_goseq/GOplot_DESeq2_PrimEffect_Upregulated_BP.pdf", sep =''), width=10, height=18)
print(ggarrange(BP_UP_plot,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()      


pdf(paste("../Output/GO/DESeq2_goseq/GOplot_DESeq2_PrimEffect_Downregulated_BP.pdf", sep =''), width=10, height=18)
print(ggarrange(BP_DOWN_plot,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()      


```

Plotting: Day 0
```{r plots_Day0.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d0.UP$category[GO_d0.UP$over_represented_pvalue<.05] # change twice here
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d0.UP, by="category") # change here
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- # 
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d0.DOWN$category[GO_d0.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d0.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d0.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d0.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 0 Upregulated DEGs") +
  geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 0 Downregulated DEGs") +
  geom_label(aes(x = 0.5, y = 1.5, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d0.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day0/GOplot_d0.Primary.pdf", GO.plot_d0.Primary, width = 21, height = 21, units = c("in"))
getwd()
```

Plotting: Day 7
```{r plots_Day7.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d7.UP$category[GO_d7.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d7.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d7.DOWN$category[GO_d7.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d7.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d7.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d7.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 7 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 2, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 7 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 2, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d7.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day7/GOplot_d7.Primary.pdf", GO.plot_d7.Primary, width = 21, height = 21, units = c("in"))
```

Plotting: Day 14
```{r plots_Day14.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d14.UP$category[GO_d14.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d14.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d14.DOWN$category[GO_d14.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d14.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d14.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d14.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 14 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 20, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 14 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 6, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d14.Primary<- grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day14/GOplot_d14.Primary.pdf", GO.plot_d14.Primary, width = 21, height = 21, units = c("in"))
```

Plotting: Day 21
```{r plots_Day21.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d21.UP$category[GO_d21.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d21.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d21.DOWN$category[GO_d21.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d21.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d21.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d21.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 21 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 7, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 21 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 3, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d21.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day21/GOplot_d21.Primary.pdf", GO.plot_d21.Primary, width = 21, height = 21, units = c("in"))
```

Plotting: Constant DEGs Days 7 14 21 affected by Primary treatment history
```{r Constant_All}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_Constant.UP$category[GO_Constant.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_Constant.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
head(UP_MF)

MFplot_UP <- UP_MF %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#00BFC4') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Mol Fxn: Constant DEGs UPREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

#Cellular Processes
CCplot_UP <- UP_CC %>% 
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#7CAE00') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Cell Comp: Constant DEGs UPREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

# Biological Processes 
BPplot_UP <- UP_BP %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#69b3a2') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Biol Proc: Constant DEGs UPREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 


# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_Constant.DOWN$category[GO_Constant.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_Constant.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Function
MFplot_DOWN <- DOWN_MF %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#00BFC4') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Mol Fxn: Constant DEGs DOWNREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% 
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#7CAE00') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Cell Comp: Constant DEGs DOWNREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#69b3a2') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Biol Proc: Constant DEGs DOWNREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 


# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(Contant_upreg)[1] # call num upregulated genes
num_down <-dim(Contant_dwnreg)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Constant Upregulated DEGs") +
  geom_label(aes(x = 1, y = 3, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Constant Downregulated DEGs") +
  geom_label(aes(x = 1, y = 3, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_Constant<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/GO/GOplot_Constant.Primary.pdf", GO.plot_Constant, width = 21, height = 21, units = c("in"))
```















### GO_MWU - not using this for the paper but did a bit of work here... Can review it for future data
Create all files for the GO_MWU.R script
output two-column csv of just the gene.id and log2fc 
```{r GO_MWU files - Primary_Treatment}
DE_d0.P <- data.frame(DE_d0.primary[,c(2,4)])
write.csv(DE_d0.P,"C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d0.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d7.P <- data.frame(DE_d7.primary[,c(2,4)])
write.csv(DE_d7.P,"C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d7.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d14.P <- data.frame(DE_d14.primary[,c(2,4)])
write.csv(DE_d14.P,"C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d14.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d21.P <- data.frame(DE_d21.primary[,c(2,4)])
write.csv(DE_d21.P,"C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d21.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

```

### Genewalk

link to tutorial: https://churchman.med.harvard.edu/genewalk/#tutorial
Citation (bioRxiv preprint): Ietswaart, R., Gyori, B. M., Bachman, J. A., Sorger, P. K., & Churchman, L. S. (2019). GeneWalk identifies relevant gene functions for a biological context using network representation learning. bioRxiv, 755579.
About: Genewalk is curently designed to perform function network analysis of DEGs with human and mouse identifiers

```{r prep genewalk files}
Geoduck.gene_HGNC.list <- Geoduck_annotation %>% dplyr::select(c("V1","V6")) # call the geoduck gene ID and the HGNC in the annotation file
colnames(Geoduck.gene_HGNC.list) <- c("gene.ID", "HGNC_symbol") # rename columns
# note that the HGNC column contains the HNC code with the model taxa as 'HUMAN, MOUSE, CHICK, etc.'
Geoduck.gene_HGNC.list$HGNC<- sapply(strsplit((Geoduck.gene_HGNC.list$HGNC), "_"), '[', 1) # this splits the HGNC by the delimiter "_" and calls just the HGNC code - note the model taxa is now absent from the dataframe! 


# change the col name of the gene.ID  in the DE datasets

# DAY 0 ------------------------------------------- #

# UPREGULATED

GW.d0.primary.UPREG <- merge((DE_d0.primary.UPREG[order(DE_d0.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d0.upreg.percHGNC <- ((dim(na.omit(GW.d0.primary.UPREG))[1]) / (dim(DE_d0.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d0.primary.DWNREG <- merge((DE_d0.primary.DWNREG[order(DE_d0.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d0.downreg.percHGNC <- ((dim(na.omit(GW.d0.primary.DWNREG))[1]) / (dim(DE_d0.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 7 ------------------------------------------- #

# UPREGULATED
GW.d7.primary.UPREG <- merge((DE_d7.primary.UPREG[order(DE_d7.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d7.upreg.percHGNC <- ((dim(na.omit(GW.d7.primary.UPREG))[1]) / (dim(DE_d7.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d7.primary.DWNREG <- merge((DE_d7.primary.DWNREG[order(DE_d7.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d7.downreg.percHGNC <- ((dim(na.omit(GW.d7.primary.DWNREG))[1]) / (dim(DE_d7.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 14 ------------------------------------------- #

# UPREGULATED
GW.d14.primary.UPREG <- merge((DE_d14.primary.UPREG[order(DE_d14.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d14.upreg.percHGNC <- ((dim(na.omit(GW.d14.primary.UPREG))[1]) / (dim(DE_d14.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d14.primary.DWNREG <- merge((DE_d14.primary.DWNREG[order(DE_d14.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC") # sort to order by greatest logfold change and merge with HGNC list
d14.downreg.percHGNC <- ((dim(na.omit(GW.d14.primary.DWNREG))[1]) / (dim(DE_d14.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 21 ------------------------------------------- #

# UPREGULATED
GW.d21.primary.UPREG <- merge((DE_d21.primary.UPREG[order(DE_d21.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d21.upreg.percHGNC <- ((dim(na.omit(GW.d21.primary.UPREG))[1]) / (dim(DE_d21.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d21.primary.DWNREG <- merge((DE_d21.primary.DWNREG[order(DE_d21.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d21.downreg.percHGNC <- ((dim(na.omit(GW.d21.primary.DWNREG))[1]) / (dim(DE_d21.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID


### OUTPUT Genewalk tables as csv wihtout header
GW.d0.primary.UPREG.om   <- na.omit(GW.d0.primary.UPREG)
GW.d0.primary.DWNREG.om  <- na.omit(GW.d0.primary.DWNREG)

GW.d7.primary.UPREG.om   <- na.omit(GW.d7.primary.UPREG)
GW.d7.primary.DWNREG.om  <- na.omit(GW.d7.primary.DWNREG) 

GW.d14.primary.UPREG.om  <- na.omit(GW.d14.primary.UPREG)
GW.d14.primary.DWNREG.om <- na.omit(GW.d14.primary.DWNREG)

GW.d21.primary.UPREG.om  <- na.omit(GW.d21.primary.UPREG)
GW.d21.primary.DWNREG.om <- na.omit(GW.d21.primary.DWNREG)

path_genewalk='C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/Genewalk_files/' # call path
# output files 
write.table(GW.d0.primary.UPREG.om,paste(path_genewalk,"Day0_upreg.csv", sep =''),sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d0.primary.DWNREG.om, paste(path_genewalk,"Day0_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)

write.table(GW.d7.primary.UPREG.om,paste(path_genewalk,"Day7_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d7.primary.DWNREG.om,paste(path_genewalk,"Day7_downreg.csv", sep =''), sep=",", col.names=FALSE, row.names=FALSE, quote=FALSE)

write.table(GW.d14.primary.UPREG.om,paste(path_genewalk,"Day14_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d14.primary.DWNREG.om,paste(path_genewalk,"Day14_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)

write.table(GW.d21.primary.UPREG.om,paste(path_genewalk,"Day21_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d21.primary.DWNREG.om,paste(path_genewalk,"Day21_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
```

```{r Genewalk_knitr_table}
GW_summarytable <- data.frame(matrix(nrow = 4, ncol = 8)) # create a new data table
colnames(GW_summarytable)<-c('Day', 'Treatment', 'DEGs_upreg', 'numHGNC_upreg', 'percHGNC_upreg', 'DEGs_dwnreg', 'numHGNC_dwnreg', 'percHGNC_dwnreg') 
GW_summarytable$Day <- c("0","7","14","21")
GW_summarytable$Treatment <- "Ambient_v_Moderate"
GW_summarytable$DEGs_upreg <- c((dim(DE_d0.primary.UPREG)[1]), (dim(DE_d7.primary.UPREG)[1]), (dim(DE_d14.primary.UPREG)[1]), (dim(DE_d21.primary.UPREG)[1]))
GW_summarytable$numHGNC_upreg <- c((dim(na.omit(GW.d0.primary.UPREG))[1]),(dim(na.omit(GW.d7.primary.UPREG))[1]),(dim(na.omit(GW.d14.primary.UPREG))[1]),(dim(na.omit(GW.d21.primary.UPREG))[1]))
GW_summarytable$percHGNC_upreg <- c(d0.upreg.percHGNC,d7.upreg.percHGNC,d14.upreg.percHGNC,d21.upreg.percHGNC)
GW_summarytable$DEGs_dwnreg <- c((dim(DE_d0.primary.DWNREG)[1]), (dim(DE_d7.primary.DWNREG)[1]), (dim(DE_d14.primary.DWNREG)[1]), (dim(DE_d21.primary.DWNREG)[1]))
GW_summarytable$numHGNC_dwnreg<- c((dim(na.omit(GW.d0.primary.DWNREG))[1]),(dim(na.omit(GW.d7.primary.DWNREG))[1]),(dim(na.omit(GW.d14.primary.DWNREG))[1]),(dim(na.omit(GW.d21.primary.DWNREG))[1]))
GW_summarytable$percHGNC_dwnreg <- c(d0.downreg.percHGNC,d7.downreg.percHGNC,d14.downreg.percHGNC,d21.downreg.percHGNC)

knitr::kable(GW_summarytable, caption = "Summary table of Genewalk files - total DEGs and mapping abund HGNC IDs")
# Write csv
write.csv(GW_summarytable,paste(path_genewalk,"Summary_HGNC_hits.csv"))
```

