---
title: "GO_Analysis_DESeq2_10cpm"
author: "Samuel Gurr"
date: "February 1, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


NOTE: after runinng this once... you can load the liibraries (below) and skip to 'the 'Find GOslim terms' to load the .csv files for plotting
Load libraries
```{r load_libraries, include = TRUE}
# load libraries - notes show the install command needed to install (pre installed)
library(goseq)
library(dplyr)
library(forcats)
library(ggplot2)
library(gridExtra)
library(tidyr)
library(grDevices)
library(reshape2)
library(Rmisc)
library(ggpubr)
library(tibble)
library(hrbrthemes)
library(gridExtra)
library(tidyr)
library(zoo)
library(ComplexHeatmap)
library(circlize)
library(GSEABase)
library(data.table)
library(stringr)
library(tidyverse)
library(readxl)
```

Set working directory and outpath
```{r set_wd, include = TRUE}

# SET WORKING DIRECTORY AND LOAD DATA
setwd("C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/")
path_out = 'C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Output/DESeq2/' # personnal computer

```

LOAD DATA:
Annotation and DE data (from DESeq2 Rmd script)
```{r Load_annot_and_DE_files, include = TRUE}


# DE files - all genes and filtered raw counts
## main pairwise effect DEGs (salinity, temperature, OA - high vs. low for each)
DE_d2.Temperature <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/DESeq2/Day2_larva/Day2.Temperature_DESeq2results.csv", sep=',', header=TRUE)
DE_d2.Salinity    <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/DESeq2/Day2_larva/Day2.Salinity_DESeq2results.csv", sep=',', header=TRUE)
DE_d2.OA          <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/DESeq2/Day2_larva/Day2.OA_DESeq2results.csv", sep=',', header=TRUE)
colnames(DE_d2.Temperature)[2]  == "TranscriptID" # TRUE
colnames(DE_d2.Salinity)[2]     == "TranscriptID" # TRUE
colnames(DE_d2.OA)[2]           == "TranscriptID" # TRUE

DE_d18.Salinity    <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/DESeq2/Day18_spat/Day18.salinity_DESeq2results.csv", sep=',', header=TRUE)
DE_d18.OA          <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Data/TagSeq/DESeq2/Day18_spat/Day18.OA_DESeq2results.csv", sep=',', header=TRUE)
colnames(DE_d18.Salinity)[2]     == "TranscriptID" # TRUE
colnames(DE_d18.OA)[2]           == "TranscriptID" # TRUE

#  mapped read counts for each Day (addressed separately in DESeq2)
d2.counts   <- read.csv(file="../Data/TagSeq/Filtered_counts/filtered_counts_5cpm_50perc/day2.filtered_5cpm50perc.csv", sep=',', header=TRUE) %>% dplyr::rename("TranscriptID" = "X")
d18.counts  <- read.csv(file="../Data/TagSeq/Filtered_counts/filtered_counts_5cpm_50perc/day18.counts.filtered_5cpm50perc.csv", sep=',', header=TRUE) %>% dplyr::rename("TranscriptID" = "X")

# annotation master data 
# first the crassotrea virginica reference - containing the transcript IDs and functional annotation (nOTE: no KEGG or GO term-level annotation in this file!)
Cvirginica_annot_reference  <- read.csv(file="../Data/TagSeq/Seq_details/seq_id_master.csv", sep=',', header=TRUE) %>% 
                                dplyr::select(c('TranscriptID','Function','GeneID')) %>% 
                                dplyr::mutate(TranscriptID = gsub(" ", "", TranscriptID)) %>% # remove the space at the end of each transcript ID
                                dplyr::mutate(Protein_name = gsub("\\s\\(LOC.*|\\sLOC111.*", "", perl=TRUE, Function)) %>% 
                                dplyr::select(!Function)
nrow(Cvirginica_annot_reference) #66625 including all genes
nrow(Cvirginica_annot_reference %>% dplyr::filter(grepl('uncharacterized', Protein_name))) #27164 EXCLUDING protein names labeled  as 'uncharacterized'


# GO terms: upload bpth the Cvirginica and the C gigas GO terms

Cvirg_genome3.0  <- read.delim2(file="../Data/TagSeq/Seq_details/GCF_002022765.2_C_virginica-3.0_genomic.txt", skip = 4, header=FALSE) %>%
                        dplyr::select(c('V3','V4','V5','V9')) %>% 
                        dplyr::rename(Seqtype = V3, start = V4, end = V5, info = V9) %>% 
                        dplyr::filter(Seqtype %in% c('start_codon','stop_codon')) %>%   
                        dplyr::mutate(GeneID = ((str_match(info, "gene_id\\s*(.*?);\\s*transcript"))[,2]) ) %>%
                        dplyr::mutate(TranscriptID = ((str_match(info, "transcript_id\\s*(.*?);\\s*db_xref"))[,2]) ) %>% 
                        dplyr::select(!info) 
# Cvirg_genome3.0 %>% group_by(TranscriptID) %>% mutate(Length = (start[Seqtype == 'start_codon'] - start[Seqtype == 'stop_codon']) )

# CvirgLength             <- data.frame(matrix(nrow = length(unique(Cvirg_genome3.0$TranscriptID)), ncol = 3)) 
# CvirgLength$TransciptID <- (Cvirg_genome3.0 %>% group_by(TranscriptID) %>% filter(row_number()==1))$TranscriptID
# CvirgLength$start       <- (Cvirg_genome3.0 %>% group_by(TranscriptID) %>% filter(row_number()==1))$start
# CvirgLength$end         <- (Cvirg_genome3.0 %>% group_by(TranscriptID) %>% filter(row_number()==2))$start
# 
# for (i in 1:nrow(CvirgLength)) {
#   if (CvirgLength$start[i] > CvirgLength$end[i]) {
#       CvirgLength$Length[i] <- CvirgLength$start[i] - CvirgLength$end[i]
#       } else (CvirgLength$Length[i] <-CvirgLength$end[i] - CvirgLength$start[i])
# 
# }
# View(CvirgLength)


Cvirginica_GOterms  <- unique(read.csv(file="../Data/TagSeq/Seq_details/Cviginiva_GOterms.csv", sep=',', header=TRUE) %>% 
                        dplyr::select(c('GeneID' ,'Annotation_GO_ID', 'Length')) %>% 
                        dplyr::group_by(GeneID) %>% 
                        dplyr::filter(Length == max(Length)))

Cgigas_GOterms      <- read.csv(file="../Data/TagSeq/Seq_details/Cgigas_refs/Cgigas_GOterms_uniprot.csv", sep=',', header=TRUE) %>%    
                        dplyr::select(c('Entry.name','Protein.names','Gene.ontology.IDs'))


  
# KEGG IDs: upload the Crassostrea gigas (Pacific oyster) KEGG reference 


Cgigas_KEGGIDs      <- as.data.frame(read.delim2(file = "C:/Users/samjg/Documents/Bioinformatics/refs/Cgigas/T03920_(2021_06_23 21_09_09 UTC).nuc", header=FALSE)) %>% 
                        dplyr::filter(grepl(">crg:", V1)) %>% 
                        dplyr::mutate(Cgigas_KEGGID = gsub("\\s.*|>", "",V1)) %>% 
                        dplyr::mutate(Cgigas_Protein_name = gsub("\\sLOC105.*|*.;", "", substring(V1, 16))) %>% 
                        dplyr::select(!V1)

blastx_CgigasCvirg  <- read.delim2(file="../Data/TagSeq/Seq_details/Cgigas_refs/crgKEGG_diamond_out.txt",header=FALSE) %>% 
                          dplyr::rename(qseqid = V1,sseqid = V2,pident = V3,length = V4,mismatch = V5,gapopen = V6,qstart = V7,qend = V8,sstart = V9,send = V10,evalue = V11,bitscore = V12)  %>% # rename columns 
                          dplyr::group_by(qseqid)

# GOslim
slim <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database - # call goslim_generic.obo terms as 'slim'

```

```{r Master reference }

#  assemble a datafile of the best blast hist for Cvirginica to the Cgigas protein database to obtain KEGG IDs
blastx_CgigasCvirg_besthits   <- blastx_CgigasCvirg %>% 
                                    dplyr::group_by(qseqid) %>% # group by the query C virginica transcript IDs
                                    dplyr::filter(evalue == min(evalue)) %>% # call the lowest E value (also highest bitscore) indicating the BEST blast hit
                                    dplyr::select(c('qseqid','sseqid', 'length'))  %>%  # call the two target columns - the query (C virginica transcript IDs) and the database of the blast (Cgigas KEGG IDs)
                                    dplyr::rename(Cvirginica_TranscriptID = qseqid, Cgigas_KEGGID = sseqid, Cgigas_length = length)# rename these columns for clarity

# merge the Cvirginica annotated reference with the Cvirnginca GO terms (contains the gene length!) 
Cvirginica_GOterms_MERGE <- merge(Cvirginica_annot_reference, Cvirginica_GOterms) %>% 
    dplyr::rename(Cvirginica_GeneID = GeneID, Cvirginica_length = Length, Cvirginica_TranscriptID = TranscriptID, Cvirginica_Protein_name = Protein_name)
nrow(Cvirginica_GOterms_MERGE) # 61891

```

```{r }
library(magrittr)
# create a referecne from the Geoduck annotation file to merge with the DE sumamry lists 
Master_ref    <- (as.data.frame(merge( (merge(Cgigas_KEGGIDs, blastx_CgigasCvirg_besthits)), Cvirginica_GOterms_MERGE)))[,c(1,5,6,8,2,3,4,7)]
nrow(Master_ref) # 66437

# write master reference
write.csv(Master_ref, file = "../Data/TagSeq/Seq_details/Seq_Reference_Master.csv", row.names = FALSE)



# ref$Function # View the new column Function


# Day 2 larva temperature pairwise main effect (high v. low)
DE_d2.Temperaturey_SHORT <- DE_d2.Temperature %>% 
                              dplyr::select("TranscriptID", 'log2FoldChange', 'padj') %>% 
                              dplyr::filter(!padj > 0.05) %>% 
                              dplyr::mutate(Dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange < 0 ~ "down")) %>%  
                              dplyr::mutate(Effect = "Day2_Temperature") %>% 
                              dplyr::mutate(time = "d2_larva")
# Day 2 larva Salinity pairwise main effect (high v. low)
DE_d2.Salinity_SHORT     <- DE_d2.Salinity %>% 
                              dplyr::select("TranscriptID", 'log2FoldChange', 'padj') %>% 
                              dplyr::filter(!padj > 0.05) %>% 
                              dplyr::mutate(Dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange < 0 ~ "down")) %>%  
                              dplyr::mutate(Effect = "Day2_Salinity") %>% 
                              dplyr::mutate(time = "d2_larva")
# Day 2 larva OA pairwise main effect (high v. low)
DE_d2.OA_SHORT           <- DE_d2.OA %>% 
                              dplyr::select("TranscriptID", 'log2FoldChange', 'padj') %>% 
                              dplyr::filter(!padj > 0.05) %>% 
                              dplyr::mutate(Dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange < 0 ~ "down")) %>%  
                              dplyr::mutate(Effect = "Day2_OA") %>% 
                              dplyr::mutate(time = "d2_larva")


# Day 18 larva Salinity pairwise main effect (high v. low)
DE_d18.Salinity_SHORT     <- DE_d18.Salinity %>% 
                              dplyr::select("TranscriptID", 'log2FoldChange', 'padj') %>% 
                              dplyr::filter(!padj > 0.05) %>% 
                              dplyr::mutate(Dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange < 0 ~ "down")) %>%  
                              dplyr::mutate(Effect = "Day18_Salinity") %>% 
                              dplyr::mutate(time = "d18_spat")
# Day 18 larva OA pairwise main effect (high v. low)
DE_d18.OA_SHORT           <- DE_d18.OA %>% 
                              dplyr::select("TranscriptID", 'log2FoldChange', 'padj') %>% 
                              dplyr::filter(!padj > 0.05) %>% 
                              dplyr::mutate(Dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange < 0 ~ "down")) %>%  
                              dplyr::mutate(Effect = "Day18_OA") %>% 
                              dplyr::mutate(time = "d18_spat")

# merge to a master file
DE_ALL           <- rbind(DE_d2.Temperaturey_SHORT, DE_d2.Salinity_SHORT, DE_d2.OA_SHORT, 
                   DE_d18.Salinity_SHORT, DE_d18.OA_SHORT)

DE_ALL_merged    <- merge( (DE_ALL %>% dplyr::rename(Cvirginica_TranscriptID=TranscriptID)), Master_ref, by="Cvirginica_TranscriptID")
DE_Master_with_annotation <- DE_ALL_merged[!duplicated(DE_ALL_merged[,c('Cvirginica_TranscriptID','log2FoldChange', 'Effect', 'padj', 'Cvirginica_length')]),]

write.csv(DE_Master_with_annotation, file = "../Output/DESeq2/DE_All_Annotation.csv", row.names = FALSE)

```

Prep Differential Expression data (from DESeq2) by Date and direction of DE 
```{r}

DE_Master_with_annotation <- read.csv(file="C:/Users/samjg/Documents/Github_repositories/Cvirginica_multistressor/RAnalysis/Output/DESeq2/DE_All_Annotation.csv", sep=',', header=TRUE)

# Objective here is to call the DE genes (up and down) 
# DAY 2 Temperature------------------------------------------- #
DE_d2.Temperature.UPREG  <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day2_Temperature") %>% dplyr::filter(Dir %in% 'up')               # UPREG call
DE_d2.Temperature.DWNREG <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day2_Temperature") %>% dplyr::filter(Dir %in% 'down')             # DWNREG call
# DAY 2 Salinity   ------------------------------------------- #
DE_d2.Salinity.UPREG     <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day2_Salinity") %>% dplyr::filter(Dir %in% 'up')               # UPREG call
DE_d2.Salinity.DWNREG    <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day2_Salinity") %>% dplyr::filter(Dir %in% 'down')             # DWNREG call
# DAY 2 OA         ------------------------------------------- #
DE_d2.OA.UPREG           <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day2_OA") %>% dplyr::filter(Dir %in% 'up')               # UPREG call
DE_d2.OA.DWNREG          <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day2_OA") %>% dplyr::filter(Dir %in% 'down')             # DWNREG call


# Day 18 Salinity   ------------------------------------------- #
DE_d18.Salinity.UPREG    <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day18_Salinity") %>% dplyr::filter(Dir %in% 'up')               # UPREG call
DE_d18.Salinity.DWNREG   <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day18_Salinity") %>% dplyr::filter(Dir %in% 'down')             # DWNREG call
# Day 18 OA         ------------------------------------------- #
DE_d18.OA.UPREG          <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day18_OA") %>% dplyr::filter(Dir %in% 'up')               # UPREG call
DE_d18.OA.DWNREG         <- DE_Master_with_annotation %>% dplyr::filter(Effect %in% "Day18_OA") %>% dplyr::filter(Dir %in% 'down')             # DWNREG call



# All genes - and all by common abiotic challenge  (----------------------------)not including uncharacterized annotations)  -------------------------------------------------------- #
All.Salinity.UPREG        <- rbind(DE_d2.Salinity.UPREG,DE_d18.Salinity.UPREG) %>% add_count(Cvirginica_TranscriptID)# UPREG call
Constant_UPREG_Salinity   <- unique(All.Salinity.UPREG %>% dplyr::filter(n %in% 2) %>%  dplyr::select(c('Cvirginica_Protein_name','Cvirginica_TranscriptID')) %>% dplyr::filter(!Cvirginica_Protein_name %in% " uncharacterized")  )

All.Salinity.DWNREG       <- rbind(DE_d2.Salinity.DWNREG,DE_d18.Salinity.DWNREG) %>% add_count(Cvirginica_TranscriptID)# UPREG call
Constant_DWNREG_Salinity  <- unique(All.Salinity.DWNREG %>% dplyr::filter(n %in% 2) %>% dplyr::select(c('Cvirginica_Protein_name','Cvirginica_TranscriptID')) %>% dplyr::filter(!Cvirginica_Protein_name %in% " uncharacterized")  )


All.OA.UPREG              <- rbind(DE_d2.OA.UPREG,DE_d18.OA.UPREG) %>% add_count(Cvirginica_TranscriptID)# UPREG call
Constant_UPREG_OA         <- unique(All.OA.UPREG %>% dplyr::filter(n %in% 2) %>%  dplyr::select(c('Cvirginica_Protein_name','Cvirginica_TranscriptID')) %>% dplyr::filter(!Cvirginica_Protein_name %in% " uncharacterized")  )

All.OA.DWNREG             <- rbind(DE_d2.OA.DWNREG,DE_d18.OA.DWNREG) %>% add_count(Cvirginica_TranscriptID)# UPREG call
Constant_DWNREG_OA        <- unique(All.OA.DWNREG %>% dplyr::filter(n %in% 2) %>%  dplyr::select(c('Cvirginica_Protein_name','Cvirginica_TranscriptID')) %>% dplyr::filter(!Cvirginica_Protein_name %in% " uncharacterized")  )

```

goseq
Prepare dataframe(s) and vectors for goseq 
(1) Format 'GO.term' for goseq from the P.generosa annotation .fna file 'Geoduck_annotation'
```{r GO.terms_data}
# call the GO terms
Cvirginica_GOterms                <- as.data.frame(Master_ref) %>% dplyr::select(c('Cvirginica_TranscriptID','Annotation_GO_ID'))
colnames(Cvirginica_GOterms)[1:2] <- c('transcript.ID', 'GO.terms') # call gene name and the GO terms - (Uniprot ID 'V5')
splitted                          <- strsplit(as.character(Cvirginica_GOterms$GO.terms), ";") #slit into multiple GO ids by delimiter'; ' remember the space after ; is needed here! without this you will only call the first listed GO term for each gene!
GO.terms                          <- data.frame(v1 = rep.int(Cvirginica_GOterms$transcript.ID, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
```


(2) Unique Genes - vector based on all unique mapped reads  - Remember this is based on the genes for EACH sampling day becasue DEeq2 was run separately (i.e. days 0, 7, 14, and 21)
(3) Gene length 
Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
```{r Unique_genes}
# Prepare dataframe(s) and vectors for goseq 
# Format 'GO.term' for goseq from the P.generosa annotation .fna file 'Geoduck_annotation'
IDvector.d2         <- as.vector(unique(d2.counts$TranscriptID))   # call unique genes (those filtered and used in DESEq2) on day0 - 'IDvector'
IDvector.d18        <- as.vector(unique(d18.counts$TranscriptID))  # call unique genes (those filtered and used in DESEq2) on day7 - 'IDvector'
GO_unique.genes.all <- as.vector(unique(Master_ref$Cvirginica_TranscriptID)) # call all unique genes for GO analysis (goseq)

# Gene length 
GO_gene.length  <- unique(Master_ref %>% dplyr::select(c("Cvirginica_TranscriptID","Cvirginica_length")))

# merge length with counts data
length_vector   <- GO_gene.length$Cvirginica_length

GeneLength.d2   <-  dplyr::inner_join(d2.counts, unique((GO_gene.length %>% dplyr::rename(TranscriptID = Cvirginica_TranscriptID)), by='TranscriptID'))

GeneLength.d18  <- dplyr::inner_join(d18.counts, unique((GO_gene.length %>% dplyr::rename(TranscriptID = Cvirginica_TranscriptID)), by='TranscriptID'))

# call length values for goseq - confirms that the IDvector and length_vector are the same!!!
length_vector.d2 <- GeneLength.d2$Cvirginica_length    # length vector for all unique reads address in DESeq2 on day 0 
sum(sapply(length_vector.d2,length)) == dim(d2.counts)[1] #should be TRUE

length_vector.d18 <- GeneLength.d18$Cvirginica_length    # length vector for all unique reads address in DESeq2 on day 7
sum(sapply(length_vector.d18,length)) == dim(d18.counts)[1] #should be TRUE

```

(4) Call DE datasets (from DESeq2) - merge to the genes list 'GO_unique.genes.all' to create a binary vector 0 = not DE'd; 1 = DE'd
```{r Binary_DEGs - must run the prep steps above}
# (1) DEGs: Pairwise main effects


# DAY 2 larvae ------------------------------------------- #


# TEMPERATURE 
nrow(DE_d2.Temperature.UPREG) # 341
DE.d2._Temperature.UP    <-as.integer(GO_unique.genes.all%in%(DE_d2.Temperature.UPREG$Cvirginica_TranscriptID)) # Day 2-temperature, up
names(DE.d2._Temperature.UP)=GO_unique.genes.all
sum(sapply(DE.d2._Temperature.UP,length)) # 59032
sum(DE.d2._Temperature.UP == 1) # should be 341

nrow(DE_d2.Temperature.DWNREG) # 251
DE.d2._Temperature.DOWN    <-as.integer(GO_unique.genes.all%in%(DE_d2.Temperature.DWNREG$Cvirginica_TranscriptID)) # Day 2-temperature, up
names(DE.d2._Temperature.DOWN)=GO_unique.genes.all
sum(sapply(DE.d2._Temperature.DOWN,length)) # 59032
sum(DE.d2._Temperature.DOWN == 1) # should be 251


# SALINITY
nrow(DE_d2.Salinity.UPREG) # 318
DE.d2._Salinity.UP    <-as.integer(GO_unique.genes.all%in%(DE_d2.Salinity.UPREG$Cvirginica_TranscriptID)) # Day 2-Salinity, up
names(DE.d2._Salinity.UP)=GO_unique.genes.all
sum(sapply(DE.d2._Salinity.UP,length)) # 59032
sum(DE.d2._Salinity.UP == 1) # should be 318

nrow(DE_d2.Salinity.DWNREG) # 355
DE.d2._Salinity.DOWN    <-as.integer(GO_unique.genes.all%in%(DE_d2.Salinity.DWNREG$Cvirginica_TranscriptID)) # Day 2-Salinity, up
names(DE.d2._Salinity.DOWN)=GO_unique.genes.all
sum(sapply(DE.d2._Salinity.DOWN,length)) # 59032
sum(DE.d2._Salinity.DOWN == 1) # should be 355

# OA
nrow(DE_d2.OA.UPREG) # 81
DE.d2._OA.UP    <-as.integer(GO_unique.genes.all%in%(DE_d2.OA.UPREG$Cvirginica_TranscriptID)) # Day 2-OA, up
names(DE.d2._OA.UP)=GO_unique.genes.all
sum(sapply(DE.d2._OA.UP,length)) # 59032
sum(DE.d2._OA.UP == 1) # should be 81

nrow(DE_d2.OA.DWNREG) # 70
DE.d2._OA.DOWN    <-as.integer(GO_unique.genes.all%in%(DE_d2.OA.DWNREG$Cvirginica_TranscriptID)) # Day 2-OA, up
names(DE.d2._OA.DOWN)=GO_unique.genes.all
sum(sapply(DE.d2._OA.DOWN,length)) # 59032
sum(DE.d2._OA.DOWN == 1) # should be 70

# DAY 18 spat ------------------------------------------- #

# SALINITY
nrow(DE_d18.Salinity.UPREG) # 476
DE.d18._Salinity.UP    <-as.integer(GO_unique.genes.all%in%(DE_d18.Salinity.UPREG$Cvirginica_TranscriptID)) # Day 2-Salinity, up
names(DE.d18._Salinity.UP)=GO_unique.genes.all
sum(sapply(DE.d18._Salinity.UP,length)) # 59032
sum(DE.d18._Salinity.UP == 1) # should be 476

nrow(DE_d18.Salinity.DWNREG) # 494
DE.d18._Salinity.DOWN    <-as.integer(GO_unique.genes.all%in%(DE_d18.Salinity.DWNREG$Cvirginica_TranscriptID)) # Day 2-Salinity, up
names(DE.d18._Salinity.DOWN)=GO_unique.genes.all
sum(sapply(DE.d18._Salinity.DOWN,length)) # 59032
sum(DE.d18._Salinity.DOWN == 1) # should be 494

# OA
nrow(DE_d18.OA.UPREG) # 185
DE.d18._OA.UP    <-as.integer(GO_unique.genes.all%in%(DE_d18.OA.UPREG$Cvirginica_TranscriptID)) # Day 2-OA, up
names(DE.d18._OA.UP)=GO_unique.genes.all
sum(sapply(DE.d18._OA.UP,length)) # 59032
sum(DE.d18._OA.UP == 1) # should be 185

nrow(DE_d18.OA.DWNREG) # 160
DE.d18._OA.DOWN    <-as.integer(GO_unique.genes.all%in%(DE_d18.OA.DWNREG$Cvirginica_TranscriptID)) # Day 2-OA, up
names(DE.d18._OA.DOWN)=GO_unique.genes.all
sum(sapply(DE.d18._OA.DOWN,length)) # 59032
sum(DE.d18._OA.DOWN == 1) # should be 160


```

Let's review what we have for goseq
```{r NOTE_checklist}
# Review what we have for goseq....

# (1) Go terms ---------------- #
#  head(GO.terms)

# (2) ID vector --------------- #
# head(IDvector.d0)
# head(IDvector.d7)
# head(IDvector.d14)
# head(IDvector.d21)

# (3) Length vector ----------- #
# head(length_vector.d0)
# head(length_vector.d7)
# head(length_vector.d14)
# head(length_vector.d21)

# (4) Binary DEG vectors ------ # 
# head(DE.d21._Primary.UP)    # example of d21 prmary ambient_v_moderate upregulated DEGs
# head(DE.d21._Primary.DOWN)  # example of d21 prmary ambient_v_moderate downregulated DEGs
# head(DE.d21._Primary.ALL)   # example of d21 prmary ambient_v_moderate all DEGs
```

It's goseq time!!!
Calculate Probability Weighting Function
```{r pwf}
#Calculate Probability Weighting Function (using 'nullp')


# DAY 2 larvae ------------------------------------------- #

# Temperature
TEMP.pwf_d2.UP    <- nullp(DE.d2._Temperature.UP,    id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
TEMP.pwf_d2.DOWN  <- nullp(DE.d2._Temperature.DOWN,  id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
# Salinity
SAL.pwf_d2.UP     <- nullp(DE.d2._Salinity.UP,    id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
SAL.pwf_d2.DOWN   <- nullp(DE.d2._Salinity.DOWN,  id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
# OA
OA.pwf_d2.UP      <- nullp(DE.d2._OA.UP,    id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
OA.pwf_d2.DOWN    <- nullp(DE.d2._OA.DOWN,  id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene


# DAY 18 spat ------------------------------------------- #

# Salinity
SAL.pwf_d18.UP     <- nullp(DE.d18._Salinity.UP,    id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
SAL.pwf_d18.DOWN   <- nullp(DE.d18._Salinity.DOWN,  id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
# OA
OA.pwf_d18.UP      <- nullp(DE.d18._OA.UP,    id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
OA.pwf_d18.DOWN    <- nullp(DE.d18._OA.DOWN,  id=GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene

```

Run goseq
```{r goseq}


# DAY 2 larvae  ------------------------------------------- #

# Temperature
Temp.GO_d2.UP       <- goseq(TEMP.pwf_d2.UP,  GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

Temp.GO_d2.DOWN    <- goseq(TEMP.pwf_d2.DOWN,   GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

# Salinity
Salinity.GO_d2.UP    <- goseq(SAL.pwf_d2.UP,  GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

Salinity.GO_d2.DOWN <- goseq(SAL.pwf_d2.DOWN,   GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

# OA
OA.GO_d2.UP        <- goseq(OA.pwf_d2.UP,  GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

OA.GO_d2.DOWN     <- goseq(OA.pwf_d2.DOWN,   GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)


# DAY 18 spat ------------------------------------------- #


# Salinity
Salinity.GO_d18.UP        <- goseq(SAL.pwf_d18.UP,  GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

Salinity.GO_d18.DOWN     <- goseq(SAL.pwf_d18.DOWN,   GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

# OA
OA.GO_d18.UP        <- goseq(OA.pwf_d18.UP,  GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)

OA.GO_d18.DOWN     <- goseq(OA.pwf_d18.DOWN,   GO_unique.genes.all, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
```

Find only enriched GO terms that are statistically significant at cutoff
```{r - Sig GO terms Day 2 larvae}


##### NOTE #####
# each of the GO analysis below contains two data sets: 
# (1) the UNFILTERED data (i.e. d0.UP.05 = day 0, upregualted genes, significant GO terms < 0.05)
# (2) the FILTERED data (i.e. d0.UP.05_filtered = day 0, upregualted genes, significant GO terms < 0.05 AND filtered for BP terms >= 2 genes per term + MG terms >= 2 per term)
# ....review  the compileGO.R script for the cumulative spreadsheet of ALL UNFILTERED GO terms for the funtional analysis of each pairwise DE below!


# Day 2 larvae - Pairwise main effects  ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::;

# TEMPETAURE ::::::::::::::::::::::::::::::::::::::

# UPREGULATED GENES - those with higher expression from ambient (control) stress priming
Temp.d2.UP.05           <- Temp.GO_d2.UP$category[Temp.GO_d2.UP$over_represented_pvalue<.05]
Temp.d2.UP.05           <- data.frame(Temp.d2.UP.05)
colnames(Temp.d2.UP.05) <- c("category")
Temp.d2.UP.05           <- merge(Temp.d2.UP.05, Temp.GO_d2.UP, by="category")
Temp.d2.UP.05           <- Temp.d2.UP.05[order(Temp.d2.UP.05$ontology, Temp.d2.UP.05$over_represented_pvalue, -Temp.d2.UP.05$numDEInCat),]
Temp.d2.UP.05$term      <- as.factor(Temp.d2.UP.05$term)
Temp.d2.UP.05_filtered  <- Temp.d2.UP.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(Temp.d2.UP.05_filtered, ontology=="BP")) #number sig BP terms == 9
nrow(filter(Temp.d2.UP.05_filtered, ontology=="MF")) #number sig MF terms == 5
nrow(Temp.d2.UP.05_filtered) # == 20

write.csv(Temp.d2.UP.05_filtered, file = "../Output/DESeq2/Day2_larva/GO_analysis/GO_Temperature_UpregulatedDEGs.csv", row.names = FALSE)


#DOWNREGULATED GENES  - those with higher expression from moderate stress priming
Temp.d2.DOWN.05           <- Temp.GO_d2.DOWN$category[Temp.GO_d2.DOWN$over_represented_pvalue<.05]
Temp.d2.DOWN.05           <- data.frame(Temp.d2.DOWN.05)
colnames(Temp.d2.DOWN.05) <- c("category")
Temp.d2.DOWN.05           <- merge(Temp.d2.DOWN.05, Temp.GO_d2.DOWN, by="category")
Temp.d2.DOWN.05           <- Temp.d2.DOWN.05[order(Temp.d2.DOWN.05$ontology, Temp.d2.DOWN.05$over_represented_pvalue, -Temp.d2.DOWN.05$numDEInCat),]
Temp.d2.DOWN.05$term      <- as.factor(Temp.d2.DOWN.05$term)
Temp.d2.DOWN.05_filtered  <- Temp.d2.DOWN.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(Temp.d2.DOWN.05_filtered, ontology=="BP")) #number sig BP terms == 4
nrow(filter(Temp.d2.DOWN.05_filtered, ontology=="MF")) #number sig MF terms == 5
nrow(Temp.d2.DOWN.05_filtered) # == 16

write.csv(Temp.d2.DOWN.05_filtered, file = "../Output/DESeq2/Day2_larva/GO_analysis/GO_Temperature_DownregulatedDEGs.csv", row.names = FALSE)


# SALINITY ::::::::::::::::::::::::::::::::::::::

# UPREGULATED GENES - those with higher expression from ambient (control) stress priming
Salinity.d2.UP.05           <- Salinity.GO_d2.UP$category[Salinity.GO_d2.UP$over_represented_pvalue<.05]
Salinity.d2.UP.05           <- data.frame(Salinity.d2.UP.05)
colnames(Salinity.d2.UP.05) <- c("category")
Salinity.d2.UP.05           <- merge(Salinity.d2.UP.05, Salinity.GO_d2.UP, by="category")
Salinity.d2.UP.05           <- Salinity.d2.UP.05[order(Salinity.d2.UP.05$ontology, Salinity.d2.UP.05$over_represented_pvalue, -Salinity.d2.UP.05$numDEInCat),]
Salinity.d2.UP.05$term      <- as.factor(Salinity.d2.UP.05$term)
Salinity.d2.UP.05_filtered  <- Salinity.d2.UP.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(Salinity.d2.UP.05_filtered, ontology=="BP")) #number sig BP terms == 5
nrow(filter(Salinity.d2.UP.05_filtered, ontology=="MF")) #number sig MF terms == 5
nrow(Salinity.d2.UP.05_filtered) # == 15

write.csv(Salinity.d2.UP.05_filtered, file = "../Output/DESeq2/Day2_larva/GO_analysis/GO_Salinity_UpregulatedDEGs.csv", row.names = FALSE)

#DOWNREGULATED GENES  - those with higher expression from moderate stress priming
Salinity.d2.DOWN.05           <- Salinity.GO_d2.DOWN$category[Salinity.GO_d2.DOWN$over_represented_pvalue<.05]
Salinity.d2.DOWN.05           <- data.frame(Salinity.d2.DOWN.05)
colnames(Salinity.d2.DOWN.05) <- c("category")
Salinity.d2.DOWN.05           <- merge(Salinity.d2.DOWN.05, Salinity.GO_d2.DOWN, by="category")
Salinity.d2.DOWN.05           <- Salinity.d2.DOWN.05[order(Salinity.d2.DOWN.05$ontology, Salinity.d2.DOWN.05$over_represented_pvalue, -Salinity.d2.DOWN.05$numDEInCat),]
Salinity.d2.DOWN.05$term      <- as.factor(Salinity.d2.DOWN.05$term)
Salinity.d2.DOWN.05_filtered  <- Salinity.d2.DOWN.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(Salinity.d2.DOWN.05_filtered, ontology=="BP")) #number sig BP terms == 11
nrow(filter(Salinity.d2.DOWN.05_filtered, ontology=="MF")) #number sig MF terms == 7
nrow(Salinity.d2.DOWN.05_filtered) # == 26

write.csv(Salinity.d2.DOWN.05_filtered, file = "../Output/DESeq2/Day2_larva/GO_analysis/GO_Salinity_DownregulatedDEGs.csv", row.names = FALSE)

# OA:::::::::::: ::::::::::::::::::::::::::::::::::::::

# UPREGULATED GENES - those with higher expression from ambient (control) stress priming
OA.d2.UP.05           <- OA.GO_d2.UP$category[OA.GO_d2.UP$over_represented_pvalue<.05]
OA.d2.UP.05           <- data.frame(OA.d2.UP.05)
colnames(OA.d2.UP.05) <- c("category")
OA.d2.UP.05           <- merge(OA.d2.UP.05, OA.GO_d2.UP, by="category")
OA.d2.UP.05           <- OA.d2.UP.05[order(OA.d2.UP.05$ontology, OA.d2.UP.05$over_represented_pvalue, -OA.d2.UP.05$numDEInCat),]
OA.d2.UP.05$term      <- as.factor(OA.d2.UP.05$term)
OA.d2.UP.05_filtered  <- OA.d2.UP.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(OA.d2.UP.05_filtered, ontology=="BP")) #number sig BP terms == 2
nrow(filter(OA.d2.UP.05_filtered, ontology=="MF")) #number sig MF terms == 0
nrow(OA.d2.UP.05_filtered) # == 3

write.csv(OA.d2.UP.05_filtered, file = "../Output/DESeq2/Day2_larva/GO_analysis/GO_OA_UpregulatedDEGs.csv", row.names = FALSE)

#DOWNREGULATED GENES  - those with higher expression from moderate stress priming
OA.d2.DOWN.05           <- OA.GO_d2.DOWN$category[OA.GO_d2.DOWN$over_represented_pvalue<.05]
OA.d2.DOWN.05           <- data.frame(OA.d2.DOWN.05)
colnames(OA.d2.DOWN.05) <- c("category")
OA.d2.DOWN.05           <- merge(OA.d2.DOWN.05, OA.GO_d2.DOWN, by="category")
OA.d2.DOWN.05           <- OA.d2.DOWN.05[order(OA.d2.DOWN.05$ontology, OA.d2.DOWN.05$over_represented_pvalue, -OA.d2.DOWN.05$numDEInCat),]
OA.d2.DOWN.05$term      <- as.factor(OA.d2.DOWN.05$term)
OA.d2.DOWN.05_filtered  <- OA.d2.DOWN.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(OA.d2.DOWN.05_filtered, ontology=="BP")) #number sig BP terms == 3
nrow(filter(OA.d2.DOWN.05_filtered, ontology=="MF")) #number sig MF terms == 1
nrow(OA.d2.DOWN.05_filtered) # == 7

write.csv(OA.d2.DOWN.05_filtered, file = "../Output/DESeq2/Day2_larva/GO_analysis/GO_OA_DownregulatedDEGs.csv", row.names = FALSE)

```


Find only enriched GO terms that are statistically significant at cutoff
```{r - Sig GO terms Day 18 spat}


##### NOTE #####
# each of the GO analysis below contains two data sets: 
# (1) the UNFILTERED data (i.e. d0.UP.05 = day 0, upregualted genes, significant GO terms < 0.05)
# (2) the FILTERED data (i.e. d0.UP.05_filtered = day 0, upregualted genes, significant GO terms < 0.05 AND filtered for BP terms >= 2 genes per term + MG terms >= 2 per term)
# ....review  the compileGO.R script for the cumulative spreadsheet of ALL UNFILTERED GO terms for the funtional analysis of each pairwise DE below!


# Day 18 spat - Pairwise main effects  ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::;


# SALINITY ::::::::::::::::::::::::::::::::::::::

# UPREGULATED GENES - those with higher expression from ambient (control) stress priming
Salinity.d18.UP.05           <- Salinity.GO_d18.UP$category[Salinity.GO_d18.UP$over_represented_pvalue<.05]
Salinity.d18.UP.05           <- data.frame(Salinity.d18.UP.05)
colnames(Salinity.d18.UP.05) <- c("category")
Salinity.d18.UP.05           <- merge(Salinity.d18.UP.05, Salinity.GO_d18.UP, by="category")
Salinity.d18.UP.05           <- Salinity.d18.UP.05[order(Salinity.d18.UP.05$ontology, Salinity.d18.UP.05$over_represented_pvalue, -Salinity.d18.UP.05$numDEInCat),]
Salinity.d18.UP.05$term      <- as.factor(Salinity.d18.UP.05$term)
Salinity.d18.UP.05_filtered  <- Salinity.d18.UP.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(Salinity.d18.UP.05_filtered, ontology=="BP")) #number sig BP terms == 10
nrow(filter(Salinity.d18.UP.05_filtered, ontology=="MF")) #number sig MF terms == 6
nrow(Salinity.d18.UP.05_filtered) # == 20

write.csv(Salinity.d18.UP.05_filtered, file = "../Output/DESeq2/Day18_spat/GO_analysis/GO_Salinity_UpregulatedDEGs.csv", row.names = FALSE)


#DOWNREGULATED GENES  - those with higher expression from moderate stress priming
Salinity.d18.DOWN.05           <- Salinity.GO_d18.DOWN$category[Salinity.GO_d18.DOWN$over_represented_pvalue<.05]
Salinity.d18.DOWN.05           <- data.frame(Salinity.d18.DOWN.05)
colnames(Salinity.d18.DOWN.05) <- c("category")
Salinity.d18.DOWN.05           <- merge(Salinity.d18.DOWN.05, Salinity.GO_d18.DOWN, by="category")
Salinity.d18.DOWN.05           <- Salinity.d18.DOWN.05[order(Salinity.d18.DOWN.05$ontology, Salinity.d18.DOWN.05$over_represented_pvalue, -Salinity.d18.DOWN.05$numDEInCat),]
Salinity.d18.DOWN.05$term      <- as.factor(Salinity.d18.DOWN.05$term)
Salinity.d18.DOWN.05_filtered  <- Salinity.d18.DOWN.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(Salinity.d18.DOWN.05_filtered, ontology=="BP")) #number sig BP terms == 20
nrow(filter(Salinity.d18.DOWN.05_filtered, ontology=="MF")) #number sig MF terms == 14
nrow(Salinity.d18.DOWN.05_filtered) # == 49

write.csv(Salinity.d18.DOWN.05_filtered, file = "../Output/DESeq2/Day18_spat/GO_analysis/GO_Salinity_DownregulatedDEGs.csv", row.names = FALSE)


# OA:::::::::::: ::::::::::::::::::::::::::::::::::::::

# UPREGULATED GENES - those with higher expression from ambient (control) stress priming
OA.d18.UP.05           <- OA.GO_d18.UP$category[OA.GO_d18.UP$over_represented_pvalue<.05]
OA.d18.UP.05           <- data.frame(OA.d18.UP.05)
colnames(OA.d18.UP.05) <- c("category")
OA.d18.UP.05           <- merge(OA.d18.UP.05, OA.GO_d18.UP, by="category")
OA.d18.UP.05           <- OA.d18.UP.05[order(OA.d18.UP.05$ontology, OA.d18.UP.05$over_represented_pvalue, -OA.d18.UP.05$numDEInCat),]
OA.d18.UP.05$term      <- as.factor(OA.d18.UP.05$term)
OA.d18.UP.05_filtered  <- OA.d18.UP.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(OA.d18.UP.05_filtered, ontology=="BP")) #number sig BP terms == 16
nrow(filter(OA.d18.UP.05_filtered, ontology=="MF")) #number sig MF terms == 9
nrow(OA.d18.UP.05_filtered) # == 35

write.csv(OA.d18.UP.05_filtered, file = "../Output/DESeq2/Day18_spat/GO_analysis/GO_OA_UpregulatedDEGs.csv", row.names = FALSE)

#DOWNREGULATED GENES  - those with higher expression from moderate stress priming
OA.d18.DOWN.05           <- OA.GO_d18.DOWN$category[OA.GO_d18.DOWN$over_represented_pvalue<.05]
OA.d18.DOWN.05           <- data.frame(OA.d18.DOWN.05)
colnames(OA.d18.DOWN.05) <- c("category")
OA.d18.DOWN.05           <- merge(OA.d18.DOWN.05, OA.GO_d18.DOWN, by="category")
OA.d18.DOWN.05           <- OA.d18.DOWN.05[order(OA.d18.DOWN.05$ontology, OA.d18.DOWN.05$over_represented_pvalue, -OA.d18.DOWN.05$numDEInCat),]
OA.d18.DOWN.05$term      <- as.factor(OA.d18.DOWN.05$term)
OA.d18.DOWN.05_filtered  <- OA.d18.DOWN.05 %>% filter(!(numDEInCat<2 & ontology == "BP"), !(numDEInCat<2 & ontology == "MF"))
nrow(filter(OA.d18.DOWN.05_filtered, ontology=="BP")) #number sig BP terms == 3
nrow(filter(OA.d18.DOWN.05_filtered, ontology=="MF")) #number sig MF terms == 5
nrow(OA.d18.DOWN.05_filtered) # == 11

write.csv(OA.d18.DOWN.05_filtered, file = "../Output/DESeq2/Day18_spat/GO_analysis/GO_OA_DownregulatedDEGs.csv", row.names = FALSE)

```
Correct any un-annotated terms/ontologies
```{r}
NAs.ontology <- d0.UP.05 %>% subset(is.na(term))
print(NAs.ontology)
NAs.ontology <- d0.DOWN.05 %>% subset(is.na(term))
print(NAs.ontology)

NAs.ontology <- d7.UP.05 %>% subset(is.na(term))
print(NAs.ontology)
NAs.ontology <- d7.DOWN.05 %>% subset(is.na(term))
print(NAs.ontology)

NAs.ontology <- d14.UP.05 %>% subset(is.na(term))
print(NAs.ontology)
NAs.ontology <- d14.DOWN.05 %>% subset(is.na(term))
print(NAs.ontology)

NAs.ontology <- d21.UP.05 %>% subset(is.na(term))
print(NAs.ontology)
NAs.ontology <- d21.DOWN.05 %>% subset(is.na(term))
print(NAs.ontology)

# NO CASES OF NA
```


GOslim analysis - data reduction from the many GO terms to more broad function/processes
Read in files if previous steps are already run
```{r}


# load Day 2 larva GO results (main effect models - within treatment)
Salinity_d2.UP.05       <- read.csv("../Output/DESeq2/Day2_larva/GO_analysis/GO_Salinity_UpregulatedDEGs.csv") %>%  
                              dplyr::mutate(dir =  "up") %>% 
                              dplyr::mutate(Day = '2') %>% 
                              dplyr::mutate(challenge = 'salinity')

Salinity_d2.DOWN.05     <- read.csv("../Output/DESeq2/Day2_larva/GO_analysis/GO_Salinity_DownregulatedDEGs.csv") %>%  
                              dplyr::mutate(dir =  "down") %>% 
                              dplyr::mutate(Day = '2') %>% 
                              dplyr::mutate(challenge = 'salinity')
                              
Temperature_d2.UP.05    <- read.csv("../Output/DESeq2/Day2_larva/GO_analysis/GO_Temperature_UpregulatedDEGs.csv") %>%  
                              dplyr::mutate(dir =  "up") %>% 
                              dplyr::mutate(Day = '2') %>% 
                              dplyr::mutate(challenge = 'temperature')
                              
Temperature_d2.DOWN.05  <- read.csv("../Output/DESeq2/Day2_larva/GO_analysis/GO_Temperature_DownregulatedDEGs.csv") %>%  
                              dplyr::mutate(dir =  "down") %>% 
                              dplyr::mutate(Day = '2') %>% 
                              dplyr::mutate(challenge = 'temperature')
                              
OA_d2.UP.05             <- read.csv("../Output/DESeq2/Day2_larva/GO_analysis/GO_OA_UpregulatedDEGs.csv") %>%  
                              dplyr::mutate(dir =  "up") %>% 
                              dplyr::mutate(Day = '2') %>% 
                              dplyr::mutate(challenge = 'OA')
                              
OA_d2.DOWN.05           <- read.csv("../Output/DESeq2/Day2_larva/GO_analysis/GO_OA_DownregulatedDEGs.csv") %>%  
                              dplyr::mutate(dir =  "down") %>% 
                              dplyr::mutate(Day = '2') %>% 
                              dplyr::mutate(challenge = 'OA')

# load Day 18 spat GO results (main effect models - within treatment)

OA_d18.UP.05         <- read.csv("../Output/DESeq2/Day18_spat/GO_analysis/GO_OA_UpregulatedDEGs.csv") %>%  
                              dplyr::mutate(dir =  "up") %>% 
                              dplyr::mutate(Day = '18') %>% 
                              dplyr::mutate(challenge = 'OA')

OA_d18.DOWN.05       <- read.csv("../Output/DESeq2/Day18_spat/GO_analysis/GO_OA_DownregulatedDEGs.csv")%>%  
                              dplyr::mutate(dir =  "down") %>% 
                              dplyr::mutate(Day = '18') %>% 
                              dplyr::mutate(challenge = 'OA')

Salinity_d18.UP.05   <- read.csv("../Output/DESeq2/Day18_spat/GO_analysis/GO_Salinity_UpregulatedDEGs.csv")  %>%  
                              dplyr::mutate(dir =  "up") %>% 
                              dplyr::mutate(Day = '18') %>% 
                              dplyr::mutate(challenge = 'salinity')

Salinity_d18.DOWN.05 <- read.csv("../Output/DESeq2/Day18_spat/GO_analysis/GO_Salinity_DownregulatedDEGs.csv") %>%  
                              dplyr::mutate(dir =  "down") %>% 
                              dplyr::mutate(Day = '18') %>% 
                              dplyr::mutate(challenge = 'salinity')

Master_goseq_results      <- rbind(Salinity_d2.UP.05, Temperature_d2.UP.05, OA_d2.UP.05,Salinity_d18.UP.05,OA_d18.UP.05,
                                   Salinity_d2.DOWN.05, Temperature_d2.DOWN.05, OA_d2.DOWN.05,Salinity_d18.DOWN.05,OA_d18.DOWN.05) %>%                                   dplyr::filter(!ontology %in% 'CC') # Master GOseq OF ALL PRIMARY EFFECTS







DEG_d2.Temperature <- DE_d2.Temperature %>% dplyr::filter(padj < 0.05) %>%  dplyr::mutate(dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange <0 ~ "down")) %>% dplyr::mutate(Day = '2') %>% dplyr::mutate(challenge = 'temperature')  

DEGs_d2.Salinity   <- DE_d2.Salinity    %>% dplyr::filter(padj < 0.05) %>%  dplyr::mutate(dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange <0 ~ "down")) %>% dplyr::mutate(Day = '2') %>% dplyr::mutate(challenge = 'salinity')  

DEGs_d2.OA         <- DE_d2.OA          %>% dplyr::filter(padj < 0.05) %>%  dplyr::mutate(dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange <0 ~ "down"))  %>% dplyr::mutate(Day = '2') %>% dplyr::mutate(challenge = 'OA')  


DEGs_d18.Salinity   <- DE_d18.Salinity    %>% dplyr::filter(padj < 0.05) %>%  dplyr::mutate(dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange <0 ~ "down")) %>% dplyr::mutate(Day = '18') %>% dplyr::mutate(challenge = 'salinity')  

DEGs_d18.OA         <- DE_d18.OA          %>% dplyr::filter(padj < 0.05) %>%  dplyr::mutate(dir = case_when(log2FoldChange > 0 ~ "up", log2FoldChange <0 ~ "down")) %>% dplyr::mutate(Day = '18')  %>% dplyr::mutate(challenge = 'OA')      


Master_DESeq2_results <- rbind(DEG_d2.Temperature[,c('TranscriptID','dir','Day', 'challenge')],
                               DEGs_d2.Salinity[,c('TranscriptID','dir','Day', 'challenge')], 
                               DEGs_d2.OA[,c('TranscriptID','dir','Day', 'challenge')],
                               DEGs_d18.Salinity[,c('TranscriptID','dir','Day', 'challenge')],
                               DEGs_d18.OA[,c('TranscriptID','dir','Day', 'challenge')]) # Master DESeq2 

```


```{r GO slim loop}

GO.terms
Master_goseq_results
Master_DESeq2_results
######################################################################################################################################################
###################################################################################################################################################### #
#
#
# PRIMARY EFFECT  GOSLIM ANALYSIS LOOP
#
#
###########################################################################################################################################

# call all the module colors and days to loop GOslim analysis 
GOslimLoop_vars <- unique(Master_goseq_results[c(8:10)]) 

# FOR LOOP FOR GOSLIM ANALYSIS AND CSV OUTPUTS 
for (i in 1:nrow(GOslimLoop_vars)) {
  # call the target dataset
  
  #goseq_res         <- Master_goseq_results %>%  dplyr::filter(dir %in% GOslimLoop_vars[i,1], Day %in% GOslimLoop_vars[i,2])
  
  goseq_res         <- Master_goseq_results %>%  dplyr::filter(dir %in% GOslimLoop_vars[i,1], Day %in% GOslimLoop_vars[i,2])
  
  if (GOslimLoop_vars[1,1] == 'up') {DESeq2_res <- Master_DESeq2_results %>%  dplyr::filter(dir == 'up') 
  } else {DESeq2_res <- Master_DESeq2_results %>%  dplyr::filter(dir = 'down') }
  DESeq2_res_2       <- DESeq2_res  %>%  
                          dplyr::filter(Day %in% GOslimLoop_vars[i,2]) %>%  
                          dplyr::filter(challenge %in% GOslimLoop_vars[i,3])
  gene_names         <- DESeq2_res_2$TranscriptID # all gene IDs in the particular WGCNA module 
  # Biological Function - run GOslim
  goseq_res_BP        <- goseq_res %>%  filter(ontology=="BP") # BP - all GO terms upregulated
  BP_GOcollection     <- GOCollection(goseq_res_BP$category)
  GOslims_BP          <- data.frame(goSlim(BP_GOcollection, slim, "BP")) #Find common parent terms to slim down our list
  GOslims_BP$category <- row.names(GOslims_BP) #save rownames as category
  
  # Molecular Function - run GOslim
  goseq_res_MF        <- goseq_res %>%  filter(ontology=="MF") # BP - all GO terms upregulated
  MF_GOcollection     <- GOCollection(goseq_res_MF$category)
  GOslims_MF          <- data.frame(goSlim(MF_GOcollection, slim, "MF")) #Find common parent terms to slim down our list
  GOslims_MF$category <- row.names(GOslims_MF) #save rownames as category
  
  # ====================================================================================
  # Get mapped terms - add to the GOslims datatable 
  # from Sam White's Biostars [post](https://support.bioconductor.org/p/128407/#128409).
  # ====================================================================================
  # Write function mappedIds to get the query terms that mapped to the slim categories 
  # ...in other words, add a column to your slim dataframe with all the GO terms from goseq
  mappedIds <-  function(df, collection, OFFSPRING) {  #the command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
    map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
    mapped <- lapply(map, intersect, ids(collection)) #Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) #Add column "go_terms" with matching terms 
    df #show resulting dataframe
  }
   
  BPslim_Mapped <- mappedIds(GOslims_BP, BP_GOcollection, GOBPOFFSPRING)
  MFslim_Mapped <- mappedIds(GOslims_MF, MF_GOcollection, GOMFOFFSPRING)

  # BIOLOGICAL PROCESS
  BPslim             <- BPslim_Mapped %>% dplyr::filter(!Term %in% "biological_process") %>% dplyr::filter(!Count == 0) #filter out empty slims and term "biological process" and slims with < 2 GO terms (omitted the Count>=2)
  BPsplitted         <- strsplit(as.character(BPslim$go_terms), ";") #split into multiple GO ids
  BPslim$BPsplitted  <- BPsplitted
  for (n in 1:nrow(BPslim)) {
    table       <- data.frame(GOlist = unlist(MFslim[n,6])) # call the MFsplitted column of characters and create a small table to filter
    table       <- unique(table)
    
    Pgen_module  <- Cvirginica_GOterms_MERGE %>% dplyr::filter(Cvirginica_TranscriptID %in% gene_names) # d7_Mod.Brown$X calls the gene names in the origin Module membership dataframe at the start of this script
    Pgen_loop    <- Pgen_module %>% dplyr::filter(Annotation_GO_ID %in% table$GOlist) # filter Gene IDs with the GO term
    Pgen_geneIDs <- Pgen_loop[2] # ommit the GO.terms to call unique gene calls 
    Pgen_geneIDs <- unique(Pgen_geneIDs) # call unique Gene calls (unique genes that had the GO term within each of the GOslim bins)
   
    BPslim$Gene.Count[n] <- nrow(Pgen_geneIDs)  # count of unique GeneIDs in each GOslim bin
    BPslim$Gene.IDs[[n]] <- vapply((Pgen_geneIDs$gene.ID), paste, collapse = ";", character(1L))
    } # end n in 1:nrow
  
  BPslim_A <- data.frame(Term = rep.int(BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
  BPslim_B <- merge(BPslim_A, BPslim, by="Term") #Add back counts, term, and category info
  BPslim_C <- unique(setDT(BPslim_B)[order(go_term, -Gene.Count)], by = "category") #remove duplicate offspring terms, keeping only those in the larger umbrella term (Count number)
  BPslim_final <- BPslim_C[,c(1,5,3,2,6,8:9)]
  colnames(BPslim_final) <- c("slim_term", "slim_cat", "GO_count", "GO_terms", "GO_list", "Gene_count", "Gene_IDs")
  BPslim_final[["Gene_IDs"]] <- vapply(unname(BPslim_final$Gene_IDs), paste, collapse = ";", character(1L)) # convert from a list to simply a charaacter string with ; delimiter
  #BPslim_final <- data.frame(slim_term=BPslim_C$Term, slim_cat=BPslim_C$category, category=BPslim_C$go_term, Gene.Count=BPslim_C$Gene.Count, GO.Count=BPslim_C$Count, Gene.IDs=BPslim_C$Gene.IDs) #rename columns) #rename columns
  BPslim_final$module_day <- paste(GOslimLoop_vars[i,2], GOslimLoop_vars[i,1], sep = '_')
  
  
        if (nrow(goseq_res_BP) >0) {
        BPslim_GOterm_summary       <- BPslim_final[,c(1,2,5,7)]
        s <- strsplit(BPslim_GOterm_summary$GO_list, split = ";")
        BPslim_GOterm_summary_2     <- data.frame(slim_term = rep(BPslim_GOterm_summary$slim_term, sapply(s, length)),
                                                  Gene_IDs  = rep(BPslim_GOterm_summary$Gene_IDs, sapply(s, length)),
                                                   GO_terms = unlist(s))
        colnames(goseq_res_BP)[1]   <- 'GO_terms'
        BPslim_GOterm_summary_final <- merge(goseq_res_BP, BPslim_GOterm_summary_2, by = 'GO_terms')
        
        
        s_2 <- strsplit(BPslim_GOterm_summary_final$Gene_IDs, split = ";")
        BPslim_GOterm_gene_annotation <- data.frame(slim_term      = rep(BPslim_GOterm_summary_final$slim_term, sapply(s_2, length)),
                                                    over_represented_pvalue = rep(BPslim_GOterm_summary_final$over_represented_pvalue, sapply(s_2, length)),
                                                    GO_terms       = rep(BPslim_GOterm_summary_final$GO_terms, sapply(s_2, length)),
                                                    term           = rep(BPslim_GOterm_summary_final$term, sapply(s_2, length)),
                                                    ontology       = rep(BPslim_GOterm_summary_final$ontology, sapply(s_2, length)),
                                                    Day            = rep(BPslim_GOterm_summary_final$Day, sapply(s_2, length)),
                                                    dir            = rep(BPslim_GOterm_summary_final$dir, sapply(s_2, length)),
                                                    PgenIDs        = unlist(s_2))
        BP_master_gene_reference <- merge(BPslim_GOterm_gene_annotation, Pgen_reference, by = "PgenIDs")
        
        } else (c(BPslim_GOterm_summary_final = NULL, BP_master_gene_reference = NULL))
  
  
  # MOLECULAR FUNCTION
  MFslim             <-  MFslim_Mapped %>% 
    dplyr::filter(!Term %in% "molecular_function") %>% 
    dplyr::filter(!go_terms == "")#filter out empty slims and term "biological process" (omitted the Count>=2)
 
  MFsplitted         <- strsplit(as.character(MFslim$go_terms), ";") #split into multiple GO ids
  
  MFslim$MFsplitted  <- MFsplitted
  
  for (m in 1:nrow(MFslim)) {
    table       <- data.frame(GOlist = unlist(MFslim[m,6])) # call the MFsplitted column of characters and create a small table to filter
    table       <- unique(table)
    
    Pgen_module  <- Cvirginica_GOterms_MERGE %>% dplyr::filter(Cvirginica_TranscriptID %in% gene_names) # d7_Mod.Brown$X calls the gene names in the origin Module membership dataframe at the start of this script
    Pgen_loop    <- Pgen_module %>% dplyr::filter(Annotation_GO_ID %in% table$GOlist) # filter Gene IDs with the GO term
    Pgen_geneIDs <- Pgen_loop[2] # ommit the GO.terms to call unique gene calls 
    Pgen_geneIDs <- unique(Pgen_geneIDs) # call unique Gene calls (unique genes that had the GO term within each of the GOslim bins)
    
    MFslim$Gene.Count[1] <- nrow(Pgen_geneIDs)  # count of unique GeneIDs in each GOslim bin
    MFslim$Gene.IDs[[m]] <- vapply((Pgen_geneIDs$Cvirginica_TranscriptID), paste, collapse = ";", character(1L))
    } # end m in 1:nrow
  
    MFslim_A <- data.frame(Term = rep.int(MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all
    MFslim_B <- merge(MFslim_A, MFslim, by="Term") #Add back counts, term, and category info
    MFslim_C <- unique(setDT(MFslim_B)[order(go_term, -Gene.Count)], by = "category") #remove duplicate offspring terms, keeping only those in the larger umbrella term (Count number)
    MFslim_final <- MFslim_C[,c(1,5,3,2,6,8:9)]
    colnames(MFslim_final) <- c("slim_term", "slim_cat", "GO_count", "GO_terms", "GO_list", "Gene_count", "Gene_IDs")
    MFslim_final[["Gene_IDs"]] <- vapply(unname(MFslim_final$Gene_IDs), paste, collapse = ";", character(1L)) # convert from a list to simply a charaacter string with ; delimiter
    # MFslim_final <- data.frame(slim_term=MFslim_C$Term, slim_cat=MFslim_C$category, category=MFslim_C$go_term, Gene.Count=MFslim_C$Gene.Count, GO.Count=MFslim_C$Count) #rename columns) #rename columns
    MFslim_final$module_day <- paste(GOslimLoop_vars[i,2], GOslimLoop_vars[i,1], sep = '_')
    
    
    
            if (nrow(goseq_res_MF) >0) {
        MFslim_GOterm_summary       <- MFslim_final[,c(1,2,5,7)]
        s <- strsplit(MFslim_GOterm_summary$GO_list, split = ";")
        MFslim_GOterm_summary_2     <- data.frame(slim_term = rep(MFslim_GOterm_summary$slim_term, sapply(s, length)),
                                                  Gene_IDs  = rep(MFslim_GOterm_summary$Gene_IDs, sapply(s, length)),
                                                   GO_terms = unlist(s))
        colnames(goseq_res_MF)[1]   <- 'GO_terms'
        MFslim_GOterm_summary_final <- merge(goseq_res_MF, MFslim_GOterm_summary_2, by = 'GO_terms')
        
        
        s_2 <- strsplit(MFslim_GOterm_summary_final$Gene_IDs, split = ";")
        MFslim_GOterm_gene_annotation <- data.frame(slim_term      = rep(MFslim_GOterm_summary_final$slim_term, sapply(s_2, length)),
                                                    over_represented_pvalue = rep(MFslim_GOterm_summary_final$over_represented_pvalue, sapply(s_2, length)),
                                                    GO_terms       = rep(MFslim_GOterm_summary_final$GO_terms, sapply(s_2, length)),
                                                    term           = rep(MFslim_GOterm_summary_final$term, sapply(s_2, length)),
                                                    ontology       = rep(MFslim_GOterm_summary_final$ontology, sapply(s_2, length)),
                                                    Day            = rep(MFslim_GOterm_summary_final$Day, sapply(s_2, length)),
                                                    dir            = rep(MFslim_GOterm_summary_final$dir, sapply(s_2, length)),
                                                    PgenIDs        = unlist(s_2))
        MF_master_gene_reference <- merge(MFslim_GOterm_gene_annotation, Pgen_reference, by = "PgenIDs")
        
        } else (c(MFslim_GOterm_summary_final = NULL, MF_master_gene_reference = NULL))
    
    
    
    
    # # save GOslim final datasets for BP and MF of each  module in their respective folder(s) by Day
    # write.csv(MFslim_final, file = paste("../Output/GO/DESeq2_goseq/", GOslimLoop_vars[i,2], "/GOslim_MolFunction_",GOslimLoop_vars[i,1], "regulated.csv", sep ='')) # save csv file              
    # write.csv(BPslim_final, file = paste("../Output/GO/DESeq2_goseq/", GOslimLoop_vars[i,2], "/GOslim_BiolProc_",GOslimLoop_vars[i,1], "regulated.csv", sep ='')) # save csv file       
    # 
    # write.csv(MFslim_GOterm_summary_final, file = paste("../Output/GO/DESeq2_goseq/", GOslimLoop_vars[i,2], "/GOterms_and_GOslim_MolFunction_",GOslimLoop_vars[i,1], "regulated.csv", sep ='')) # save csv file
    # write.csv(BPslim_GOterm_summary_final, file = paste("../Output/GO/DESeq2_goseq/", GOslimLoop_vars[i,2], "/GOterms_and_GOslim_BiolProc_",GOslimLoop_vars[i,1], "regulated.csv", sep ='')) # save csv file
    # 
    # write.csv(MF_master_gene_reference, file = paste("../Output/GO/DESeq2_goseq/", GOslimLoop_vars[i,2], "/GOterms_and_GOslim_MolFunction_",GOslimLoop_vars[i,1], "regulated_GENE_REFERENCE.csv", sep ='')) # save csv
    # write.csv(BP_master_gene_reference, file = paste("../Output/GO/DESeq2_goseq/", GOslimLoop_vars[i,2], "/GOterms_and_GOslim_BiolProc_",GOslimLoop_vars[i,1], "regulated_GENE_REFERENCE.csv", sep ='')) # save csv       
    # 
    # print(paste(GOslimLoop_vars[i,2], GOslimLoop_vars[i,1], "done", sep = ' '))
}
    
```



PLOTTING  GOslim --log1- pvalue plots 
PRIMARY  EFFECT
```{r}
# read in the outputs from the previous... (note- we did not run goslim fror Day0 DEGs because there were very very few)

# D7_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day7/primary_effect/GOterms_and_GOslim_BiolProc_Upregulated.csv")
# D7_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day7/primary_effect/GOterms_and_GOslim_BiolProc_Downregulated.csv")
# D7_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day7/primary_effect/GOterms_and_GOslim_MolFunction_Upregulated.csv")
# D7_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day7/primary_effect/GOterms_and_GOslim_MolFunction_Downregulated.csv")
# 
# D14_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOterms_and_GOslim_BiolProc_Upregulated.csv")
# D14_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOterms_and_GOslim_BiolProc_Downregulated.csv")
# D14_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOterms_and_GOslim_MolFunction_Upregulated.csv")
# D14_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOterms_and_GOslim_MolFunction_Downregulated.csv")
# 
# D21_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOterms_and_GOslim_BiolProc_Upregulated.csv")
# D21_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOterms_and_GOslim_BiolProc_Downregulated.csv")
# D21_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOterms_and_GOslim_MolFunction_Upregulated.csv")
# D21_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOterms_and_GOslim_MolFunction_Downregulated.csv")



############################################################################################################## 3
#  Molecular Function  

Master_goseq_results
GOres_Day2_MolFunction <- Master_goseq_results %>% dplyr::filter(ontology =='MF' & Day == '2')

MF_Day2_Plot <- ggplot(data = GOres_Day2_MolFunction,aes(x = dir, y = forcats::fct_rev(term))) + 
                      geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
                      scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(GOres_Day2_MolFunction$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(GOres_Day2_MolFunction$over_represented_pvalue)))),by=2))  +
                      facet_grid(~challenge, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                      theme_bw() + 
                      theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                         panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                         strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                         strip.text.x = element_text(size = 8, face = "bold"),
                                         axis.title.x = element_blank(),
                                         axis.title.y = element_text(size=8),
                                         axis.text = element_text(size = 8), legend.position = "right",
                                         plot.margin = unit(c(0,1,0,0.25), "cm"))+
                      ggtitle('GO analysis Day 2 larvae - Molecular Function') 
MF_Day2_Plot

pdf(paste("../Output/DESeq2/Day2_larva/GO_analysis/Day2_larvae_DEGs_MolFunction.pdf", sep =''), width=8,height=7)
print(MF_Day2_Plot)
dev.off()


############################################################################################################## 3
#  Biological Process    

GOres_Day2_BiolProcess <- Master_goseq_results %>% dplyr::filter(ontology =='BP' & Day == '2') %>%  dplyr::filter(!term %in% 'biological_process') 

BP_Day2_Plot <- ggplot(data = GOres_Day2_BiolProcess,aes(x = dir, y = forcats::fct_rev(term))) + 
                      geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
                      scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(GOres_Day2_MolFunction$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(GOres_Day2_MolFunction$over_represented_pvalue)))),by=2))  +
                      facet_grid(~challenge, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                      theme_bw() + 
                      theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                         panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                         strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                         strip.text.x = element_text(size = 8, face = "bold"),
                                         axis.title.x = element_blank(),
                                         axis.title.y = element_text(size=8),
                                         axis.text = element_text(size = 8), legend.position = "right",
                                         plot.margin = unit(c(0,1,0,0.25), "cm"))+
                      ggtitle('GO analysis Day 2 larvae - Biological Process') 
BP_Day2_Plot

pdf(paste("../Output/DESeq2/Day2_larva/GO_analysis/Day2_larvae_DEGs_BiolProcess.pdf", sep =''), width=8,height=7)
print(BP_Day2_Plot)
dev.off()

############################################################################################################## 3
#  Molecular Function  


GOres_Day18_MolFunction <- Master_goseq_results %>% dplyr::filter(ontology =='MF' & Day == '18')

MF_Day18_Plot <- ggplot(data = GOres_Day18_MolFunction,aes(x = dir, y = forcats::fct_rev(term))) + 
                      geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
                      scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(GOres_Day18_MolFunction$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(GOres_Day18_MolFunction$over_represented_pvalue)))),by=18))  +
                      facet_grid(~challenge, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                      theme_bw() + 
                      theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                         panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                         strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                         strip.text.x = element_text(size = 8, face = "bold"),
                                         axis.title.x = element_blank(),
                                         axis.title.y = element_text(size=8),
                                         axis.text = element_text(size = 8), legend.position = "right",
                                         plot.margin = unit(c(0,1,0,0.185), "cm"))+
                      ggtitle('GO analysis Day 18 larvae - Molecular Function') 
MF_Day18_Plot

pdf(paste("../Output/DESeq2/Day18_spat/GO_analysis/Day18_spat_DEGs_MolFunction.pdf", sep =''), width=8,height=7)
print(MF_Day18_Plot)
dev.off()

############################################################################################################## 3
#  Molecular Function  

GOres_Day18_BiolProcess <- Master_goseq_results %>% dplyr::filter(ontology =='BP' & Day == '18') %>%  dplyr::filter(!term %in% 'biological_process') 

BP_Day18_Plot <- ggplot(data = GOres_Day18_BiolProcess,aes(x = dir, y = forcats::fct_rev(term))) + 
                      geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
                      scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(GOres_Day18_MolFunction$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(GOres_Day18_MolFunction$over_represented_pvalue)))),by=18))  +
                      facet_grid(~challenge, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                      theme_bw() + 
                      theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                         panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                         strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                         strip.text.x = element_text(size = 8, face = "bold"),
                                         axis.title.x = element_blank(),
                                         axis.title.y = element_text(size=8),
                                         axis.text = element_text(size = 8), legend.position = "right",
                                         plot.margin = unit(c(0,1,0,0.185), "cm"))+
                      ggtitle('GO analysis Day 18 spat - Biological Process') 
BP_Day18_Plot


pdf(paste("../Output/DESeq2/Day18_spat/GO_analysis/Day18_spat_DEGs_BiolProcess.pdf", sep =''), width=8,height=7)
print(BP_Day18_Plot)
dev.off()

```


PLOTTING  GOslim --log1- pvalue plots 
SECOND EFFECT AND GROUP EFFECT PLOTS (DAY 7)
```{r}
# read in the outputs from the previous... (note- we did not run goslim fror Day0 DEGs because there were very very few)
# DAY 7 - SECOND TREATMENT EFFECT AMBIENT > MODERATE
# D7_Goslim_secondAvM_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day7/secondAvM/GOterms_and_GOslim_BiolProc_secondAvM_Upregulated.csv") #no upregualted terms! 
D7_Goslim_secondAvM_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day7/second_effect_AvM/GOterms_and_GOslim_BiolProc_secondAvM_Downregulated.csv")
# D7_Goslim_secondAvM_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day7/secondAvM/GOterms_and_GOslim_MolFunction_secondAvM_Upregulated.csv") #no upregualted terms! 
D7_Goslim_secondAvM_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day7/second_effect_AvM/GOterms_and_GOslim_MolFunction_secondAvM_Downregulated.csv")


# DAY 7 - GROUP TREATMENT EFFECT MA > AM
D7_Goslim_MAvAM_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GOterms_and_GOslim_BiolProc_MAvAM_Upregulated.csv") 
D7_Goslim_MAvAM_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GOterms_and_GOslim_BiolProc_MAvAM_Downregulated.csv")
# D7_Goslim_MAvAM_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GOterms_and_GOslim_MolFunction_MAvAM_Upregulated.csv") # no upregualted molecular function terms!
D7_Goslim_MAvAM_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GOterms_and_GOslim_MolFunction_MAvAM_Downregulated.csv")


############################################################################################################## 
# DAY 7 - MA  versus AM - Full model effect 
#  Biological Process Upregulated 
BP_UP_Plot_MAvAM <-ggplot(data = D7_Goslim_MAvAM_UP_BP, aes(x = effect, y = forcats::fct_rev(term))) + 
  geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
  scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(BP_UP_Master$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(BP_UP_Master$over_represented_pvalue)))),by=2))  +
  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                     strip.text.x = element_text(size = 8, face = "bold"),
                     axis.title.x = element_blank(),
                     axis.title.y = element_text(size=8),
                     axis.text = element_text(size = 8), legend.position = "right",
                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
  ggtitle('Biological Process: Day 7 Full model effect MA versus AM') +
  # facet_wrap(~ (substr(slim_term, 1,30)))
  facet_wrap((substr(slim_term, 1,30)) ~., scales="free_y", ncol= 1, strip.position="right")

pdf(paste("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/Biol_Process_Upregulated_DEGs_MAvAM.pdf", sep =''), width=15, height=15)
print(BP_UP_Plot_MAvAM)
dev.off()

#  Biological Process Upregulated 
BP_DOWN_Plot_MAvAM <-ggplot(data = D7_Goslim_MAvAM_DOWN_BP, aes(x = effect, y = forcats::fct_rev(term))) + 
  geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
  scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(BP_UP_Master$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(BP_UP_Master$over_represented_pvalue)))),by=2))  +
  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                     strip.text.x = element_text(size = 8, face = "bold"),
                     axis.title.x = element_blank(),
                     axis.title.y = element_text(size=8),
                     axis.text = element_text(size = 8), legend.position = "right",
                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
  ggtitle('Biological Process: Day 7 Full model effect MA versus AM') +
  # facet_wrap(~ (substr(slim_term, 1,30)))
  facet_wrap((substr(slim_term, 1,30)) ~., scales="free_y", ncol= 1, strip.position="right")

pdf(paste("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/Biol_Process_Downregulated_DEGs_MAvAM.pdf", sep =''), width=15, height=10)
print(BP_DOWN_Plot_MAvAM)
dev.off()


############################################################################################################## 3
#  Molecular Function Upregulated 
MF_DOWN_Plot_MAvAM <-ggplot(data = D7_Goslim_MAvAM_DOWN_MF, aes(x = effect, y = forcats::fct_rev(term))) + 
  geom_tile(aes(fill= -log10(over_represented_pvalue), width = 1)) + 
  scale_fill_gradient(low = "grey95", high = "black",limits=c(0, (ceiling(max(-log10(MF_UP_Master$over_represented_pvalue))))), breaks=seq(0, (ceiling(max(-log10(MF_UP_Master$over_represented_pvalue)))),by=2))  +
  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                     strip.text.x = element_text(size = 8, face = "bold"),
                     axis.title.x = element_blank(),
                     axis.title.y = element_text(size=8),
                     axis.text = element_text(size = 8), legend.position = "right",
                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
  ggtitle('Molecular Function: Day 7 Full model effect MA versus AM') +
  # facet_wrap(~ (substr(slim_term, 1,30)))
  facet_wrap((substr(slim_term, 1,30)) ~., scales="free_y", ncol= 1, strip.position="right")

pdf(paste("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/Mol_Function_Downregulated_DEGs_MAvAM.pdf", sep =''), width=15, height=10)
print(MF_DOWN_Plot_MAvAM)
dev.off()




```




PLOTTING  GOslim - gene count plots 
```{r}
# read in the outputs from the previous... (note- we did not run goslim fror Day0 DEGs because there were very very few)

D7_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day7/GOslim_BiolProc_Upregulated.csv")
D7_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day7/GOslim_BiolProc_Downregulated.csv")

D7_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day7/GOslim_MolFunction_Upregulated.csv")
D7_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day7/GOslim_MolFunction_Downregulated.csv")


D14_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOslim_BiolProc_Upregulated.csv")
D14_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOslim_BiolProc_Downregulated.csv")

D14_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOslim_MolFunction_Upregulated.csv")
D14_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day14/GOslim_MolFunction_Downregulated.csv")



D21_Goslim_UP_BP    <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOslim_BiolProc_Upregulated.csv")
D21_Goslim_DOWN_BP  <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOslim_BiolProc_Downregulated.csv")

D21_Goslim_UP_MF    <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOslim_MolFunction_Upregulated.csv")
D21_Goslim_DOWN_MF  <- read.csv("../Output/GO/DESeq2_goseq/Day21/GOslim_MolFunction_Downregulated.csv")



#MASTER FOR PLOTTING 
# biological process upregulated 
BP_UP_Master <- rbind(D7_Goslim_UP_BP, D14_Goslim_UP_BP, D21_Goslim_UP_BP)
BP_UP_Master$module_day <- factor(BP_UP_Master$module_day , levels = c("Day7_Up", "Day14_Up", "Day21_Up"))
# biological process donwregualted 
BP_DOWN_Master <- rbind(D7_Goslim_DOWN_BP, D14_Goslim_DOWN_BP, D21_Goslim_DOWN_BP)
BP_DOWN_Master$module_day <- factor(BP_DOWN_Master$module_day , levels = c("Day7_Down", "Day14_Down", "Day21_Down"))


MF_UP_Master <- rbind(D7_Goslim_UP_MF, D14_Goslim_UP_MF, D21_Goslim_UP_MF)
MF_UP_Master$module_day <- factor(MF_UP_Master$module_day , levels = c("Day7_Up", "Day14_Up", "Day21_Up"))
# biological process donwregualted 
MF_DOWN_Master <- rbind(D7_Goslim_DOWN_MF, D14_Goslim_DOWN_MF, D21_Goslim_DOWN_MF)
MF_DOWN_Master$module_day <- factor(MF_DOWN_Master$module_day , levels = c("Day7_Down", "Day14_Down", "Day21_Down"))


# plots
BP_UP_Plot <- BP_UP_Master %>% dplyr::filter(Gene_count >= 5) %>%  
              ggplot(aes(x = "BP_Upreg", y = forcats::fct_rev(slim_term))) + 
                geom_tile(aes(fill=Gene_count, width = 1)) + 
                scale_fill_gradient(low = "grey95", high = "grey10",limits=c(0, 20), breaks=seq(0,20,by=5)) +
                facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                   panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                   strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                   strip.text.x = element_text(size = 8, face = "bold"),
                                   axis.title.x = element_blank(),
                                   axis.title.y = element_text(size=8),
                                   axis.text = element_text(size = 8), legend.position = "right",
                                   plot.margin = unit(c(0,1,0,0.25), "cm"))+
                ggtitle('GOslim Biological Process: Upregulated DEGs Primary Effect All')

BP_DOWN_Plot <- BP_DOWN_Master %>% dplyr::filter(Gene_count >= 5) %>%  
                ggplot(aes(x = "BP_Downreg", y = forcats::fct_rev(slim_term))) + 
                  geom_tile(aes(fill=Gene_count, width = 1)) + 
                  scale_fill_gradient(low = "grey95", high = "grey10",limits=c(0, 20), breaks=seq(0,20,by=5)) +
                  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                     strip.text.x = element_text(size = 8, face = "bold"),
                                     axis.title.x = element_blank(),
                                     axis.title.y = element_text(size=8),
                                     axis.text = element_text(size = 8), legend.position = "right",
                                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
                  ggtitle('GOslim Biological Process: Downregulated DEGs Primary Effect All')





MF_UP_Plot <-MF_UP_Master %>% dplyr::filter(Gene_count >= 2) %>% 
             ggplot(aes(x = "MF_Upreg", y = forcats::fct_rev(slim_term))) + 
                geom_tile(aes(fill=Gene_count, width = 1)) + 
                scale_fill_gradient(low = "grey95", high = "grey10",limits=c(0, 20), breaks=seq(0,20,by=5)) +
                facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                   panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                   strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                   strip.text.x = element_text(size = 8, face = "bold"),
                                   axis.title.x = element_blank(),
                                   axis.title.y = element_text(size=8),
                                   axis.text = element_text(size = 8), legend.position = "right",
                                   plot.margin = unit(c(0,1,0,0.25), "cm"))+
                ggtitle('GOslim Mol Function: Upregulated DEGs Primary Effect All')

MF_DOWN_Plot <-MF_DOWN_Master %>% dplyr::filter(Gene_count >= 2) %>% 
                 ggplot(aes(x = "MF_Downreg", y = forcats::fct_rev(slim_term))) + 
                  geom_tile(aes(fill=Gene_count, width = 1)) + 
                  scale_fill_gradient(low = "grey95", high = "grey10",limits=c(0, 20), breaks=seq(0,20,by=5)) +
                  facet_grid(~module_day, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                     panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                     strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                     strip.text.x = element_text(size = 8, face = "bold"),
                                     axis.title.x = element_blank(),
                                     axis.title.y = element_text(size=8),
                                     axis.text = element_text(size = 8), legend.position = "right",
                                     plot.margin = unit(c(0,1,0,0.25), "cm"))+
                  ggtitle('GOslim Mol Function: Downregulated DEGs Primary Effect All')




# SAVE PLOTS  ------------------------------------------------------------------------------------------------- #
pdf(paste("../Output/GO/DESeq2_goseq/GOslim_PrimEff_BiolProcess.pdf", sep =''), width=20, height=8)
print(ggarrange(BP_UP_Plot, BP_DOWN_Plot,         
                plotlist = NULL,
                ncol = 2,
                nrow = 1,
                labels = NULL))
dev.off() 


pdf(paste("../Output/GO/DESeq2_goseq/GOslim_PrimEff_MolFunction.pdf", sep =''), width=20, height=8)
print(ggarrange(MF_UP_Plot, MF_DOWN_Plot,         
                plotlist = NULL,
                ncol = 2,
                nrow = 1,
                labels = NULL))
dev.off() 

```







prep
```{r prep data}
# add ontology column  ---------------------------------------------------------------------------------------- #
All.UP_BPslim_final$Ontolgy <- "BP_Upregulated_all"
All.UP_MFslim_final$Ontolgy <- "MF_Upregulated_all"

All.DOWN_BPslim_final$Ontolgy <- "BP_Downregulated_all"
All.DOWN_MFslim_final$Ontolgy <- "MF_Downregulated_all"

# BIND DATA TO FACET THE PLOTS BY UP AND DOWN REG   ---------------------------------------------------------- #
# bp
BP_All <- rbind(All.UP_BPslim_final, All.DOWN_BPslim_final) # merge the data by binding rows 
BP_All$Ont <- "BP" # create a new common clumn to call in the plot 
BP_All$Ontolgy <- factor(BP_All$Ontolgy, levels = c("BP_Upregulated_all", "BP_Downregulated_all"))# reorder the facotr level for the facet wrap plot 
BP_All_filtered <- BP_All %>%  dplyr::filter(Gene.Count > 1) # ommit all with gene counts >1
# mf
MF_All <- rbind(All.UP_MFslim_final, All.DOWN_MFslim_final) # merge the data by binding rows 
MF_All$Ont <- "MF" # create a new common clumn to call in the plot 
MF_All$Ontolgy <- factor(MF_All$Ontolgy, levels = c("MF_Upregulated_all", "MF_Downregulated_all")) # reorder the facotr level for the facet wrap plot 
MF_All_filtered <- MF_All %>%  dplyr::filter(Gene.Count > 1) # ommit all with gene counts >1
```

Stacked bar plots -  Gene Counts within GOslim term (NOTE: there are likely duplicates (or more) BETWEEN the GOslim bins - however EACH bin contains UNIQUE genes)
```{r barplots - ALL DEGs PrimEffect} 
# create color palette  --------------------------------------------------------------------------------------- #
library(RColorBrewer)
cbp1 <- c("#999999", "#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7") # color blind pallette
OrangeRed <- brewer.pal(2, "OrRd") 
GreenBlue <- brewer.pal(2, "GnBu") 


# PLOTTING --------------------------------------------------------------------------------------------------- #
BP_Plot_PrimEffect_All <- ggplot(data = BP_All_filtered, aes(x = Ont, y = Gene.Count, fill=forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE))) + # BP plot
                               geom_bar(color = "white",size=1.5,stat="identity") + # call bar and create lines between each slim term
                               scale_fill_manual(values =  rev( colorRampPalette(OrangeRed)( (nrow(BP_All)) ) ) )+ # fill with pallette accomodating the max nmber of slim terms
                               geom_text(aes(label = forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE)), colour = "black", position="stack", size = 7, vjust = 1.25) + # add labels
                               theme_classic() + # classic theme
                               ylim(0, 300) + 
                               theme(legend.position = "none",text = element_text(size=25)) +
                               facet_wrap(~Ontolgy) # facet by the upregulated and downregulated datasets

MF_Plot_PrimEffect_All <- ggplot(data = MF_All_filtered, aes(x = Ont, y = Gene.Count, fill=forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE))) + # MF plot
                               geom_bar(color = "white",size=1.5,stat="identity") + # call bar and create lines between each slim term
                               scale_fill_manual(values =  rev( colorRampPalette(GreenBlue)( (nrow(MF_All)) ) ) )+ # fill with pallette accomodating the max nmber of slim terms
                               geom_text(aes(label = forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE)), colour = "black", position="stack", size = 7, vjust = 1.25) + # add labels
                               theme_classic() + # classic theme
                               ylim(0, 100) +
                               theme(legend.position = "none",text = element_text(size=25)) +
                               facet_wrap(~Ontolgy) # facet by the upregulated and downregulated datasets

# SAVE PLOTS  ------------------------------------------------------------------------------------------------- #
pdf(paste("../Output/GO/DESeq2_goseq/GOslim_All_PrimEff_BP.pdf", sep =''), width=15, height=20)
print(ggarrange(BP_Plot_PrimEffect_All,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()     

pdf(paste("../Output/GO/DESeq2_goseq/GOslim_All_PrimEff_MF.pdf", sep =''), width=15, height=20)
print(ggarrange(MF_Plot_PrimEffect_All,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()     

```

Geomtile Heatmaps
```{r geom_tile heatmaps- ALL DEGs PrimEffect}

# plots
BP_Plot_PrimEffect_All <-ggplot(data = BP_All_filtered, aes(x = Ont, y = slim_term)) + 
                          geom_tile(aes(fill=Gene.Count, width = 1)) + 
                          scale_fill_gradient(low = "thistle1", high = "steelblue4") +
                          facet_grid(~Ontolgy, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                          theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                             panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                             strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                             strip.text.x = element_text(size = 8, face = "bold"),
                                             axis.title.x = element_blank(),
                                             axis.title.y = element_text(size=8),
                                             axis.text = element_text(size = 8), legend.position = "right",
                                             plot.margin = unit(c(0,1,0,0.25), "cm"))+
                          ggtitle('GOslim Biological Process: DESeq2 DEGs PrimEffect')

# plot MF
MF_Plot_PrimEffect_All <-ggplot(data = MF_All_filtered, aes(x = Ont, y = slim_term)) + 
                          geom_tile(aes(fill=Gene.Count, width = 1)) + 
                          scale_fill_gradient(low = "azure2", high = "springgreen4") +
                          facet_grid(~Ontolgy, labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                          theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                                             panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                                             strip.text.y = element_text(angle=0, size = 8, face = "bold"),
                                             strip.text.x = element_text(size = 8, face = "bold"),
                                             axis.title.x = element_blank(),
                                             axis.title.y = element_text(size=8),
                                             axis.text = element_text(size = 8), legend.position = "right",
                                             plot.margin = unit(c(0,1,0,0.25), "cm"))+
                          ggtitle('GOslim Molecular Function: DESeq2 DEGs PrimEffect')

pdf(paste("../Output/GO/DESeq2_goseq/GOslim_All_PrimEff_Heatmap.pdf", sep =''), width=20, height=8)
print(ggarrange(BP_Plot_PrimEffect_All, MF_Plot_PrimEffect_All,         
                plotlist = NULL,
                ncol = 2,
                nrow = 1,
                labels = NULL))
dev.off() 
```


Follow-up from the stacked blots above...
What are the genesin the GOslimss of interest
Extract genes in GOslim bins of interest and plots their LFC results (DESeq2)
```{r} 
# prep data for the for loop plotting...
# add column 'DE.direction'
All.DOWN_MFslim$DE.direction  <- "downregulated"
All.DOWN_BPslim$DE.direction  <- "downregulated"
All.UP_MFslim$DE.direction    <- "upregulated"
All.UP_BPslim$DE.direction    <- "upregulated"

# add column 'Experiment.note'
All.DOWN_MFslim$Experiment.note  <- "All_Days_PrimaryEfffect"
All.DOWN_BPslim$Experiment.note  <- "All_Days_PrimaryEfffect"
All.UP_MFslim$Experiment.note    <- "All_Days_PrimaryEfffect"
All.UP_BPslim$Experiment.note    <- "All_Days_PrimaryEfffect"

# ommit rownames
rownames(All.DOWN_MFslim)<-NULL
rownames(All.DOWN_BPslim)<-NULL
rownames(All.UP_MFslim)  <-NULL
rownames(All.UP_BPslim)  <-NULL

# merge data frames to build a master GOslim
Master_GOslim <- rbind(All.DOWN_MFslim[,c(10,7,9,8,3,4,5)], All.DOWN_BPslim[,c(10,7,9,8,3,4,5)], All.UP_MFslim[,c(10,7,9,8,3,4,5)], All.UP_BPslim[,c(10,7,9,8,3,4,5)]) # merge for a master table with rbind

# load DESeq2 results 
All_DEGs_Primary.Ambieant_v_Moderate<-read.csv("C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/10cpm/DE_PrimaryTreatment.All.csv")
All_DEGs_Primary.Ambieant_v_Moderate <- All_DEGs_Primary.Ambieant_v_Moderate[2:(ncol(All_DEGs_Primary.Ambieant_v_Moderate))] # ommit the first column - DESEq2 results data 
names(All_DEGs_Primary.Ambieant_v_Moderate)[1] <- 'Gene.ID' # rename the column 

Pgen_annot <- Geoduck_annotation[,c(1,7)]# call only the gene ID and gene names in the annotation csv
names(Pgen_annot) <- c('Gene.ID', 'Gene.name') # rename the columns 
Pgen_annot$Gene.name <- str_extract(Pgen_annot$Gene.name, "[^(]+") # call all gene names before the delimiter ( when the EC code is included)

# loop through the gene IDsin each GOslim bin, DESeq2 results heatmap facetted by sampling day in for loop
for (i in 1:nrow(Master_GOslim)) {
  
  genes <- data.frame(geneids = unlist(Master_GOslim[i,4]))
  
  if (Master_GOslim[i,3] == "downregulated") {
  
    DEG_data <- All_DEGs_Primary.Ambieant_v_Moderate %>% dplyr::filter(down %in% TRUE)  # filter only downregualted genes
    DEG_data_loop <- DEG_data %>% dplyr::filter(Gene.ID %in% genes$geneids)
    DEG_data_merge <- merge(Pgen_annot, DEG_data_loop, by ="Gene.ID")
    
    DEG_data_merge$Day_Trmt <- str_extract(DEG_data_merge$Day_Trmt, "[^_]+")
    
    Plot <- DEG_data_merge %>% #mutate(Day_Trmt = fct_relevel(Day_Trmt,  "Day7", "Day14", "Day21")) %>% 
              ggplot(aes(x = "NA", y = Gene.name)) + 
                geom_tile(aes(fill=log2FoldChange, width = 1)) + 
                scale_fill_gradient(low = "orangered3", high = "orange1") +
                facet_grid(~Day_Trmt, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                strip.text.y = element_text(angle=0, size = 11, face = "bold"),
                strip.text.x = element_text(size = 12, face = "bold"),
                axis.title.x = element_blank(),
                axis.title.y = element_text(size=15),
                axis.text = element_text(size = 12), legend.position = "right",
                plot.margin = unit(c(0,1,0,0.25), "cm"))+
                ggtitle(Master_GOslim[i,5])
    pdf(paste("../Output/GO/DESeq2_goseq/All_PrimEff_GOslim/Downregulated/",Master_GOslim[i,5],".pdf", sep =''), width=18, height=10)
    print(Plot)
    dev.off()
  
    } else {
        DEG_data <- All_DEGs_Primary.Ambieant_v_Moderate %>% dplyr::filter(up %in% TRUE)  # filter only downregualted genes
        DEG_data_loop <- DEG_data %>% dplyr::filter(Gene.ID %in% genes$geneids)
        DEG_data_merge <- merge(Pgen_annot, DEG_data_loop, by ="Gene.ID")
        
        DEG_data_merge$Day_Trmt <- str_extract(DEG_data_merge$Day_Trmt, "[^_]+")
        
        Plot <- DEG_data_merge %>% #mutate(Day_Trmt = fct_relevel(Day_Trmt,  "Day7", "Day14", "Day21")) %>% 
                  ggplot(aes(x = "NA", y = Gene.name)) + 
                    geom_tile(aes(fill=log2FoldChange, width = 1)) + 
                    scale_fill_gradient(low = "palegreen1", high = "green4") +
                    facet_grid( ~Day_Trmt, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                    theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                    panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                    strip.text.y = element_text(angle=0, size = 11, face = "bold"),
                    strip.text.x = element_text(size = 12, face = "bold"),
                    axis.title.x = element_blank(),
                    axis.title.y = element_text(size=15),
                    axis.text = element_text(size = 12), legend.position = "right",
                    plot.margin = unit(c(0,1,0,0.25), "cm"))+
                    ggtitle(Master_GOslim[i,5])
        pdf(paste("../Output/GO/DESeq2_goseq/All_PrimEff_GOslim/Upregulated/",Master_GOslim[i,5],".pdf", sep =''), width=18, height=10)
        print(Plot)
        dev.off()
  } # end of if else statement

} # end of for loop


```


```{r Constant.primary.GO table}
GO_Constant.UP_summary <- GO_Constant.UP %>% dplyr::select(c('category','numDEInCat','numInCat', 'term', 'ontology')) %>% dplyr::filter(!is.na(term)) %>% dplyr::filter(numDEInCat > 0)
GO_Constant.UP_summary$DE_dir <- "upregulated"
GO_Constant.DOWN_summary <- GO_Constant.DOWN %>% dplyr::select(c('category','numDEInCat','numInCat', 'term', 'ontology')) %>% dplyr::filter(!is.na(term)) %>% dplyr::filter(numDEInCat > 0)
GO_Constant.DOWN_summary$DE_dir <- "downregulated"

GO_constant.Master <- rbind(GO_Constant.UP_summary,GO_Constant.DOWN_summary)
GO_constant.Master$perc.occurance <- GO_constant.Master$numDEInCat / GO_constant.Master$numInCat *100

#knitr
knitr::kable(GO_constant.Master, caption = "Summary table of constant GO terms Days 7, 14, 21")
# Write csv
write.csv(GO_constant.Master,paste(path_out,"Constant.Primary_GO.terms.csv"))
```



Geom_Segment Plotting: -log1- p alue all primary efffect down regualted and upregulated genes (figure 2B)
# SECOND EFFECT (A v M main effect AND full group effect MA versus AM)
```{r All Upreg and Downreg GO terms in response to priming}

# Build a master list of all genes and GO terms 'Pgen_GOterms2'
Geoduck_annotation <- read.delim2(file="C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Seq_details/Panopea-generosa-genes-annotations.txt", header=F) # Load themaster Pgenerosa gene list 
Pgen_GOterms <- Geoduck_annotation %>% dplyr::select(c('V1','V8')) # select only two columns - those with the gene IDs and those with the GO terms
Pgen_GOterms2 <- strsplit(Pgen_GOterms$V8, split = "; ") # create a string splitting by delimiter '; ' - view the data to see that this separates each GO term entry in the string
Pgen_GOterms2 <- data.frame(gene.ID = rep(Pgen_GOterms$V1, sapply(Pgen_GOterms2, length)), Go.terms = unlist(Pgen_GOterms2)) # create new dataframe 'Pgen_GOterms2' listing genes for each GO term (MUCH longer!)
Pgen_GOterms2 <- na.omit(Pgen_GOterms2) # ommit the NAs  - genes without GO annotation

d7.UP.second_AvM.05    <- read.csv("../Output/GO/DESeq2_goseq/Day7/second_effect_AvM/GO.05.Day7_SecondEffect_AvM_Upregulated.csv")
d7.UP.second_AvM.05$Dir = "Upregulated"
d7.DOWN.second_AvM.05  <- read.csv("../Output/GO/DESeq2_goseq/Day7/second_effect_AvM/GO.05.Day7_SecondEffect_AvM_Downregulated.csv")
d7.DOWN.second_AvM.05$Dir = "Downregulated"

d7.UP.MAvAM.05    <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GO.05.Day7_GroupEffect_MAvAM_Upregulated.csv")
d7.UP.MAvAM.05$Dir = "Upregulated"
d7.DOWN.MAvAM.05  <- read.csv("../Output/GO/DESeq2_goseq/Day7/group_effect_MAvAM/GO.05.Day7_GroupEffect_MAvAM_Downregulated.csv")
d7.DOWN.MAvAM.05$Dir = "Downregulated"

# Master ALL DESeq2 significant modules with treatment
Second_goseq_results      <- rbind(d7.UP.second_AvM.05, d7.DOWN.second_AvM.05)
MAvAM_goseq_results      <- rbind(d7.UP.MAvAM.05, d7.DOWN.MAvAM.05)



# PLOTS 
# Second Treatment effect Ambient > Moderate 
AvM_BiolProcess <- Second_goseq_results %>%  dplyr::filter(ontology %in% ('BP')) %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Biological Process: Day 7 Second Treatment Ambient v Moderate") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Dir ~., scales="free_y", ncol= 1, strip.position="right")


AvM_MolFunction <- Second_goseq_results %>%  dplyr::filter(ontology %in% ('MF')) %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Molecular Function: Day 7 Second Treatment Ambient v Moderate") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Dir ~., scales="free_y", ncol= 1, strip.position="right")

pdf(paste("../Output/GO/DESeq2_goseq/Day7_SecondEffect_AvM_plots/GOplot_DESeq2_Day7_AvM_BP.pdf", sep =''), width=10, height=12)
print(ggarrange(AvM_BiolProcess,         
                plotlist = NULL,
                ncol = 1,
                nrow = 2,
            labels = NULL))
dev.off()   

pdf(paste("../Output/GO/DESeq2_goseq/Day7_SecondEffect_AvM_plots/GOplot_DESeq2_Day7_AvM_MF.pdf", sep =''), width=25, height=8)
print(ggarrange(AvM_MolFunction,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()   


# Second Treatment effect Ambient > Moderate 
MVvAM_BiolProcess <- MAvAM_goseq_results %>%  dplyr::filter(ontology %in% ('BP')) %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Biological Process: Day 7 Full Model MA v AM") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Dir ~., scales="free_y", ncol= 1, strip.position="right")


MVvAM_MolFunction <- MAvAM_goseq_results %>%  dplyr::filter(ontology %in% ('MF')) %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Molecular Function: Day 7 Full Model MA v AM") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Dir ~., scales="free_y", ncol= 1, strip.position="right")


pdf(paste("../Output/GO/DESeq2_goseq/Day7_FullModEffect_MAvAM_plots/GOplot_DESeq2_Day7_MAvAM_BP.pdf", sep =''), width=10, height=12)
print(ggarrange(MVvAM_BiolProcess,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()   

pdf(paste("../Output/GO/DESeq2_goseq/Day7_FullModEffect_MAvAM_plots/GOplot_DESeq2_Day7_MAvAM_MF.pdf", sep =''), width=10, height=8)
print(ggarrange(MVvAM_MolFunction,         
                plotlist = NULL,
                ncol = 1,
                nrow = 2,
            labels = NULL))
dev.off()      


```

Geom_Segment Plotting: -log1- p alue all primary efffect down regualted and upregulated genes (figure 2B)
# PRIMARY Effect
```{r All Upreg and Downreg GO terms in response to priming}

# Build a master list of all genes and GO terms 'Pgen_GOterms2'
Geoduck_annotation <- read.delim2(file="C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Seq_details/Panopea-generosa-genes-annotations.txt", header=F) # Load themaster Pgenerosa gene list 
Pgen_GOterms <- Geoduck_annotation %>% dplyr::select(c('V1','V8')) # select only two columns - those with the gene IDs and those with the GO terms
Pgen_GOterms2 <- strsplit(Pgen_GOterms$V8, split = "; ") # create a string splitting by delimiter '; ' - view the data to see that this separates each GO term entry in the string
Pgen_GOterms2 <- data.frame(gene.ID = rep(Pgen_GOterms$V1, sapply(Pgen_GOterms2, length)), Go.terms = unlist(Pgen_GOterms2)) # create new dataframe 'Pgen_GOterms2' listing genes for each GO term (MUCH longer!)
Pgen_GOterms2 <- na.omit(Pgen_GOterms2) # ommit the NAs  - genes without GO annotation



d0.UP.05    <- read.csv("../Output/GO/DESeq2_goseq/Day0/GO.05.Day0_PrimEffect_Upregulated.csv")
d0.UP.05$Day = "Day0"
d0.UP.05$Dir = "Upregulated"
d0.DOWN.05  <- read.csv("../Output/GO/DESeq2_goseq/Day0/GO.05.Day0_PrimEffect_Downregulated.csv")
d0.DOWN.05$Day = "Day0"
d0.DOWN.05$Dir = "Downregulated"

d7.UP.05    <- read.csv("../Output/GO/DESeq2_goseq/Day7/GO.05.Day7_PrimEffect_Upregulated.csv")
d7.UP.05$Day = "Day7"
d7.UP.05$Dir = "Upregulated"
d7.DOWN.05  <- read.csv("../Output/GO/DESeq2_goseq/Day7/GO.05.Day7_PrimEffect_Downregulated.csv")
d7.DOWN.05$Day = "Day7"
d7.DOWN.05$Dir = "Downregulated"

d14.UP.05   <- read.csv("../Output/GO/DESeq2_goseq/Day14/GO.05.Day14_PrimEffect_Upregulated.csv")
d14.UP.05$Day = "Day14"
d14.UP.05$Dir = "Upregulated"
d14.DOWN.05 <- read.csv("../Output/GO/DESeq2_goseq/Day14/GO.05.Day14_PrimEffect_Downregulated.csv")
d14.DOWN.05$Day = "Day14"
d14.DOWN.05$Dir = "Downregulated"

d21.UP.05   <- read.csv("../Output/GO/DESeq2_goseq/Day21/GO.05.Day21_PrimEffect_Upregulated.csv")
d21.UP.05$Day = "Day21"
d21.UP.05$Dir = "Upregulated"
d21.DOWN.05 <- read.csv("../Output/GO/DESeq2_goseq/Day21/GO.05.Day21_PrimEffect_Downregulated.csv")
d21.DOWN.05$Day = "Day21"
d21.DOWN.05$Dir = "Downregulated"

# Master ALL DESeq2 significant modules with treatment
Master_goseq_results      <- rbind(d0.UP.05, d0.DOWN.05, 
                              d7.UP.05, d7.DOWN.05,
                              d14.UP.05, d14.DOWN.05, 
                              d21.UP.05, d21.DOWN.05)
# View(Master_goseq_results)


# call enriched GO terms and plot 
#  UPREGULATED GENES! 
# day 0 UPREGULATED
GO_d0.UP.05.a<-d0.UP.05$category[d0.UP.05$over_represented_pvalue<.05] # change twice here
GO_d0.UP.05<-data.frame(GO_d0.UP.05.a)
colnames(GO_d0.UP.05) <- c("category")
GO_d0.UP.05 <- merge(GO_d0.UP.05, d0.UP.05, by="category") # change here
GO_d0.UP.05 <- GO_d0.UP.05[order(-GO_d0.UP.05$numDEInCat),]
GO_d0.UP.05$term <- as.factor(GO_d0.UP.05$term)

d0_UP_MF <- GO_d0.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd0_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d0_UP'))
d0_UP_BP <- GO_d0.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd0_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d0_UP'))

# day 7 UPREGULATED
GO_d7.UP.05.a<-d7.UP.05$category[d7.UP.05$over_represented_pvalue<.05] # change twice here
GO_d7.UP.05<-data.frame(GO_d7.UP.05.a)
colnames(GO_d7.UP.05) <- c("category")
GO_d7.UP.05 <- merge(GO_d7.UP.05, d7.UP.05, by="category") # change here
GO_d7.UP.05 <- GO_d7.UP.05[order(-GO_d7.UP.05$numDEInCat),]
GO_d7.UP.05$term <- as.factor(GO_d7.UP.05$term)

d7_UP_MF <- GO_d7.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d7_UP'))
d7_UP_BP <- GO_d7.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d7_UP'))

# day 14 UPREGULATED
GO_d14.UP.05.a<-d14.UP.05$category[d14.UP.05$over_represented_pvalue<.05] # change twice here
GO_d14.UP.05<-data.frame(GO_d14.UP.05.a)
colnames(GO_d14.UP.05) <- c("category")
GO_d14.UP.05 <- merge(GO_d14.UP.05, d14.UP.05, by="category") # change here
GO_d14.UP.05 <- GO_d14.UP.05[order(-GO_d14.UP.05$numDEInCat),]
GO_d14.UP.05$term <- as.factor(GO_d14.UP.05$term)

d14_UP_MF <- GO_d14.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d14_UP'))
d14_UP_BP <- GO_d14.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d14_UP'))

# day 21 UPREGULATED
GO_d21.UP.05.a<-d21.UP.05$category[d21.UP.05$over_represented_pvalue<.05] # change twice here
GO_d21.UP.05<-data.frame(GO_d21.UP.05.a)
colnames(GO_d21.UP.05) <- c("category")
GO_d21.UP.05 <- merge(GO_d21.UP.05, d21.UP.05, by="category") # change here
GO_d21.UP.05 <- GO_d21.UP.05[order(-GO_d21.UP.05$numDEInCat),]
GO_d21.UP.05$term <- as.factor(GO_d21.UP.05$term)

d21_UP_MF <- GO_d21.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d21_UP'))
d21_UP_BP <- GO_d21.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d21_UP'))


#======================================================================= #
# bind the rows of MF and BP in the clustered modules (those twith primary treatment effect in same pattern/directionality)
MF_UP <- rbind(d0_UP_MF, d7_UP_MF, d14_UP_MF, d21_UP_MF) 
BP_UP <- rbind(d0_UP_BP, d7_UP_BP, d14_UP_BP, d21_UP_BP)


MF_UP$term <- gsub(",.*$", "", MF_UP$term)
MF_UP$Day = factor(MF_UP$Day, levels = c("Day0", "Day7", "Day14", "Day21"))
MF_UP_plot <- MF_UP  %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Molecular Function: Upregulated DEGs Primary Effect (higher by ambient history)") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Day ~., scales="free_y", ncol= 1, strip.position="right")
                  #facet_wrap(~ Day,nrow = 1)

BP_UP$term <- gsub(",.*$", "", BP_UP$term)
BP_UP$Day = factor(BP_UP$Day, levels = c("Day0", "Day7", "Day14", "Day21"))
BP_UP_plot <- BP_UP  %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Biological Process: Upregulated DEGs Primary Effect (higher by ambient history)") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Day ~., scales="free_y", ncol= 1, strip.position="right")



# Downregulated plots 
# day 0 DOWNREGULATED
GO_d0.DOWN.05.a<-d0.DOWN.05$category[d0.DOWN.05$over_represented_pvalue<.05] # change twice here
GO_d0.DOWN.05<-data.frame(GO_d0.DOWN.05.a)
colnames(GO_d0.DOWN.05) <- c("category")
GO_d0.DOWN.05 <- merge(GO_d0.DOWN.05, d0.DOWN.05, by="category") # change here
GO_d0.DOWN.05 <- GO_d0.DOWN.05[order(-GO_d0.DOWN.05$numDEInCat),]
GO_d0.DOWN.05$term <- as.factor(GO_d0.DOWN.05$term)

d0_DOWN_MF <- GO_d0.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd0_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d0_DOWN'))
d0_DOWN_BP <- GO_d0.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd0_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d0_DOWN'))

# day 7 DOWNREGULATED
GO_d7.DOWN.05.a<-d7.DOWN.05$category[d7.DOWN.05$over_represented_pvalue<.05] # change twice here
GO_d7.DOWN.05<-data.frame(GO_d7.DOWN.05.a)
colnames(GO_d7.DOWN.05) <- c("category")
GO_d7.DOWN.05 <- merge(GO_d7.DOWN.05, d7.DOWN.05, by="category") # change here
GO_d7.DOWN.05 <- GO_d7.DOWN.05[order(-GO_d7.DOWN.05$numDEInCat),]
GO_d7.DOWN.05$term <- as.factor(GO_d7.DOWN.05$term)

d7_DOWN_MF <- GO_d7.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d7_DOWN'))
d7_DOWN_BP <- GO_d7.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d7_DOWN'))

# day 14 DOWNREGULATED
GO_d14.DOWN.05.a<-d14.DOWN.05$category[d14.DOWN.05$over_represented_pvalue<.05] # change twice here
GO_d14.DOWN.05<-data.frame(GO_d14.DOWN.05.a)
colnames(GO_d14.DOWN.05) <- c("category")
GO_d14.DOWN.05 <- merge(GO_d14.DOWN.05, d14.DOWN.05, by="category") # change here
GO_d14.DOWN.05 <- GO_d14.DOWN.05[order(-GO_d14.DOWN.05$numDEInCat),]
GO_d14.DOWN.05$term <- as.factor(GO_d14.DOWN.05$term)

d14_DOWN_MF <- GO_d14.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d14_DOWN'))
d14_DOWN_BP <- GO_d14.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d14_DOWN'))

# day 21 DOWNREGULATED
GO_d21.DOWN.05.a<-d21.DOWN.05$category[d21.DOWN.05$over_represented_pvalue<.05] # change twice here
GO_d21.DOWN.05<-data.frame(GO_d21.DOWN.05.a)
colnames(GO_d21.DOWN.05) <- c("category")
GO_d21.DOWN.05 <- merge(GO_d21.DOWN.05, d21.DOWN.05, by="category") # change here
GO_d21.DOWN.05 <- GO_d21.DOWN.05[order(-GO_d21.DOWN.05$numDEInCat),]
GO_d21.DOWN.05$term <- as.factor(GO_d21.DOWN.05$term)

d21_DOWN_MF <- GO_d21.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d21_DOWN'))
d21_DOWN_BP <- GO_d21.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d21_DOWN'))


#======================================================================= #
# bind the rows of MF and BP in the clustered modules (those twith primary treatment effect in same pattern/directionality)
MF_DOWN <- rbind(d0_DOWN_MF, d7_DOWN_MF, d14_DOWN_MF, d21_DOWN_MF) 
BP_DOWN <- rbind(d0_DOWN_BP, d7_DOWN_BP, d14_DOWN_BP, d21_DOWN_BP)


MF_DOWN$term <- gsub(",.*$", "", MF_DOWN$term)
MF_DOWN$Day = factor(MF_DOWN$Day, levels = c("Day0", "Day7", "Day14", "Day21"))
MF_DOWN_plot <- MF_DOWN  %>% mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Molecular Function: Downregulated DEGs Primary Effect (higher by Moderate history)") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Day ~., scales="free_y", ncol= 1, strip.position="right")

BP_DOWN$term <- gsub(",.*$", "", BP_DOWN$term)
BP_DOWN$Day = factor(BP_DOWN$Day, levels = c("Day0", "Day7", "Day14", "Day21"))
BP_DOWN_plot <- BP_DOWN    %>%  mutate(term = fct_reorder(term, -log10(over_represented_pvalue))) %>%
                  #mutate(term = fct_reorder(term, Day)) %>%
                  ggplot(aes(x=term, y=-log10(over_represented_pvalue))) +
                  geom_segment( aes(x=term ,xend=term, y=1, yend=-log10(over_represented_pvalue)), color="grey95", size = 3) +
                  geom_point(size=3, shape = 15, colour = "grey70") +
                  coord_flip() +
                  theme(
                    panel.grid.minor.y = element_blank(),
                    panel.grid.major.y = element_blank(),
                    legend.position="bottom"
                  ) +
                  xlab("") +
                  ylab("") +
                  ggtitle("Biological Process: Downregulated DEGs Primary Effect (higher by Moderate history)") +
                  #geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
                  theme_bw() + #Set background color 
                  theme(panel.border = element_blank(), # Set border
                        panel.grid.major = element_blank(), #Set major gridlines
                        panel.grid.minor = element_blank(), #Set minor gridlines
                        axis.line = element_line(colour = "black"), #Set axes color
                        plot.background=element_blank()) + #Set the plot background #set title attributes
                  #geom_hline(yintercept=0.25, linetype="dashed",  color = "black", size=2) +
                  facet_wrap(Day ~., scales="free_y", ncol= 1, strip.position="right")

# MOLECULAR FUNCTION GRAPHS OUTPUT 

pdf(paste("../Output/GO/DESeq2_goseq/GOplot_DESeq2_PrimEffect_Upregulated_MF.pdf", sep =''), width=10, height=12)
print(ggarrange(MF_UP_plot,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()      


pdf(paste("../Output/GO/DESeq2_goseq/GOplot_DESeq2_PrimEffect_Downregulated_MF.pdf", sep =''), width=10, height=12)
print(ggarrange(MF_DOWN_plot,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()      


# BIOLOFICAL PROCESS GRAPHS OUTPUT

pdf(paste("../Output/GO/DESeq2_goseq/GOplot_DESeq2_PrimEffect_Upregulated_BP.pdf", sep =''), width=10, height=18)
print(ggarrange(BP_UP_plot,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()      


pdf(paste("../Output/GO/DESeq2_goseq/GOplot_DESeq2_PrimEffect_Downregulated_BP.pdf", sep =''), width=10, height=18)
print(ggarrange(BP_DOWN_plot,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()      


```

Plotting: Day 0
```{r plots_Day0.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d0.UP$category[GO_d0.UP$over_represented_pvalue<.05] # change twice here
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d0.UP, by="category") # change here
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- # 
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d0.DOWN$category[GO_d0.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d0.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d0.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d0.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 0 Upregulated DEGs") +
  geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 0 Downregulated DEGs") +
  geom_label(aes(x = 0.5, y = 1.5, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d0.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day0/GOplot_d0.Primary.pdf", GO.plot_d0.Primary, width = 21, height = 21, units = c("in"))
getwd()
```

Plotting: Day 7
```{r plots_Day7.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d7.UP$category[GO_d7.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d7.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d7.DOWN$category[GO_d7.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d7.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d7.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d7.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 7 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 2, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 7 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 2, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d7.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day7/GOplot_d7.Primary.pdf", GO.plot_d7.Primary, width = 21, height = 21, units = c("in"))
```

Plotting: Day 14
```{r plots_Day14.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d14.UP$category[GO_d14.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d14.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d14.DOWN$category[GO_d14.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d14.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d14.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d14.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 14 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 20, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 14 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 6, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d14.Primary<- grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day14/GOplot_d14.Primary.pdf", GO.plot_d14.Primary, width = 21, height = 21, units = c("in"))
```

Plotting: Day 21
```{r plots_Day21.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d21.UP$category[GO_d21.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d21.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d21.DOWN$category[GO_d21.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d21.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d21.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d21.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 21 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 7, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 21 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 3, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d21.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day21/GOplot_d21.Primary.pdf", GO.plot_d21.Primary, width = 21, height = 21, units = c("in"))
```

Plotting: Constant DEGs Days 7 14 21 affected by Primary treatment history
```{r Constant_All}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_Constant.UP$category[GO_Constant.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_Constant.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
head(UP_MF)

MFplot_UP <- UP_MF %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#00BFC4') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Mol Fxn: Constant DEGs UPREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

#Cellular Processes
CCplot_UP <- UP_CC %>% 
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#7CAE00') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Cell Comp: Constant DEGs UPREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

# Biological Processes 
BPplot_UP <- UP_BP %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#69b3a2') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Biol Proc: Constant DEGs UPREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 


# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_Constant.DOWN$category[GO_Constant.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_Constant.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Function
MFplot_DOWN <- DOWN_MF %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#00BFC4') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Mol Fxn: Constant DEGs DOWNREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% 
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#7CAE00') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Cell Comp: Constant DEGs DOWNREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#69b3a2') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Biol Proc: Constant DEGs DOWNREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 


# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(Contant_upreg)[1] # call num upregulated genes
num_down <-dim(Contant_dwnreg)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Constant Upregulated DEGs") +
  geom_label(aes(x = 1, y = 3, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Constant Downregulated DEGs") +
  geom_label(aes(x = 1, y = 3, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_Constant<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/GO/GOplot_Constant.Primary.pdf", GO.plot_Constant, width = 21, height = 21, units = c("in"))
```















### GO_MWU - not using this for the paper but did a bit of work here... Can review it for future data
Create all files for the GO_MWU.R script
output two-column csv of just the gene.id and log2fc 
```{r GO_MWU files - Primary_Treatment}
DE_d0.P <- data.frame(DE_d0.primary[,c(2,4)])
write.csv(DE_d0.P,"C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d0.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d7.P <- data.frame(DE_d7.primary[,c(2,4)])
write.csv(DE_d7.P,"C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d7.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d14.P <- data.frame(DE_d14.primary[,c(2,4)])
write.csv(DE_d14.P,"C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d14.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d21.P <- data.frame(DE_d21.primary[,c(2,4)])
write.csv(DE_d21.P,"C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d21.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

```

### Genewalk

link to tutorial: https://churchman.med.harvard.edu/genewalk/#tutorial
Citation (bioRxiv preprint): Ietswaart, R., Gyori, B. M., Bachman, J. A., Sorger, P. K., & Churchman, L. S. (2019). GeneWalk identifies relevant gene functions for a biological context using network representation learning. bioRxiv, 755579.
About: Genewalk is curently designed to perform function network analysis of DEGs with human and mouse identifiers

```{r prep genewalk files}
Geoduck.gene_HGNC.list <- Geoduck_annotation %>% dplyr::select(c("V1","V6")) # call the geoduck gene ID and the HGNC in the annotation file
colnames(Geoduck.gene_HGNC.list) <- c("gene.ID", "HGNC_symbol") # rename columns
# note that the HGNC column contains the HNC code with the model taxa as 'HUMAN, MOUSE, CHICK, etc.'
Geoduck.gene_HGNC.list$HGNC<- sapply(strsplit((Geoduck.gene_HGNC.list$HGNC), "_"), '[', 1) # this splits the HGNC by the delimiter "_" and calls just the HGNC code - note the model taxa is now absent from the dataframe! 


# change the col name of the gene.ID  in the DE datasets

# DAY 0 ------------------------------------------- #

# UPREGULATED

GW.d0.primary.UPREG <- merge((DE_d0.primary.UPREG[order(DE_d0.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d0.upreg.percHGNC <- ((dim(na.omit(GW.d0.primary.UPREG))[1]) / (dim(DE_d0.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d0.primary.DWNREG <- merge((DE_d0.primary.DWNREG[order(DE_d0.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d0.downreg.percHGNC <- ((dim(na.omit(GW.d0.primary.DWNREG))[1]) / (dim(DE_d0.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 7 ------------------------------------------- #

# UPREGULATED
GW.d7.primary.UPREG <- merge((DE_d7.primary.UPREG[order(DE_d7.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d7.upreg.percHGNC <- ((dim(na.omit(GW.d7.primary.UPREG))[1]) / (dim(DE_d7.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d7.primary.DWNREG <- merge((DE_d7.primary.DWNREG[order(DE_d7.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d7.downreg.percHGNC <- ((dim(na.omit(GW.d7.primary.DWNREG))[1]) / (dim(DE_d7.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 14 ------------------------------------------- #

# UPREGULATED
GW.d14.primary.UPREG <- merge((DE_d14.primary.UPREG[order(DE_d14.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d14.upreg.percHGNC <- ((dim(na.omit(GW.d14.primary.UPREG))[1]) / (dim(DE_d14.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d14.primary.DWNREG <- merge((DE_d14.primary.DWNREG[order(DE_d14.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC") # sort to order by greatest logfold change and merge with HGNC list
d14.downreg.percHGNC <- ((dim(na.omit(GW.d14.primary.DWNREG))[1]) / (dim(DE_d14.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 21 ------------------------------------------- #

# UPREGULATED
GW.d21.primary.UPREG <- merge((DE_d21.primary.UPREG[order(DE_d21.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d21.upreg.percHGNC <- ((dim(na.omit(GW.d21.primary.UPREG))[1]) / (dim(DE_d21.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d21.primary.DWNREG <- merge((DE_d21.primary.DWNREG[order(DE_d21.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d21.downreg.percHGNC <- ((dim(na.omit(GW.d21.primary.DWNREG))[1]) / (dim(DE_d21.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID


### OUTPUT Genewalk tables as csv wihtout header
GW.d0.primary.UPREG.om   <- na.omit(GW.d0.primary.UPREG)
GW.d0.primary.DWNREG.om  <- na.omit(GW.d0.primary.DWNREG)

GW.d7.primary.UPREG.om   <- na.omit(GW.d7.primary.UPREG)
GW.d7.primary.DWNREG.om  <- na.omit(GW.d7.primary.DWNREG) 

GW.d14.primary.UPREG.om  <- na.omit(GW.d14.primary.UPREG)
GW.d14.primary.DWNREG.om <- na.omit(GW.d14.primary.DWNREG)

GW.d21.primary.UPREG.om  <- na.omit(GW.d21.primary.UPREG)
GW.d21.primary.DWNREG.om <- na.omit(GW.d21.primary.DWNREG)

path_genewalk='C:/Users/samjg/Documents/Github_repositories/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/Genewalk_files/' # call path
# output files 
write.table(GW.d0.primary.UPREG.om,paste(path_genewalk,"Day0_upreg.csv", sep =''),sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d0.primary.DWNREG.om, paste(path_genewalk,"Day0_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)

write.table(GW.d7.primary.UPREG.om,paste(path_genewalk,"Day7_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d7.primary.DWNREG.om,paste(path_genewalk,"Day7_downreg.csv", sep =''), sep=",", col.names=FALSE, row.names=FALSE, quote=FALSE)

write.table(GW.d14.primary.UPREG.om,paste(path_genewalk,"Day14_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d14.primary.DWNREG.om,paste(path_genewalk,"Day14_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)

write.table(GW.d21.primary.UPREG.om,paste(path_genewalk,"Day21_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d21.primary.DWNREG.om,paste(path_genewalk,"Day21_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
```

```{r Genewalk_knitr_table}
GW_summarytable <- data.frame(matrix(nrow = 4, ncol = 8)) # create a new data table
colnames(GW_summarytable)<-c('Day', 'Treatment', 'DEGs_upreg', 'numHGNC_upreg', 'percHGNC_upreg', 'DEGs_dwnreg', 'numHGNC_dwnreg', 'percHGNC_dwnreg') 
GW_summarytable$Day <- c("0","7","14","21")
GW_summarytable$Treatment <- "Ambient_v_Moderate"
GW_summarytable$DEGs_upreg <- c((dim(DE_d0.primary.UPREG)[1]), (dim(DE_d7.primary.UPREG)[1]), (dim(DE_d14.primary.UPREG)[1]), (dim(DE_d21.primary.UPREG)[1]))
GW_summarytable$numHGNC_upreg <- c((dim(na.omit(GW.d0.primary.UPREG))[1]),(dim(na.omit(GW.d7.primary.UPREG))[1]),(dim(na.omit(GW.d14.primary.UPREG))[1]),(dim(na.omit(GW.d21.primary.UPREG))[1]))
GW_summarytable$percHGNC_upreg <- c(d0.upreg.percHGNC,d7.upreg.percHGNC,d14.upreg.percHGNC,d21.upreg.percHGNC)
GW_summarytable$DEGs_dwnreg <- c((dim(DE_d0.primary.DWNREG)[1]), (dim(DE_d7.primary.DWNREG)[1]), (dim(DE_d14.primary.DWNREG)[1]), (dim(DE_d21.primary.DWNREG)[1]))
GW_summarytable$numHGNC_dwnreg<- c((dim(na.omit(GW.d0.primary.DWNREG))[1]),(dim(na.omit(GW.d7.primary.DWNREG))[1]),(dim(na.omit(GW.d14.primary.DWNREG))[1]),(dim(na.omit(GW.d21.primary.DWNREG))[1]))
GW_summarytable$percHGNC_dwnreg <- c(d0.downreg.percHGNC,d7.downreg.percHGNC,d14.downreg.percHGNC,d21.downreg.percHGNC)

knitr::kable(GW_summarytable, caption = "Summary table of Genewalk files - total DEGs and mapping abund HGNC IDs")
# Write csv
write.csv(GW_summarytable,paste(path_genewalk,"Summary_HGNC_hits.csv"))
```

